--- a/net/minecraft/data/worldgen/features/CaveFeatures.java
+++ b/net/minecraft/data/worldgen/features/CaveFeatures.java
@@ -264,13 +_,13 @@
         WeightedStateProvider weightedStateProvider = new WeightedStateProvider(
             SimpleWeightedRandomList.<BlockState>builder()
                 .add(Blocks.CAVE_VINES_PLANT.defaultBlockState(), 4)
-                .add(Blocks.CAVE_VINES_PLANT.defaultBlockState().setValue(CaveVines.BERRIES, Boolean.valueOf(true)), 1)
+                .add(Blocks.CAVE_VINES_PLANT.defaultBlockState().setValue(CaveVines.BERRIES, true), 1)
         );
         RandomizedIntStateProvider randomizedIntStateProvider = new RandomizedIntStateProvider(
             new WeightedStateProvider(
                 SimpleWeightedRandomList.<BlockState>builder()
                     .add(Blocks.CAVE_VINES.defaultBlockState(), 4)
-                    .add(Blocks.CAVE_VINES.defaultBlockState().setValue(CaveVines.BERRIES, Boolean.valueOf(true)), 1)
+                    .add(Blocks.CAVE_VINES.defaultBlockState().setValue(CaveVines.BERRIES, true), 1)
             ),
             CaveVinesBlock.AGE,
             UniformInt.of(23, 25)
