--- a/net/minecraft/data/worldgen/features/TreeFeatures.java
+++ b/net/minecraft/data/worldgen/features/TreeFeatures.java
@@ -288,14 +_,14 @@
                 BlockStateProvider.simple(
                     Blocks.BROWN_MUSHROOM_BLOCK
                         .defaultBlockState()
-                        .setValue(HugeMushroomBlock.UP, Boolean.valueOf(true))
-                        .setValue(HugeMushroomBlock.DOWN, Boolean.valueOf(false))
+                        .setValue(HugeMushroomBlock.UP, true)
+                        .setValue(HugeMushroomBlock.DOWN, false)
                 ),
                 BlockStateProvider.simple(
                     Blocks.MUSHROOM_STEM
                         .defaultBlockState()
-                        .setValue(HugeMushroomBlock.UP, Boolean.valueOf(false))
-                        .setValue(HugeMushroomBlock.DOWN, Boolean.valueOf(false))
+                        .setValue(HugeMushroomBlock.UP, false)
+                        .setValue(HugeMushroomBlock.DOWN, false)
                 ),
                 3
             )
@@ -305,12 +_,12 @@
             HUGE_RED_MUSHROOM,
             Feature.HUGE_RED_MUSHROOM,
             new HugeMushroomFeatureConfiguration(
-                BlockStateProvider.simple(Blocks.RED_MUSHROOM_BLOCK.defaultBlockState().setValue(HugeMushroomBlock.DOWN, Boolean.valueOf(false))),
+                BlockStateProvider.simple(Blocks.RED_MUSHROOM_BLOCK.defaultBlockState().setValue(HugeMushroomBlock.DOWN, false)),
                 BlockStateProvider.simple(
                     Blocks.MUSHROOM_STEM
                         .defaultBlockState()
-                        .setValue(HugeMushroomBlock.UP, Boolean.valueOf(false))
-                        .setValue(HugeMushroomBlock.DOWN, Boolean.valueOf(false))
+                        .setValue(HugeMushroomBlock.UP, false)
+                        .setValue(HugeMushroomBlock.DOWN, false)
                 ),
                 2
             )
@@ -554,7 +_,7 @@
                             0,
                             new RandomizedIntStateProvider(
                                 BlockStateProvider.simple(
-                                    Blocks.MANGROVE_PROPAGULE.defaultBlockState().setValue(MangrovePropaguleBlock.HANGING, Boolean.valueOf(true))
+                                    Blocks.MANGROVE_PROPAGULE.defaultBlockState().setValue(MangrovePropaguleBlock.HANGING, true)
                                 ),
                                 MangrovePropaguleBlock.AGE,
                                 UniformInt.of(0, 4)
@@ -605,7 +_,7 @@
                             0,
                             new RandomizedIntStateProvider(
                                 BlockStateProvider.simple(
-                                    Blocks.MANGROVE_PROPAGULE.defaultBlockState().setValue(MangrovePropaguleBlock.HANGING, Boolean.valueOf(true))
+                                    Blocks.MANGROVE_PROPAGULE.defaultBlockState().setValue(MangrovePropaguleBlock.HANGING, true)
                                 ),
                                 MangrovePropaguleBlock.AGE,
                                 UniformInt.of(0, 4)
