--- a/net/minecraft/world/level/material/WaterFluid.java
+++ b/net/minecraft/world/level/material/WaterFluid.java
@@ -100,7 +_,7 @@
 
     @Override
     public BlockState createLegacyBlock(FluidState state) {
-        return Blocks.WATER.defaultBlockState().setValue(LiquidBlock.LEVEL, Integer.valueOf(getLegacyLevel(state)));
+        return Blocks.WATER.defaultBlockState().setValue(LiquidBlock.LEVEL, getLegacyLevel(state));
     }
 
     @Override
@@ -110,12 +_,12 @@
 
     @Override
     public int getDropOff(LevelReader level) {
-        return 1;
+        return level.getWorldBorder().world.plazmaConfig().block.waterDropOffSize; // Plazma - Configurable water dropOff block
     }
 
     @Override
     public int getTickDelay(LevelReader level) {
-        return 5;
+        return level.getWorldBorder().world.plazmaConfig().block.waterFlowingTick; // Plazma - Configurable water flowing speed
     }
 
     @Override
