--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -67,7 +_,7 @@
 
     public PistonBaseBlock(boolean isSticky, BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(EXTENDED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(EXTENDED, false));
         this.isSticky = isSticky;
     }
 
@@ -119,7 +_,7 @@
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
-        return this.defaultBlockState().setValue(FACING, context.getNearestLookingDirection().getOpposite()).setValue(EXTENDED, Boolean.valueOf(false));
+        return this.defaultBlockState().setValue(FACING, context.getNearestLookingDirection().getOpposite()).setValue(EXTENDED, false);
     }
 
     private void checkIfExtend(Level level, BlockPos pos, BlockState state) {
@@ -180,7 +_,7 @@
             return false;
         }
         // Paper end - Protect Bedrock and End Portal/Frames from being destroyed
-        BlockState blockState = state.setValue(EXTENDED, Boolean.valueOf(true));
+        BlockState blockState = state.setValue(EXTENDED, true);
         if (!level.isClientSide) {
             boolean neighborSignal = this.getNeighborSignal(level, pos, direction);
             if (neighborSignal && (id == 1 || id == 2)) {
