--- a/net/minecraft/world/item/equipment/trim/ArmorTrim.java
+++ b/net/minecraft/world/item/equipment/trim/ArmorTrim.java
@@ -21,7 +_,7 @@
         instance -> instance.group(
                 TrimMaterial.CODEC.fieldOf("material").forGetter(ArmorTrim::material),
                 TrimPattern.CODEC.fieldOf("pattern").forGetter(ArmorTrim::pattern),
-                Codec.BOOL.optionalFieldOf("show_in_tooltip", Boolean.valueOf(true)).forGetter(trim -> trim.showInTooltip)
+                Codec.BOOL.optionalFieldOf("show_in_tooltip", true).forGetter(trim -> trim.showInTooltip)
             )
             .apply(instance, ArmorTrim::new)
     );
