--- a/net/minecraft/server/level/ThreadedLevelLightEngine.java
+++ b/net/minecraft/server/level/ThreadedLevelLightEngine.java
@@ -55,7 +_,7 @@
             return;
         }
 
-        final Long ticketId = Long.valueOf(this.chunkWorkCounter.getAndIncrement());
+        final Long ticketId = this.chunkWorkCounter.getAndIncrement();
         final ChunkPos pos = new ChunkPos(chunkX, chunkZ);
         world.getChunkSource().addRegionTicket(ca.spottedleaf.moonrise.patches.starlight.light.StarLightInterface.CHUNK_WORK_TICKET, pos, ca.spottedleaf.moonrise.patches.starlight.light.StarLightInterface.REGION_LIGHT_TICKET_LEVEL, ticketId);
 
