--- a/ca/spottedleaf/dataconverter/minecraft/converters/entity/ConverterFlattenEntity.java
+++ b/ca/spottedleaf/dataconverter/minecraft/converters/entity/ConverterFlattenEntity.java
@@ -339,7 +_,7 @@
 
     public static int getBlockId(final String block) {
         final Integer ret = BLOCK_NAME_TO_ID.get(block);
-        return ret == null ? 0 : ret.intValue();
+        return ret == null ? 0 : ret;
     }
 
     @Override
