--- a/net/minecraft/world/level/redstone/ExperimentalRedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/ExperimentalRedstoneWireEvaluator.java
@@ -51,7 +_,7 @@
                     i1 |= 128;
                 }
 
-                level.setBlock(blockPos, blockState.setValue(RedStoneWireBlock.POWER, Integer.valueOf(i)), i1);
+                level.setBlock(blockPos, blockState.setValue(RedStoneWireBlock.POWER, i), i1);
             } else {
                 objectIterator.remove();
             }
