--- a/net/minecraft/world/inventory/AnvilMenu.java
+++ b/net/minecraft/world/inventory/AnvilMenu.java
@@ -157,8 +_,7 @@
             ItemStack itemStack = item.copy();
             ItemStack item1 = this.inputSlots.getItem(1);
             ItemEnchantments.Mutable mutable = new ItemEnchantments.Mutable(EnchantmentHelper.getEnchantmentsForCrafting(itemStack));
-            l += (long)item.getOrDefault(DataComponents.REPAIR_COST, Integer.valueOf(0)).intValue()
-                + item1.getOrDefault(DataComponents.REPAIR_COST, Integer.valueOf(0)).intValue();
+            l += (long) item.getOrDefault(DataComponents.REPAIR_COST, 0) + item1.getOrDefault(DataComponents.REPAIR_COST, 0);
             this.repairItemCountCost = 0;
             if (!item1.isEmpty()) {
                 boolean hasStoredEnchantments = item1.has(DataComponents.STORED_ENCHANTMENTS);
@@ -351,9 +_,9 @@
             }
 
             if (!itemStack.isEmpty()) {
-                int minxx = itemStack.getOrDefault(DataComponents.REPAIR_COST, Integer.valueOf(0));
-                if (minxx < item1.getOrDefault(DataComponents.REPAIR_COST, Integer.valueOf(0))) {
-                    minxx = item1.getOrDefault(DataComponents.REPAIR_COST, Integer.valueOf(0));
+                int minxx = itemStack.getOrDefault(DataComponents.REPAIR_COST, 0);
+                if (minxx < item1.getOrDefault(DataComponents.REPAIR_COST, 0)) {
+                    minxx = item1.getOrDefault(DataComponents.REPAIR_COST, 0);
                 }
 
                 if (i1 != i || i1 == 0) {
