--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -49,11 +_,11 @@
     public final boolean onlineMode = this.get("online-mode", true);
     public final boolean preventProxyConnections = this.get("prevent-proxy-connections", false);
     public final String serverIp = this.get("server-ip", "");
-    public final String serverName = this.get("server-name", "Unknown Server"); // Purpur - Bring back server name
+    public final String serverName = this.get("server-name", "A Plazma Server"); // Purpur - Bring back server name // Plazma - no
     public final boolean pvp = this.get("pvp", true);
     public final boolean allowFlight = this.get("allow-flight", false);
-    public final String motd = this.get("motd", "A Minecraft Server");
-    public final String bugReportLink = this.get("bug-report-link", "");
+    public final String motd = this.get("motd", "A Plazma Server"); // Plazma - Rebrand
+    public final String bugReportLink = this.get("bug-report-link", "https://github.com/PlazmaMC/PlazmaBukkit/issues"); // Plazma - Rebrand
     public final boolean forceGameMode = this.get("force-gamemode", false);
     public final boolean enforceWhitelist = this.get("enforce-whitelist", false);
     public final Difficulty difficulty = this.get(
