--- a/net/minecraft/server/dedicated/Settings.java
+++ b/net/minecraft/server/dedicated/Settings.java
@@ -216,7 +_,7 @@
     }
 
     protected int get(String key, int defaultValue) {
-        return this.get(key, wrapNumberDeserializer(Integer::parseInt), Integer.valueOf(defaultValue));
+        return this.get(key, wrapNumberDeserializer(Integer::parseInt), defaultValue);
     }
 
     protected Settings<T>.MutableValue<Integer> getMutable(String key, int defaultValue) {
