--- a/net/minecraft/world/item/AdventureModePredicate.java
+++ b/net/minecraft/world/item/AdventureModePredicate.java
@@ -27,7 +_,7 @@
                 ExtraCodecs.nonEmptyList(BlockPredicate.CODEC.listOf())
                     .fieldOf("predicates")
                     .forGetter(adventureModePredicate -> adventureModePredicate.predicates),
-                Codec.BOOL.optionalFieldOf("show_in_tooltip", Boolean.valueOf(true)).forGetter(AdventureModePredicate::showInTooltip)
+                Codec.BOOL.optionalFieldOf("show_in_tooltip", true).forGetter(AdventureModePredicate::showInTooltip)
             )
             .apply(instance, AdventureModePredicate::new)
     );
