--- a/net/minecraft/world/item/EnderEyeItem.java
+++ b/net/minecraft/world/item/EnderEyeItem.java
@@ -41,7 +_,7 @@
         } else if (level.isClientSide) {
             return InteractionResult.SUCCESS;
         } else {
-            BlockState blockState1 = blockState.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(true));
+            BlockState blockState1 = blockState.setValue(EndPortalFrameBlock.HAS_EYE, true);
             // Paper start
             if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(context.getPlayer(), clickedPos, blockState1)) {
                 return InteractionResult.PASS;
