--- a/net/minecraft/world/item/FireChargeItem.java
+++ b/net/minecraft/world/item/FireChargeItem.java
@@ -58,7 +_,7 @@
             }
             // CraftBukkit end
             this.playSound(level, clickedPos);
-            level.setBlockAndUpdate(clickedPos, blockState.setValue(BlockStateProperties.LIT, Boolean.valueOf(true)));
+            level.setBlockAndUpdate(clickedPos, blockState.setValue(BlockStateProperties.LIT, true));
             level.gameEvent(context.getPlayer(), GameEvent.BLOCK_CHANGE, clickedPos);
             flag = true;
         }
