--- a/net/minecraft/world/item/ItemStack.java
+++ b/net/minecraft/world/item/ItemStack.java
@@ -614,7 +_,7 @@
     }
 
     public int getMaxStackSize() {
-        return this.getOrDefault(DataComponents.MAX_STACK_SIZE, Integer.valueOf(1));
+        return this.getOrDefault(DataComponents.MAX_STACK_SIZE, 1);
     }
 
     public boolean isStackable() {
@@ -650,7 +_,7 @@
     // Purpur end - Add option to mend the most damaged equipment first
 
     public int getDamageValue() {
-        return Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, Integer.valueOf(0)), 0, this.getMaxDamage());
+        return Mth.clamp(this.getOrDefault(DataComponents.DAMAGE, 0), 0, this.getMaxDamage());
     }
 
     public void setDamageValue(int damage) {
@@ -658,7 +_,7 @@
     }
 
     public int getMaxDamage() {
-        return this.getOrDefault(DataComponents.MAX_DAMAGE, Integer.valueOf(0));
+        return this.getOrDefault(DataComponents.MAX_DAMAGE, 0);
     }
 
     public boolean isBroken() {
