--- a/net/minecraft/world/item/JukeboxPlayable.java
+++ b/net/minecraft/world/item/JukeboxPlayable.java
@@ -30,7 +_,7 @@
     public static final Codec<JukeboxPlayable> CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
                 EitherHolder.codec(Registries.JUKEBOX_SONG, JukeboxSong.CODEC).fieldOf("song").forGetter(JukeboxPlayable::song),
-                Codec.BOOL.optionalFieldOf("show_in_tooltip", Boolean.valueOf(true)).forGetter(JukeboxPlayable::showInTooltip)
+                Codec.BOOL.optionalFieldOf("show_in_tooltip", true).forGetter(JukeboxPlayable::showInTooltip)
             )
             .apply(instance, JukeboxPlayable::new)
     );
