--- a/net/minecraft/world/item/ShovelItem.java
+++ b/net/minecraft/world/item/ShovelItem.java
@@ -61,7 +_,7 @@
 
                 CampfireBlock.dowse(context.getPlayer(), level, clickedPos, blockState);
                 }; // Paper
-                blockState2 = blockState.setValue(CampfireBlock.LIT, Boolean.valueOf(false));
+                blockState2 = blockState.setValue(CampfireBlock.LIT, false);
             }
 
             if (blockState2 != null) {
