--- a/net/minecraft/world/level/block/entity/vault/VaultConfig.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultConfig.java
@@ -26,9 +_,9 @@
     static Codec<VaultConfig> CODEC = RecordCodecBuilder.<VaultConfig>create(
             instance -> instance.group(
                     ResourceKey.codec(Registries.LOOT_TABLE).lenientOptionalFieldOf("loot_table", DEFAULT.lootTable()).forGetter(VaultConfig::lootTable),
-                    Codec.DOUBLE.lenientOptionalFieldOf("activation_range", Double.valueOf(DEFAULT.activationRange())).forGetter(VaultConfig::activationRange),
+                    Codec.DOUBLE.lenientOptionalFieldOf("activation_range", DEFAULT.activationRange()).forGetter(VaultConfig::activationRange),
                     Codec.DOUBLE
-                        .lenientOptionalFieldOf("deactivation_range", Double.valueOf(DEFAULT.deactivationRange()))
+                        .lenientOptionalFieldOf("deactivation_range", DEFAULT.deactivationRange())
                         .forGetter(VaultConfig::deactivationRange),
                     ItemStack.lenientOptionalFieldOf("key_item").forGetter(VaultConfig::keyItem),
                     ResourceKey.codec(Registries.LOOT_TABLE)
