--- a/net/minecraft/world/level/block/entity/vault/VaultServerData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultServerData.java
@@ -21,11 +_,11 @@
         instance -> instance.group(
                 UUIDUtil.CODEC_LINKED_SET.lenientOptionalFieldOf("rewarded_players", Set.of()).forGetter(vaultServerData -> vaultServerData.rewardedPlayers),
                 Codec.LONG
-                    .lenientOptionalFieldOf("state_updating_resumes_at", Long.valueOf(0L))
+                    .lenientOptionalFieldOf("state_updating_resumes_at", 0L)
                     .forGetter(vaultServerData -> vaultServerData.stateUpdatingResumesAt),
                 ItemStack.CODEC.listOf().lenientOptionalFieldOf("items_to_eject", List.of()).forGetter(vaultServerData -> vaultServerData.itemsToEject),
                 Codec.INT
-                    .lenientOptionalFieldOf("total_ejections_needed", Integer.valueOf(0))
+                    .lenientOptionalFieldOf("total_ejections_needed", 0)
                     .forGetter(vaultServerData -> vaultServerData.totalEjectionsNeeded)
             )
             .apply(instance, VaultServerData::new)
