--- a/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
+++ b/net/minecraft/world/level/block/entity/vault/VaultSharedData.java
@@ -18,7 +_,7 @@
                 ItemStack.lenientOptionalFieldOf("display_item").forGetter(vaultSharedData -> vaultSharedData.displayItem),
                 UUIDUtil.CODEC_LINKED_SET.lenientOptionalFieldOf("connected_players", Set.of()).forGetter(vaultSharedData -> vaultSharedData.connectedPlayers),
                 Codec.DOUBLE
-                    .lenientOptionalFieldOf("connected_particles_range", Double.valueOf(VaultConfig.DEFAULT.deactivationRange()))
+                    .lenientOptionalFieldOf("connected_particles_range", VaultConfig.DEFAULT.deactivationRange())
                     .forGetter(vaultSharedData -> vaultSharedData.connectedParticlesRange)
             )
             .apply(instance, VaultSharedData::new)
