--- a/net/minecraft/network/chat/contents/NbtContents.java
+++ b/net/minecraft/network/chat/contents/NbtContents.java
@@ -27,7 +_,7 @@
     public static final MapCodec<NbtContents> CODEC = RecordCodecBuilder.mapCodec(
         instance -> instance.group(
                 Codec.STRING.fieldOf("nbt").forGetter(NbtContents::getNbtPath),
-                Codec.BOOL.lenientOptionalFieldOf("interpret", Boolean.valueOf(false)).forGetter(NbtContents::isInterpreting),
+                Codec.BOOL.lenientOptionalFieldOf("interpret", false).forGetter(NbtContents::isInterpreting),
                 ComponentSerialization.CODEC.lenientOptionalFieldOf("separator").forGetter(NbtContents::getSeparator),
                 DataSource.CODEC.forGetter(NbtContents::getDataSource)
             )
