--- a/net/minecraft/world/entity/animal/Bee.java
+++ b/net/minecraft/world/entity/animal/Bee.java
@@ -784,7 +_,7 @@
 
     public static boolean attractsBees(BlockState state) {
         return state.is(BlockTags.BEE_ATTRACTIVE)
-            && !state.getValueOrElse(BlockStateProperties.WATERLOGGED, Boolean.valueOf(false))
+            && !state.getValueOrElse(BlockStateProperties.WATERLOGGED, false)
             && (!state.is(Blocks.SUNFLOWER) || state.getValue(DoublePlantBlock.HALF) == DoubleBlockHalf.UPPER);
     }
 
@@ -1080,12 +_,12 @@
                         } else if (block instanceof StemBlock) {
                             int ageValue = blockState.getValue(StemBlock.AGE);
                             if (ageValue < 7) {
-                                blockState1 = blockState.setValue(StemBlock.AGE, Integer.valueOf(ageValue + 1));
+                                blockState1 = blockState.setValue(StemBlock.AGE, ageValue + 1);
                             }
                         } else if (blockState.is(Blocks.SWEET_BERRY_BUSH)) {
                             int ageValue = blockState.getValue(SweetBerryBushBlock.AGE);
                             if (ageValue < 3) {
-                                blockState1 = blockState.setValue(SweetBerryBushBlock.AGE, Integer.valueOf(ageValue + 1));
+                                blockState1 = blockState.setValue(SweetBerryBushBlock.AGE, ageValue + 1);
                             }
                         } else if (blockState.is(Blocks.CAVE_VINES) || blockState.is(Blocks.CAVE_VINES_PLANT)) {
                             BonemealableBlock bonemealableBlock = (BonemealableBlock)blockState.getBlock();
