--- a/net/minecraft/world/entity/animal/Fox.java
+++ b/net/minecraft/world/entity/animal/Fox.java
@@ -1068,7 +_,7 @@
 
         private void pickSweetBerries(BlockState state) {
             int ageValue = state.getValue(SweetBerryBushBlock.AGE);
-            state.setValue(SweetBerryBushBlock.AGE, Integer.valueOf(1));
+            state.setValue(SweetBerryBushBlock.AGE, 1);
             if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(Fox.this, this.blockPos, state.setValue(SweetBerryBushBlock.AGE, 1))) return; // CraftBukkit - call EntityChangeBlockEvent
             int i = 1 + Fox.this.level().random.nextInt(2) + (ageValue == 3 ? 1 : 0);
             ItemStack itemBySlot = Fox.this.getItemBySlot(EquipmentSlot.MAINHAND);
@@ -1082,7 +_,7 @@
             }
 
             Fox.this.playSound(SoundEvents.SWEET_BERRY_BUSH_PICK_BERRIES, 1.0F, 1.0F);
-            Fox.this.level().setBlock(this.blockPos, state.setValue(SweetBerryBushBlock.AGE, Integer.valueOf(1)), 2);
+            Fox.this.level().setBlock(this.blockPos, state.setValue(SweetBerryBushBlock.AGE, 1), 2);
             Fox.this.level().gameEvent(GameEvent.BLOCK_CHANGE, this.blockPos, GameEvent.Context.of(Fox.this));
         }
 
