--- a/net/minecraft/world/entity/animal/Rabbit.java
+++ b/net/minecraft/world/entity/animal/Rabbit.java
@@ -669,7 +_,7 @@
                         level.destroyBlock(blockPos, true, this.rabbit);
                     } else {
                         if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.rabbit, blockPos, blockState.setValue(CarrotBlock.AGE, ageValue - 1))) return; // CraftBukkit // Paper - fix wrong block state
-                        level.setBlock(blockPos, blockState.setValue(CarrotBlock.AGE, Integer.valueOf(ageValue - 1)), 2);
+                        level.setBlock(blockPos, blockState.setValue(CarrotBlock.AGE, ageValue - 1), 2);
                         level.gameEvent(GameEvent.BLOCK_CHANGE, blockPos, GameEvent.Context.of(this.rabbit));
                         level.levelEvent(2001, blockPos, Block.getId(blockState));
                     }
