--- a/net/minecraft/server/network/ServerGamePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerGamePacketListenerImpl.java
@@ -937,7 +_,7 @@
                     default -> Blocks.COMMAND_BLOCK.defaultBlockState();
                 };
                 BlockState blockState2 = blockState1.setValue(CommandBlock.FACING, direction)
-                    .setValue(CommandBlock.CONDITIONAL, Boolean.valueOf(packet.isConditional()));
+                    .setValue(CommandBlock.CONDITIONAL, packet.isConditional());
                 if (blockState2 != blockState) {
                     this.player.level().setBlock(pos, blockState2, 2);
                     blockEntity.setBlockState(blockState2);
