From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Sat, 13 Sep 2025 07:40:45 +0900
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index fb1fe2651e53a9bf46b3632c638e13eea9dcda93..816fbbd0fdb429b8c0d9c5bf1f9f33f0cf2fa8d2 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +25,23 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
+
+    // Plazma start
+    /**
+     * The brand id for Plazma.
+     */
+    Key BRAND_PLAZMA_ID = Key.key("plazmamc", "plazma");
+
+    /**
+     * Supported brands of the server.
+     */
+    java.util.List<Key> SUPPORTED_BRANDS = java.util.Arrays.asList(
+            BRAND_PAPER_ID,
+            BRAND_PURPUR_ID,
+            BRAND_PLAZMA_ID
+    );
+    // Plazma end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
