From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Sun, 14 Sep 2025 00:05:22 +0900
Subject: [PATCH] Rebrand


diff --git a/src/main/java/io/papermc/paper/ServerBuildInfo.java b/src/main/java/io/papermc/paper/ServerBuildInfo.java
index 652ff54e7c50412503725d628bfe72ed03059790..2ab1933ecb56378673d1b924e97924fd6c7d13fd 100644
--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -19,6 +19,36 @@ public interface ServerBuildInfo {
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
+    // Purpur start
+    /**
+     * The brand id for Purpur.
+     */
+    Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
+    // Purpur end
+
+    // Plazma start
+    /**
+     * The brand id for Plazma.
+     */
+    Key BRAND_PLAZMA_ID = Key.key("plazmamc", "plazma");
+
+    /**
+     * Gets the brand name of the server.
+     *
+     * @return the brand name of the server (e.g. "Paper")
+     */
+    String BRAND_NAME = buildInfo().brandName();
+
+    /**
+     * Supported brands of the server.
+     */
+    java.util.List<Key> SUPPORTED_BRANDS = java.util.Arrays.asList(
+            BRAND_PAPER_ID,
+            BRAND_PURPUR_ID,
+            BRAND_PLAZMA_ID
+    );
+    // Plazma end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
