--- a/net/minecraft/world/entity/item/FallingBlockEntity.java
+++ b/net/minecraft/world/entity/item/FallingBlockEntity.java
@@ -92,7 +_,7 @@
             pos.getY(),
             pos.getZ() + 0.5,
             blockState.hasProperty(BlockStateProperties.WATERLOGGED)
-                ? blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(false))
+                ? blockState.setValue(BlockStateProperties.WATERLOGGED, false)
                 : blockState
         );
         if (!CraftEventFactory.callEntityChangeBlockEvent(fallingBlockEntity, pos, blockState.getFluidState().createLegacyBlock())) return fallingBlockEntity; // CraftBukkit
@@ -202,7 +_,7 @@
                             if (canBeReplaced && flag3) {
                                 if (this.blockState.hasProperty(BlockStateProperties.WATERLOGGED)
                                     && this.level().getFluidState(blockPos).getType() == Fluids.WATER) {
-                                    this.blockState = this.blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true));
+                                    this.blockState = this.blockState.setValue(BlockStateProperties.WATERLOGGED, true);
                                 }
 
                                 // CraftBukkit start
