--- a/net/minecraft/world/level/block/AbstractCandleBlock.java
+++ b/net/minecraft/world/level/block/AbstractCandleBlock.java
@@ -100,7 +_,7 @@
     }
 
     private static void setLit(LevelAccessor level, BlockState state, BlockPos pos, boolean lit) {
-        level.setBlock(pos, state.setValue(LIT, Boolean.valueOf(lit)), 11);
+        level.setBlock(pos, state.setValue(LIT, lit), 11);
     }
 
     @Override
