--- a/net/minecraft/world/level/block/AbstractSkullBlock.java
+++ b/net/minecraft/world/level/block/AbstractSkullBlock.java
@@ -24,7 +_,7 @@
     public AbstractSkullBlock(SkullBlock.Type type, BlockBehaviour.Properties properties) {
         super(properties);
         this.type = type;
-        this.registerDefaultState(this.stateDefinition.any().setValue(POWERED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(POWERED, false));
     }
 
     @Override
@@ -67,7 +_,7 @@
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
-        return this.defaultBlockState().setValue(POWERED, Boolean.valueOf(context.getLevel().hasNeighborSignal(context.getClickedPos())));
+        return this.defaultBlockState().setValue(POWERED, context.getLevel().hasNeighborSignal(context.getClickedPos()));
     }
 
     @Override
@@ -75,7 +_,7 @@
         if (!level.isClientSide) {
             boolean hasNeighborSignal = level.hasNeighborSignal(pos);
             if (hasNeighborSignal != state.getValue(POWERED)) {
-                level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(hasNeighborSignal)), 2);
+                level.setBlock(pos, state.setValue(POWERED, hasNeighborSignal), 2);
             }
         }
     }
