--- a/net/minecraft/world/level/block/AttachedStemBlock.java
+++ b/net/minecraft/world/level/block/AttachedStemBlock.java
@@ -84,7 +_,7 @@
         if (!neighborState.is(this.fruit) && direction == state.getValue(FACING)) {
             Optional<Block> optional = level.registryAccess().lookupOrThrow(Registries.BLOCK).getOptional(this.stem);
             if (optional.isPresent()) {
-                return optional.get().defaultBlockState().trySetValue(StemBlock.AGE, Integer.valueOf(7));
+                return optional.get().defaultBlockState().trySetValue(StemBlock.AGE, 7);
             }
         }
 
