--- a/net/minecraft/world/level/block/BambooStalkBlock.java
+++ b/net/minecraft/world/level/block/BambooStalkBlock.java
@@ -52,7 +_,7 @@
     public BambooStalkBlock(BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(AGE, Integer.valueOf(0)).setValue(LEAVES, BambooLeaves.NONE).setValue(STAGE, Integer.valueOf(0))
+            this.stateDefinition.any().setValue(AGE, 0).setValue(LEAVES, BambooLeaves.NONE).setValue(STAGE, 0)
         );
     }
 
@@ -99,10 +_,10 @@
             BlockState blockState = context.getLevel().getBlockState(context.getClickedPos().below());
             if (blockState.is(BlockTags.BAMBOO_PLANTABLE_ON)) {
                 if (blockState.is(Blocks.BAMBOO_SAPLING)) {
-                    return this.defaultBlockState().setValue(AGE, Integer.valueOf(0));
+                    return this.defaultBlockState().setValue(AGE, 0);
                 } else if (blockState.is(Blocks.BAMBOO)) {
                     int i = blockState.getValue(AGE) > 0 ? 1 : 0;
-                    return this.defaultBlockState().setValue(AGE, Integer.valueOf(i));
+                    return this.defaultBlockState().setValue(AGE, i);
                 } else {
                     BlockState blockState1 = context.getLevel().getBlockState(context.getClickedPos().above());
                     return blockState1.is(Blocks.BAMBOO)
