--- a/net/minecraft/world/level/block/BaseRailBlock.java
+++ b/net/minecraft/world/level/block/BaseRailBlock.java
@@ -143,7 +_,7 @@
         BlockState blockState = super.defaultBlockState();
         Direction horizontalDirection = context.getHorizontalDirection();
         boolean flag1 = horizontalDirection == Direction.EAST || horizontalDirection == Direction.WEST;
-        return blockState.setValue(this.getShapeProperty(), flag1 ? RailShape.EAST_WEST : RailShape.NORTH_SOUTH).setValue(WATERLOGGED, Boolean.valueOf(flag));
+        return blockState.setValue(this.getShapeProperty(), flag1 ? RailShape.EAST_WEST : RailShape.NORTH_SOUTH).setValue(WATERLOGGED, flag);
     }
 
     public abstract Property<RailShape> getShapeProperty();
