--- a/net/minecraft/world/level/block/BaseCoralWallFanBlock.java
+++ b/net/minecraft/world/level/block/BaseCoralWallFanBlock.java
@@ -43,7 +_,7 @@
 
     protected BaseCoralWallFanBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, Boolean.valueOf(true)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, true));
     }
 
     @Override
