--- a/net/minecraft/world/level/block/BeehiveBlock.java
+++ b/net/minecraft/world/level/block/BeehiveBlock.java
@@ -77,7 +_,7 @@
 
     public BeehiveBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(HONEY_LEVEL, Integer.valueOf(0)).setValue(FACING, Direction.NORTH));
+        this.registerDefaultState(this.stateDefinition.any().setValue(HONEY_LEVEL, 0).setValue(FACING, Direction.NORTH));
     }
 
     @Override
@@ -206,7 +_,7 @@
     }
 
     public void resetHoneyLevel(Level level, BlockState state, BlockPos pos) {
-        level.setBlock(pos, state.setValue(HONEY_LEVEL, Integer.valueOf(0)), 3);
+        level.setBlock(pos, state.setValue(HONEY_LEVEL, 0), 3);
     }
 
     @Override
