--- a/net/minecraft/world/level/block/BellBlock.java
+++ b/net/minecraft/world/level/block/BellBlock.java
@@ -67,7 +_,7 @@
     public BellBlock(BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(ATTACHMENT, BellAttachType.FLOOR).setValue(POWERED, Boolean.valueOf(false))
+            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(ATTACHMENT, BellAttachType.FLOOR).setValue(POWERED, false)
         );
     }
 
@@ -79,7 +_,7 @@
                 this.attemptToRing(level, pos, null);
             }
 
-            level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(hasNeighborSignal)), 3);
+            level.setBlock(pos, state.setValue(POWERED, hasNeighborSignal), 3);
         }
     }
 
