--- a/net/minecraft/world/level/block/BigDripleafBlock.java
+++ b/net/minecraft/world/level/block/BigDripleafBlock.java
@@ -88,7 +_,7 @@
     protected BigDripleafBlock(BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(WATERLOGGED, Boolean.valueOf(false)).setValue(FACING, Direction.NORTH).setValue(TILT, Tilt.NONE)
+            this.stateDefinition.any().setValue(WATERLOGGED, false).setValue(FACING, Direction.NORTH).setValue(TILT, Tilt.NONE)
         );
         this.shapesCache = this.getShapeForEachState(BigDripleafBlock::calculateShape);
     }
@@ -129,7 +_,7 @@
     protected static boolean place(LevelAccessor level, BlockPos pos, FluidState fluidState, Direction direction) {
         BlockState blockState = Blocks.BIG_DRIPLEAF
             .defaultBlockState()
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidState.isSourceOfType(Fluids.WATER)))
+            .setValue(WATERLOGGED, fluidState.isSourceOfType(Fluids.WATER))
             .setValue(FACING, direction);
         return level.setBlock(pos, blockState, 3);
     }
@@ -306,7 +_,7 @@
         FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
         boolean flag = blockState.is(Blocks.BIG_DRIPLEAF) || blockState.is(Blocks.BIG_DRIPLEAF_STEM);
         return this.defaultBlockState()
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidState.isSourceOfType(Fluids.WATER)))
+            .setValue(WATERLOGGED, fluidState.isSourceOfType(Fluids.WATER))
             .setValue(FACING, flag ? blockState.getValue(FACING) : context.getHorizontalDirection().getOpposite());
     }
 
