--- a/net/minecraft/world/level/block/BigDripleafStemBlock.java
+++ b/net/minecraft/world/level/block/BigDripleafStemBlock.java
@@ -40,7 +_,7 @@
 
     protected BigDripleafStemBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, Boolean.valueOf(false)).setValue(FACING, Direction.NORTH));
+        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, false).setValue(FACING, Direction.NORTH));
     }
 
     @Override
@@ -79,7 +_,7 @@
     protected static boolean place(LevelAccessor level, BlockPos pos, FluidState fluidState, Direction direction) {
         BlockState blockState = Blocks.BIG_DRIPLEAF_STEM
             .defaultBlockState()
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidState.isSourceOfType(Fluids.WATER)))
+            .setValue(WATERLOGGED, fluidState.isSourceOfType(Fluids.WATER))
             .setValue(FACING, direction);
         return level.setBlock(pos, blockState, 3);
     }
