--- a/net/minecraft/world/level/block/BrewingStandBlock.java
+++ b/net/minecraft/world/level/block/BrewingStandBlock.java
@@ -44,9 +_,9 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(HAS_BOTTLE[0], Boolean.valueOf(false))
-                .setValue(HAS_BOTTLE[1], Boolean.valueOf(false))
-                .setValue(HAS_BOTTLE[2], Boolean.valueOf(false))
+                .setValue(HAS_BOTTLE[0], false)
+                .setValue(HAS_BOTTLE[1], false)
+                .setValue(HAS_BOTTLE[2], false)
         );
     }
 
