--- a/net/minecraft/world/level/block/BrushableBlock.java
+++ b/net/minecraft/world/level/block/BrushableBlock.java
@@ -51,7 +_,7 @@
         this.turnsInto = turnsInto;
         this.brushSound = brushSound;
         this.brushCompletedSound = brushCompletedSound;
-        this.registerDefaultState(this.stateDefinition.any().setValue(DUSTED, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(DUSTED, 0));
     }
 
     @Override
