--- a/net/minecraft/world/level/block/BuddingAmethystBlock.java
+++ b/net/minecraft/world/level/block/BuddingAmethystBlock.java
@@ -43,7 +_,7 @@
             if (block != null) {
                 BlockState blockState1 = block.defaultBlockState()
                     .setValue(AmethystClusterBlock.FACING, direction)
-                    .setValue(AmethystClusterBlock.WATERLOGGED, Boolean.valueOf(blockState.getFluidState().getType() == Fluids.WATER));
+                    .setValue(AmethystClusterBlock.WATERLOGGED, blockState.getFluidState().getType() == Fluids.WATER);
                 // Paper start - Have Amethyst throw both spread and grow events
                 if (block == Blocks.SMALL_AMETHYST_BUD) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, pos, blockPos, blockState1, 3); // CraftBukkit
