--- a/net/minecraft/world/level/block/ButtonBlock.java
+++ b/net/minecraft/world/level/block/ButtonBlock.java
@@ -76,7 +_,7 @@
         super(properties.sound(type.soundType()));
         this.type = type;
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(POWERED, Boolean.valueOf(false)).setValue(FACE, AttachFace.WALL)
+            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(POWERED, false).setValue(FACE, AttachFace.WALL)
         );
         this.ticksToStayPressed = ticksToStayPressed;
     }
@@ -142,7 +_,7 @@
     }
 
     public void press(BlockState state, Level level, BlockPos pos, @Nullable Player player) {
-        level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(true)), 3);
+        level.setBlock(pos, state.setValue(POWERED, true), 3);
         this.updateNeighbours(state, level, pos);
         level.scheduleTick(pos, this, this.ticksToStayPressed);
         this.playSound(player, level, pos, true);
@@ -229,7 +_,7 @@
                 return;
             }
             // CraftBukkit end
-            level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(flag)), 3);
+            level.setBlock(pos, state.setValue(POWERED, flag), 3);
             this.updateNeighbours(state, level, pos);
             this.playSound(null, level, pos, flag);
             level.gameEvent(abstractArrow, flag ? GameEvent.BLOCK_ACTIVATE : GameEvent.BLOCK_DEACTIVATE, pos);
