--- a/net/minecraft/world/level/block/CakeBlock.java
+++ b/net/minecraft/world/level/block/CakeBlock.java
@@ -53,7 +_,7 @@
 
     protected CakeBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(BITES, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(BITES, 0));
     }
 
     @Override
@@ -128,7 +_,7 @@
             // Paper - move up
             level.gameEvent(player, GameEvent.EAT, pos);
             if (bitesValue < 6) {
-                level.setBlock(pos, state.setValue(BITES, Integer.valueOf(bitesValue + 1)), 3);
+                level.setBlock(pos, state.setValue(BITES, bitesValue + 1), 3);
             } else {
                 level.removeBlock(pos, false);
                 level.gameEvent(player, GameEvent.BLOCK_DESTROY, pos);
