--- a/net/minecraft/world/level/block/CandleCakeBlock.java
+++ b/net/minecraft/world/level/block/CandleCakeBlock.java
@@ -53,7 +_,7 @@
 
     protected CandleCakeBlock(Block candleBlock, BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(LIT, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(LIT, false));
         if (candleBlock instanceof CandleBlock candleBlock1) {
             BY_CANDLE.put(candleBlock1, this);
             this.candleBlock = candleBlock1;
