--- a/net/minecraft/world/level/block/CandleBlock.java
+++ b/net/minecraft/world/level/block/CandleBlock.java
@@ -67,9 +_,9 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(CANDLES, Integer.valueOf(1))
-                .setValue(LIT, Boolean.valueOf(false))
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(CANDLES, 1)
+                .setValue(LIT, false)
+                .setValue(WATERLOGGED, false)
         );
     }
 
@@ -99,7 +_,7 @@
         } else {
             FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
             boolean flag = fluidState.getType() == Fluids.WATER;
-            return super.getStateForPlacement(context).setValue(WATERLOGGED, Boolean.valueOf(flag));
+            return super.getStateForPlacement(context).setValue(WATERLOGGED, flag);
         }
     }
 
@@ -149,7 +_,7 @@
     @Override
     public boolean placeLiquid(LevelAccessor level, BlockPos pos, BlockState state, FluidState fluidState) {
         if (!state.getValue(WATERLOGGED) && fluidState.getType() == Fluids.WATER) {
-            BlockState blockState = state.setValue(WATERLOGGED, Boolean.valueOf(true));
+            BlockState blockState = state.setValue(WATERLOGGED, true);
             if (state.getValue(LIT)) {
                 extinguish(null, blockState, level, pos);
             } else {
