--- a/net/minecraft/world/level/block/ChainBlock.java
+++ b/net/minecraft/world/level/block/ChainBlock.java
@@ -36,7 +_,7 @@
 
     public ChainBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, Boolean.valueOf(false)).setValue(AXIS, Direction.Axis.Y));
+        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, false).setValue(AXIS, Direction.Axis.Y));
     }
 
     @Override
@@ -57,7 +_,7 @@
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
         boolean flag = fluidState.getType() == Fluids.WATER;
-        return super.getStateForPlacement(context).setValue(WATERLOGGED, Boolean.valueOf(flag));
+        return super.getStateForPlacement(context).setValue(WATERLOGGED, flag);
     }
 
     @Override
