--- a/net/minecraft/world/level/block/ChestBlock.java
+++ b/net/minecraft/world/level/block/ChestBlock.java
@@ -156,7 +_,7 @@
     protected ChestBlock(Supplier<BlockEntityType<? extends ChestBlockEntity>> blockEntityType, BlockBehaviour.Properties properties) {
         super(properties, blockEntityType);
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(TYPE, ChestType.SINGLE).setValue(WATERLOGGED, Boolean.valueOf(false))
+            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(TYPE, ChestType.SINGLE).setValue(WATERLOGGED, false)
         );
     }
 
@@ -249,7 +_,7 @@
         return this.defaultBlockState()
             .setValue(FACING, opposite)
             .setValue(TYPE, chestType)
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+            .setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
     }
 
     @Override
