--- a/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
+++ b/net/minecraft/world/level/block/ChiseledBookShelfBlock.java
@@ -55,7 +_,7 @@
         BlockState blockState = this.stateDefinition.any().setValue(HorizontalDirectionalBlock.FACING, Direction.NORTH);
 
         for (BooleanProperty booleanProperty : SLOT_OCCUPIED_PROPERTIES) {
-            blockState = blockState.setValue(booleanProperty, Boolean.valueOf(false));
+            blockState = blockState.setValue(booleanProperty, false);
         }
 
         this.registerDefaultState(blockState);
