--- a/net/minecraft/world/level/block/ChorusPlantBlock.java
+++ b/net/minecraft/world/level/block/ChorusPlantBlock.java
@@ -27,12 +_,12 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(NORTH, Boolean.valueOf(false))
-                .setValue(EAST, Boolean.valueOf(false))
-                .setValue(SOUTH, Boolean.valueOf(false))
-                .setValue(WEST, Boolean.valueOf(false))
-                .setValue(UP, Boolean.valueOf(false))
-                .setValue(DOWN, Boolean.valueOf(false))
+                .setValue(NORTH, false)
+                .setValue(EAST, false)
+                .setValue(SOUTH, false)
+                .setValue(WEST, false)
+                .setValue(UP, false)
+                .setValue(DOWN, false)
         );
     }
 
@@ -50,12 +_,12 @@
         BlockState blockState4 = level.getBlockState(pos.south());
         BlockState blockState5 = level.getBlockState(pos.west());
         Block block = state.getBlock();
-        return state.trySetValue(DOWN, Boolean.valueOf(blockState.is(block) || blockState.is(Blocks.CHORUS_FLOWER) || blockState.is(Blocks.END_STONE)))
-            .trySetValue(UP, Boolean.valueOf(blockState1.is(block) || blockState1.is(Blocks.CHORUS_FLOWER)))
-            .trySetValue(NORTH, Boolean.valueOf(blockState2.is(block) || blockState2.is(Blocks.CHORUS_FLOWER)))
-            .trySetValue(EAST, Boolean.valueOf(blockState3.is(block) || blockState3.is(Blocks.CHORUS_FLOWER)))
-            .trySetValue(SOUTH, Boolean.valueOf(blockState4.is(block) || blockState4.is(Blocks.CHORUS_FLOWER)))
-            .trySetValue(WEST, Boolean.valueOf(blockState5.is(block) || blockState5.is(Blocks.CHORUS_FLOWER)));
+        return state.trySetValue(DOWN, blockState.is(block) || blockState.is(Blocks.CHORUS_FLOWER) || blockState.is(Blocks.END_STONE))
+            .trySetValue(UP, blockState1.is(block) || blockState1.is(Blocks.CHORUS_FLOWER))
+            .trySetValue(NORTH, blockState2.is(block) || blockState2.is(Blocks.CHORUS_FLOWER))
+            .trySetValue(EAST, blockState3.is(block) || blockState3.is(Blocks.CHORUS_FLOWER))
+            .trySetValue(SOUTH, blockState4.is(block) || blockState4.is(Blocks.CHORUS_FLOWER))
+            .trySetValue(WEST, blockState5.is(block) || blockState5.is(Blocks.CHORUS_FLOWER));
     }
 
     @Override
@@ -77,7 +_,7 @@
             boolean flag = neighborState.is(this)
                 || neighborState.is(Blocks.CHORUS_FLOWER)
                 || direction == Direction.DOWN && neighborState.is(Blocks.END_STONE);
-            return state.setValue(PROPERTY_BY_DIRECTION.get(direction), Boolean.valueOf(flag));
+            return state.setValue(PROPERTY_BY_DIRECTION.get(direction), flag);
         }
     }
 
