--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -42,7 +_,7 @@
     protected ChorusFlowerBlock(Block plant, BlockBehaviour.Properties properties) {
         super(properties);
         this.plant = plant;
-        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, 0));
     }
 
     @Override
@@ -125,14 +_,14 @@
                         level.setBlock(pos, ChorusPlantBlock.getStateWithConnections(level, pos, this.plant.defaultBlockState()), 2);
                     } else {
                         // CraftBukkit start - add event
-                        if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
+                        if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, 5), 2)) {
                         this.placeDeadFlower(level, pos);
                         }
                         // CraftBukkit end
                     }
                 } else {
                     // CraftBukkit start - add event
-                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, Integer.valueOf(5)), 2)) {
+                    if (org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, this.defaultBlockState().setValue(ChorusFlowerBlock.AGE, 5), 2)) {
                     this.placeDeadFlower(level, pos);
                     }
                     // CraftBukkit end
@@ -142,12 +_,12 @@
     }
 
     private void placeGrownFlower(Level level, BlockPos pos, int age) {
-        level.setBlock(pos, this.defaultBlockState().setValue(AGE, Integer.valueOf(age)), 2);
+        level.setBlock(pos, this.defaultBlockState().setValue(AGE, age), 2);
         level.levelEvent(1033, pos, 0);
     }
 
     private void placeDeadFlower(Level level, BlockPos pos) {
-        level.setBlock(pos, this.defaultBlockState().setValue(AGE, Integer.valueOf(5)), 2);
+        level.setBlock(pos, this.defaultBlockState().setValue(AGE, 5), 2);
         level.levelEvent(1034, pos, 0);
     }
 
@@ -265,7 +_,7 @@
         }
 
         if (!flag) {
-            level.setBlock(branchPos.above(i), Blocks.CHORUS_FLOWER.defaultBlockState().setValue(AGE, Integer.valueOf(5)), 2);
+            level.setBlock(branchPos.above(i), Blocks.CHORUS_FLOWER.defaultBlockState().setValue(AGE, 5), 2);
         }
     }
 
