--- a/net/minecraft/world/level/block/ComposterBlock.java
+++ b/net/minecraft/world/level/block/ComposterBlock.java
@@ -186,7 +_,7 @@
 
     public ComposterBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(LEVEL, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(LEVEL, 0));
     }
 
     public static void handleFill(Level level, BlockPos pos, boolean success) {
@@ -344,7 +_,7 @@
     }
 
     static BlockState empty(@Nullable Entity entity, BlockState state, LevelAccessor level, BlockPos pos) {
-        BlockState blockState = state.setValue(LEVEL, Integer.valueOf(0));
+        BlockState blockState = state.setValue(LEVEL, 0);
         level.setBlock(pos, blockState, 3);
         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(entity, blockState));
         return blockState;
@@ -377,7 +_,7 @@
             return state;
         } else {
             int i = levelValue + 1;
-            BlockState blockState = state.setValue(LEVEL, Integer.valueOf(i));
+            BlockState blockState = state.setValue(LEVEL, i);
             // Paper start - move the EntityChangeBlockEvent here to avoid conflict later for the compost events
             if (entity != null && !org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(entity, pos, blockState)) {
                 return null;
