--- a/net/minecraft/world/level/block/CopperBulbBlock.java
+++ b/net/minecraft/world/level/block/CopperBulbBlock.java
@@ -26,7 +_,7 @@
 
     public CopperBulbBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.defaultBlockState().setValue(LIT, Boolean.valueOf(false)).setValue(POWERED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.defaultBlockState().setValue(LIT, false).setValue(POWERED, false));
     }
 
     @Override
@@ -52,7 +_,7 @@
                 level.playSound(null, pos, blockState.getValue(LIT) ? SoundEvents.COPPER_BULB_TURN_ON : SoundEvents.COPPER_BULB_TURN_OFF, SoundSource.BLOCKS);
             }
 
-            level.setBlock(pos, blockState.setValue(POWERED, Boolean.valueOf(hasNeighborSignal)), 3);
+            level.setBlock(pos, blockState.setValue(POWERED, hasNeighborSignal), 3);
         }
     }
 
