--- a/net/minecraft/world/level/block/CreakingHeartBlock.java
+++ b/net/minecraft/world/level/block/CreakingHeartBlock.java
@@ -44,7 +_,7 @@
     protected CreakingHeartBlock(BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(
-            this.defaultBlockState().setValue(AXIS, Direction.Axis.Y).setValue(ACTIVE, Boolean.valueOf(false)).setValue(NATURAL, Boolean.valueOf(false))
+            this.defaultBlockState().setValue(AXIS, Direction.Axis.Y).setValue(ACTIVE, false).setValue(NATURAL, false)
         );
     }
 
@@ -96,7 +_,7 @@
     private static BlockState updateState(BlockState state, LevelReader level, BlockPos pos) {
         boolean hasRequiredLogs = hasRequiredLogs(state, level, pos);
         boolean flag = !state.getValue(ACTIVE);
-        return hasRequiredLogs && flag ? state.setValue(ACTIVE, Boolean.valueOf(true)) : state;
+        return hasRequiredLogs && flag ? state.setValue(ACTIVE, true) : state;
     }
 
     public static boolean hasRequiredLogs(BlockState state, LevelReader level, BlockPos pos) {
