--- a/net/minecraft/world/level/block/DaylightDetectorBlock.java
+++ b/net/minecraft/world/level/block/DaylightDetectorBlock.java
@@ -38,7 +_,7 @@
 
     public DaylightDetectorBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(POWER, Integer.valueOf(0)).setValue(INVERTED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(POWER, 0).setValue(INVERTED, false));
     }
 
     @Override
@@ -71,7 +_,7 @@
         i = Mth.clamp(i, 0, 15);
         if (state.getValue(POWER) != i) {
             i = org.bukkit.craftbukkit.event.CraftEventFactory.callRedstoneChange(level, pos, state.getValue(DaylightDetectorBlock.POWER), i).getNewCurrent(); // CraftBukkit - Call BlockRedstoneEvent
-            level.setBlock(pos, state.setValue(POWER, Integer.valueOf(i)), 3);
+            level.setBlock(pos, state.setValue(POWER, i), 3);
         }
     }
 
