--- a/net/minecraft/world/level/block/DetectorRailBlock.java
+++ b/net/minecraft/world/level/block/DetectorRailBlock.java
@@ -41,9 +_,9 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(POWERED, Boolean.valueOf(false))
+                .setValue(POWERED, false)
                 .setValue(SHAPE, RailShape.NORTH_SOUTH)
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(WATERLOGGED, false)
         );
     }
 
@@ -104,7 +_,7 @@
             }
             // CraftBukkit end
             if (flag && !poweredValue) {
-                BlockState blockState = state.setValue(POWERED, Boolean.valueOf(true));
+                BlockState blockState = state.setValue(POWERED, true);
                 level.setBlock(pos, blockState, 3);
                 this.updatePowerToConnected(level, pos, blockState, true);
                 level.updateNeighborsAt(pos, this);
@@ -113,7 +_,7 @@
             }
 
             if (!flag && poweredValue) {
-                BlockState blockState = state.setValue(POWERED, Boolean.valueOf(false));
+                BlockState blockState = state.setValue(POWERED, false);
                 level.setBlock(pos, blockState, 3);
                 this.updatePowerToConnected(level, pos, blockState, false);
                 level.updateNeighborsAt(pos, this);
