--- a/net/minecraft/world/level/block/DispenserBlock.java
+++ b/net/minecraft/world/level/block/DispenserBlock.java
@@ -67,7 +_,7 @@
 
     protected DispenserBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(TRIGGERED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(TRIGGERED, false));
     }
 
     @Override
@@ -127,9 +_,9 @@
         boolean triggeredValue = state.getValue(TRIGGERED);
         if (flag && !triggeredValue) {
             level.scheduleTick(pos, this, 4);
-            level.setBlock(pos, state.setValue(TRIGGERED, Boolean.valueOf(true)), 2);
+            level.setBlock(pos, state.setValue(TRIGGERED, true), 2);
         } else if (!flag && triggeredValue) {
-            level.setBlock(pos, state.setValue(TRIGGERED, Boolean.valueOf(false)), 2);
+            level.setBlock(pos, state.setValue(TRIGGERED, false), 2);
         }
     }
 
