--- a/net/minecraft/world/level/block/DoorBlock.java
+++ b/net/minecraft/world/level/block/DoorBlock.java
@@ -66,9 +_,9 @@
             this.stateDefinition
                 .any()
                 .setValue(FACING, Direction.NORTH)
-                .setValue(OPEN, Boolean.valueOf(false))
+                .setValue(OPEN, false)
                 .setValue(HINGE, DoorHingeSide.LEFT)
-                .setValue(POWERED, Boolean.valueOf(false))
+                .setValue(POWERED, false)
                 .setValue(HALF, DoubleBlockHalf.LOWER)
         );
     }
@@ -150,8 +_,8 @@
             return this.defaultBlockState()
                 .setValue(FACING, context.getHorizontalDirection())
                 .setValue(HINGE, this.getHinge(context))
-                .setValue(POWERED, Boolean.valueOf(flag))
-                .setValue(OPEN, Boolean.valueOf(flag))
+                .setValue(POWERED, flag)
+                .setValue(OPEN, flag)
                 .setValue(HALF, DoubleBlockHalf.LOWER);
         } else {
             return null;
@@ -222,7 +_,7 @@
 
     public void setOpen(@Nullable Entity entity, Level level, BlockState state, BlockPos pos, boolean _open) {
         if (state.is(this) && state.getValue(OPEN) != _open) {
-            level.setBlock(pos, state.setValue(OPEN, Boolean.valueOf(_open)), 10);
+            level.setBlock(pos, state.setValue(OPEN, _open), 10);
             this.playSound(entity, level, pos, _open);
             level.gameEvent(entity, _open ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
         }
@@ -251,7 +_,7 @@
                 level.gameEvent(null, flag ? GameEvent.BLOCK_OPEN : GameEvent.BLOCK_CLOSE, pos);
             }
 
-            level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(flag)).setValue(OPEN, Boolean.valueOf(flag)), 2);
+            level.setBlock(pos, state.setValue(POWERED, flag).setValue(OPEN, flag), 2);
         }
     }
 
