--- a/net/minecraft/world/level/block/DoublePlantBlock.java
+++ b/net/minecraft/world/level/block/DoublePlantBlock.java
@@ -94,7 +_,7 @@
 
     public static BlockState copyWaterloggedFrom(LevelReader level, BlockPos pos, BlockState state) {
         return state.hasProperty(BlockStateProperties.WATERLOGGED)
-            ? state.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(level.isWaterAt(pos)))
+            ? state.setValue(BlockStateProperties.WATERLOGGED, level.isWaterAt(pos))
             : state;
     }
 
