--- a/net/minecraft/world/level/block/EnderChestBlock.java
+++ b/net/minecraft/world/level/block/EnderChestBlock.java
@@ -52,7 +_,7 @@
 
     protected EnderChestBlock(BlockBehaviour.Properties properties) {
         super(properties, () -> BlockEntityType.ENDER_CHEST);
-        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, false));
     }
 
     @Override
@@ -70,7 +_,7 @@
         FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
         return this.defaultBlockState()
             .setValue(FACING, context.getHorizontalDirection().getOpposite())
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+            .setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
     }
 
     @Override
