--- a/net/minecraft/world/level/block/FenceBlock.java
+++ b/net/minecraft/world/level/block/FenceBlock.java
@@ -37,11 +_,11 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(NORTH, Boolean.valueOf(false))
-                .setValue(EAST, Boolean.valueOf(false))
-                .setValue(SOUTH, Boolean.valueOf(false))
-                .setValue(WEST, Boolean.valueOf(false))
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(NORTH, false)
+                .setValue(EAST, false)
+                .setValue(SOUTH, false)
+                .setValue(WEST, false)
+                .setValue(WATERLOGGED, false)
         );
         this.occlusionByIndex = this.makeShapes(2.0F, 1.0F, 16.0F, 6.0F, 15.0F);
     }
@@ -91,11 +_,11 @@
         BlockState blockState2 = level.getBlockState(blockPos2);
         BlockState blockState3 = level.getBlockState(blockPos3);
         return super.getStateForPlacement(context)
-            .setValue(NORTH, Boolean.valueOf(this.connectsTo(blockState, blockState.isFaceSturdy(level, blockPos, Direction.SOUTH), Direction.SOUTH)))
-            .setValue(EAST, Boolean.valueOf(this.connectsTo(blockState1, blockState1.isFaceSturdy(level, blockPos1, Direction.WEST), Direction.WEST)))
-            .setValue(SOUTH, Boolean.valueOf(this.connectsTo(blockState2, blockState2.isFaceSturdy(level, blockPos2, Direction.NORTH), Direction.NORTH)))
-            .setValue(WEST, Boolean.valueOf(this.connectsTo(blockState3, blockState3.isFaceSturdy(level, blockPos3, Direction.EAST), Direction.EAST)))
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+            .setValue(NORTH, this.connectsTo(blockState, blockState.isFaceSturdy(level, blockPos, Direction.SOUTH), Direction.SOUTH))
+            .setValue(EAST, this.connectsTo(blockState1, blockState1.isFaceSturdy(level, blockPos1, Direction.WEST), Direction.WEST))
+            .setValue(SOUTH, this.connectsTo(blockState2, blockState2.isFaceSturdy(level, blockPos2, Direction.NORTH), Direction.NORTH))
+            .setValue(WEST, this.connectsTo(blockState3, blockState3.isFaceSturdy(level, blockPos3, Direction.EAST), Direction.EAST))
+            .setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
     }
 
     @Override
@@ -116,9 +_,7 @@
         return direction.getAxis().isHorizontal()
             ? state.setValue(
                 PROPERTY_BY_DIRECTION.get(direction),
-                Boolean.valueOf(
-                    this.connectsTo(neighborState, neighborState.isFaceSturdy(level, neighborPos, direction.getOpposite()), direction.getOpposite())
-                )
+                this.connectsTo(neighborState, neighborState.isFaceSturdy(level, neighborPos, direction.getOpposite()), direction.getOpposite())
             )
             : super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
     }
