--- a/net/minecraft/world/level/block/FrostedIceBlock.java
+++ b/net/minecraft/world/level/block/FrostedIceBlock.java
@@ -32,7 +_,7 @@
 
     public FrostedIceBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, 0));
     }
 
     @Override
@@ -63,7 +_,7 @@
     private boolean slightlyMelt(BlockState state, Level level, BlockPos pos) {
         int ageValue = state.getValue(AGE);
         if (ageValue < 3) {
-            level.setBlock(pos, state.setValue(AGE, Integer.valueOf(ageValue + 1)), 2);
+            level.setBlock(pos, state.setValue(AGE, ageValue + 1), 2);
             return false;
         } else {
             this.melt(state, level, pos);
