--- a/net/minecraft/world/level/block/HangingRootsBlock.java
+++ b/net/minecraft/world/level/block/HangingRootsBlock.java
@@ -31,7 +_,7 @@
 
     protected HangingRootsBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, false));
     }
 
     @Override
@@ -50,7 +_,7 @@
         BlockState blockState = super.getStateForPlacement(context);
         if (blockState != null) {
             FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
-            return blockState.setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+            return blockState.setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
         } else {
             return null;
         }
