--- a/net/minecraft/world/level/block/HangingMossBlock.java
+++ b/net/minecraft/world/level/block/HangingMossBlock.java
@@ -34,7 +_,7 @@
 
     public HangingMossBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(TIP, Boolean.valueOf(true)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(TIP, true));
     }
 
     @Override
@@ -83,7 +_,7 @@
             scheduledTickAccess.scheduleTick(pos, this, 1);
         }
 
-        return state.setValue(TIP, Boolean.valueOf(!level.getBlockState(pos.below()).is(this)));
+        return state.setValue(TIP, !level.getBlockState(pos.below()).is(this));
     }
 
     @Override
@@ -128,7 +_,7 @@
     public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
         BlockPos blockPos = this.getTip(level, pos).below();
         if (this.canGrowInto(level.getBlockState(blockPos))) {
-            level.setBlockAndUpdate(blockPos, state.setValue(TIP, Boolean.valueOf(true)));
+            level.setBlockAndUpdate(blockPos, state.setValue(TIP, true));
         }
     }
 }
