--- a/net/minecraft/world/level/block/HeavyCoreBlock.java
+++ b/net/minecraft/world/level/block/HeavyCoreBlock.java
@@ -26,7 +_,7 @@
 
     public HeavyCoreBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(WATERLOGGED, false));
     }
 
     @Override
@@ -65,7 +_,7 @@
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
-        return this.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(fluidState.is(Fluids.WATER)));
+        return this.defaultBlockState().setValue(WATERLOGGED, fluidState.is(Fluids.WATER));
     }
 
     @Override
