--- a/net/minecraft/world/level/block/HopperBlock.java
+++ b/net/minecraft/world/level/block/HopperBlock.java
@@ -58,7 +_,7 @@
 
     public HopperBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.DOWN).setValue(ENABLED, Boolean.valueOf(true)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.DOWN).setValue(ENABLED, true));
     }
 
     @Override
@@ -102,7 +_,7 @@
         Direction opposite = context.getClickedFace().getOpposite();
         return this.defaultBlockState()
             .setValue(FACING, opposite.getAxis() == Direction.Axis.Y ? Direction.DOWN : opposite)
-            .setValue(ENABLED, Boolean.valueOf(true));
+            .setValue(ENABLED, true);
     }
 
     @Override
@@ -140,7 +_,7 @@
     private void checkPoweredState(Level level, BlockPos pos, BlockState state) {
         boolean flag = !level.hasNeighborSignal(pos);
         if (flag != state.getValue(ENABLED)) {
-            level.setBlock(pos, state.setValue(ENABLED, Boolean.valueOf(flag)), 2);
+            level.setBlock(pos, state.setValue(ENABLED, flag), 2);
         }
     }
 
