--- a/net/minecraft/world/level/block/JukeboxBlock.java
+++ b/net/minecraft/world/level/block/JukeboxBlock.java
@@ -36,7 +_,7 @@
 
     protected JukeboxBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(HAS_RECORD, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(HAS_RECORD, false));
     }
 
     @Override
@@ -44,7 +_,7 @@
         super.setPlacedBy(level, pos, state, placer, stack);
         CustomData customData = stack.getOrDefault(DataComponents.BLOCK_ENTITY_DATA, CustomData.EMPTY);
         if (customData.contains("RecordItem")) {
-            level.setBlock(pos, state.setValue(HAS_RECORD, Boolean.valueOf(true)), 2);
+            level.setBlock(pos, state.setValue(HAS_RECORD, true), 2);
         }
     }
 
