--- a/net/minecraft/world/level/block/LanternBlock.java
+++ b/net/minecraft/world/level/block/LanternBlock.java
@@ -35,7 +_,7 @@
 
     public LanternBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(HANGING, Boolean.valueOf(false)).setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(HANGING, false).setValue(WATERLOGGED, false));
     }
 
     @Nullable
@@ -45,9 +_,9 @@
 
         for (Direction direction : context.getNearestLookingDirections()) {
             if (direction.getAxis() == Direction.Axis.Y) {
-                BlockState blockState = this.defaultBlockState().setValue(HANGING, Boolean.valueOf(direction == Direction.UP));
+                BlockState blockState = this.defaultBlockState().setValue(HANGING, direction == Direction.UP);
                 if (blockState.canSurvive(context.getLevel(), context.getClickedPos())) {
-                    return blockState.setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+                    return blockState.setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
                 }
             }
         }
