--- a/net/minecraft/world/level/block/LayeredCauldronBlock.java
+++ b/net/minecraft/world/level/block/LayeredCauldronBlock.java
@@ -41,7 +_,7 @@
     public LayeredCauldronBlock(Biome.Precipitation precipitationType, CauldronInteraction.InteractionMap interactions, BlockBehaviour.Properties properties) {
         super(properties, interactions);
         this.precipitationType = precipitationType;
-        this.registerDefaultState(this.stateDefinition.any().setValue(LEVEL, Integer.valueOf(1)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(LEVEL, 1));
     }
 
     @Override
@@ -56,7 +_,7 @@
 
     @Override
     protected double getContentHeight(BlockState state) {
-        return (6.0 + state.getValue(LEVEL).intValue() * 3.0) / 16.0;
+        return (6.0 + state.getValue(LEVEL) * 3.0) / 16.0;
     }
 
     @Override
@@ -139,7 +_,7 @@
     @Override
     protected void receiveStalactiteDrip(BlockState state, Level level, BlockPos pos, Fluid fluid) {
         if (!this.isFull(state)) {
-            BlockState blockState = state.setValue(LEVEL, Integer.valueOf(state.getValue(LEVEL) + 1));
+            BlockState blockState = state.setValue(LEVEL, state.getValue(LEVEL) + 1);
             // CraftBukkit start
             if (!LayeredCauldronBlock.changeLevel(level, pos, blockState, null, org.bukkit.event.block.CauldronLevelChangeEvent.ChangeReason.NATURAL_FILL)) {
                 return;
