--- a/net/minecraft/world/level/block/LecternBlock.java
+++ b/net/minecraft/world/level/block/LecternBlock.java
@@ -84,7 +_,7 @@
     protected LecternBlock(BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(POWERED, Boolean.valueOf(false)).setValue(HAS_BOOK, Boolean.valueOf(false))
+            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(POWERED, false).setValue(HAS_BOOK, false)
         );
     }
 
@@ -111,7 +_,7 @@
             }
         }
 
-        return this.defaultBlockState().setValue(FACING, context.getHorizontalDirection().getOpposite()).setValue(HAS_BOOK, Boolean.valueOf(flag));
+        return this.defaultBlockState().setValue(FACING, context.getHorizontalDirection().getOpposite()).setValue(HAS_BOOK, flag); // Plazma - Remove unnecessary type casting
     }
 
     @Override
@@ -193,7 +_,7 @@
     }
 
     public static void resetBookState(@Nullable Entity entity, Level level, BlockPos pos, BlockState state, boolean hasBook) {
-        BlockState blockState = state.setValue(POWERED, Boolean.valueOf(false)).setValue(HAS_BOOK, Boolean.valueOf(hasBook));
+        BlockState blockState = state.setValue(POWERED, false).setValue(HAS_BOOK, hasBook);
         level.setBlock(pos, blockState, 3);
         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(entity, blockState));
         updateBelow(level, pos, state);
@@ -216,7 +_,7 @@
             }
         }
         // Paper end - Call BlockRedstoneEvent properly
-        level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(powered)), 3);
+        level.setBlock(pos, state.setValue(POWERED, powered), 3);
         updateBelow(level, pos, state);
     }
 
