--- a/net/minecraft/world/level/block/MangrovePropaguleBlock.java
+++ b/net/minecraft/world/level/block/MangrovePropaguleBlock.java
@@ -52,10 +_,10 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(STAGE, Integer.valueOf(0))
-                .setValue(AGE, Integer.valueOf(0))
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
-                .setValue(HANGING, Boolean.valueOf(false))
+                .setValue(STAGE, 0)
+                .setValue(AGE, 0)
+                .setValue(WATERLOGGED, false)
+                .setValue(HANGING, false)
         );
     }
 
@@ -74,7 +_,7 @@
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
         boolean flag = fluidState.getType() == Fluids.WATER;
-        return super.getStateForPlacement(context).setValue(WATERLOGGED, Boolean.valueOf(flag)).setValue(AGE, Integer.valueOf(4));
+        return super.getStateForPlacement(context).setValue(WATERLOGGED, flag).setValue(AGE, 4);
     }
 
     @Override
@@ -165,6 +_,6 @@
     }
 
     public static BlockState createNewHangingPropagule(int age) {
-        return Blocks.MANGROVE_PROPAGULE.defaultBlockState().setValue(HANGING, Boolean.valueOf(true)).setValue(AGE, Integer.valueOf(age));
+        return Blocks.MANGROVE_PROPAGULE.defaultBlockState().setValue(HANGING, true).setValue(AGE, age);
     }
 }
