--- a/net/minecraft/world/level/block/MossyCarpetBlock.java
+++ b/net/minecraft/world/level/block/MossyCarpetBlock.java
@@ -70,7 +_,7 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(BASE, Boolean.valueOf(true))
+                .setValue(BASE, true)
                 .setValue(NORTH, WallSide.NONE)
                 .setValue(EAST, WallSide.NONE)
                 .setValue(SOUTH, WallSide.NONE)
@@ -222,7 +_,7 @@
         BlockState blockState = level.getBlockState(blockPos);
         boolean isPaleMossCarpet = blockState.is(Blocks.PALE_MOSS_CARPET);
         if ((!isPaleMossCarpet || !blockState.getValue(BASE)) && (isPaleMossCarpet || blockState.canBeReplaced())) {
-            BlockState blockState1 = Blocks.PALE_MOSS_CARPET.defaultBlockState().setValue(BASE, Boolean.valueOf(false));
+            BlockState blockState1 = Blocks.PALE_MOSS_CARPET.defaultBlockState().setValue(BASE, false);
             BlockState updatedState = getUpdatedState(blockState1, level, pos.above(), true);
 
             for (Direction direction : Direction.Plane.HORIZONTAL) {
