--- a/net/minecraft/world/level/block/MultifaceBlock.java
+++ b/net/minecraft/world/level/block/MultifaceBlock.java
@@ -210,12 +_,12 @@
             if (currentState.is(this)) {
                 blockState = currentState;
             } else if (currentState.getFluidState().isSourceOfType(Fluids.WATER)) {
-                blockState = this.defaultBlockState().setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true));
+                blockState = this.defaultBlockState().setValue(BlockStateProperties.WATERLOGGED, true);
             } else {
                 blockState = this.defaultBlockState();
             }
 
-            return blockState.setValue(getFaceProperty(lookingDirection), Boolean.valueOf(true));
+            return blockState.setValue(getFaceProperty(lookingDirection), true);
         }
     }
 
@@ -247,7 +_,7 @@
 
     public static boolean hasFace(BlockState state, Direction direction) {
         BooleanProperty faceProperty = getFaceProperty(direction);
-        return state.getValueOrElse(faceProperty, Boolean.valueOf(false));
+        return state.getValueOrElse(faceProperty, false);
     }
 
     public static boolean canAttachTo(BlockGetter level, BlockPos pos, Direction direction) {
@@ -262,7 +_,7 @@
     }
 
     private static BlockState removeFace(BlockState state, BooleanProperty faceProp) {
-        BlockState blockState = state.setValue(faceProp, Boolean.valueOf(false));
+        BlockState blockState = state.setValue(faceProp, false);
         return hasAnyFace(blockState) ? blockState : Blocks.AIR.defaultBlockState();
     }
 
@@ -271,10 +_,10 @@
     }
 
     private static BlockState getDefaultMultifaceState(StateDefinition<Block, BlockState> stateDefinition) {
-        BlockState blockState = stateDefinition.any().setValue(WATERLOGGED, Boolean.valueOf(false));
+        BlockState blockState = stateDefinition.any().setValue(WATERLOGGED, false);
 
         for (BooleanProperty booleanProperty : PROPERTY_BY_DIRECTION.values()) {
-            blockState = blockState.trySetValue(booleanProperty, Boolean.valueOf(false));
+            blockState = blockState.trySetValue(booleanProperty, false);
         }
 
         return blockState;
