--- a/net/minecraft/world/level/block/NetherWartBlock.java
+++ b/net/minecraft/world/level/block/NetherWartBlock.java
@@ -34,7 +_,7 @@
 
     protected NetherWartBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, 0));
     }
 
     @Override
@@ -56,7 +_,7 @@
     protected void randomTick(BlockState state, ServerLevel level, BlockPos pos, RandomSource random) {
         int ageValue = state.getValue(AGE);
         if (ageValue < 3 && random.nextFloat() < (level.spigotConfig.wartModifier / (100.0f * 10))) { // Spigot - SPIGOT-7159: Better modifier resolution
-            state = state.setValue(AGE, Integer.valueOf(ageValue + 1));
+            state = state.setValue(AGE, ageValue + 1);
             org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state, 2); // CraftBukkit
         }
     }
