--- a/net/minecraft/world/level/block/NoteBlock.java
+++ b/net/minecraft/world/level/block/NoteBlock.java
@@ -52,8 +_,8 @@
             this.stateDefinition
                 .any()
                 .setValue(INSTRUMENT, NoteBlockInstrument.HARP)
-                .setValue(NOTE, Integer.valueOf(0))
-                .setValue(POWERED, Boolean.valueOf(false))
+                .setValue(NOTE, 0)
+                .setValue(POWERED, false)
         );
     }
 
@@ -102,7 +_,7 @@
                 state = level.getBlockState(pos); // CraftBukkit - SPIGOT-5617: update in case changed in event
             }
 
-            level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(hasNeighborSignal)), 3);
+            level.setBlock(pos, state.setValue(POWERED, hasNeighborSignal), 3);
         }
     }
 
