--- a/net/minecraft/world/level/block/PinkPetalsBlock.java
+++ b/net/minecraft/world/level/block/PinkPetalsBlock.java
@@ -54,7 +_,7 @@
 
     protected PinkPetalsBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(AMOUNT, Integer.valueOf(1)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(AMOUNT, 1));
     }
 
     @Override
@@ -82,7 +_,7 @@
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         BlockState blockState = context.getLevel().getBlockState(context.getClickedPos());
         return blockState.is(this)
-            ? blockState.setValue(AMOUNT, Integer.valueOf(Math.min(4, blockState.getValue(AMOUNT) + 1)))
+            ? blockState.setValue(AMOUNT, Math.min(4, blockState.getValue(AMOUNT) + 1))
             : this.defaultBlockState().setValue(FACING, context.getHorizontalDirection().getOpposite());
     }
 
@@ -105,7 +_,7 @@
     public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
         int amountValue = state.getValue(AMOUNT);
         if (amountValue < 4) {
-            level.setBlock(pos, state.setValue(AMOUNT, Integer.valueOf(amountValue + 1)), 2);
+            level.setBlock(pos, state.setValue(AMOUNT, amountValue + 1), 2);
         } else {
             popResource(level, pos, new ItemStack(this));
         }
