--- a/net/minecraft/world/level/block/PitcherCropBlock.java
+++ b/net/minecraft/world/level/block/PitcherCropBlock.java
@@ -141,7 +_,7 @@
     private void grow(ServerLevel level, BlockState state, BlockPos pos, int ageIncrement) {
         int min = Math.min(state.getValue(AGE) + ageIncrement, 4);
         if (this.canGrow(level, pos, state, min)) {
-            BlockState blockState = state.setValue(AGE, Integer.valueOf(min));
+            BlockState blockState = state.setValue(AGE, min);
             if (!org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, blockState, 2)) return; // Paper
             if (isDouble(min)) {
                 level.setBlock(pos.above(), blockState.setValue(HALF, DoubleBlockHalf.UPPER), 3);
