--- a/net/minecraft/world/level/block/PointedDripstoneBlock.java
+++ b/net/minecraft/world/level/block/PointedDripstoneBlock.java
@@ -87,7 +_,7 @@
                 .any()
                 .setValue(TIP_DIRECTION, Direction.UP)
                 .setValue(THICKNESS, DripstoneThickness.TIP)
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(WATERLOGGED, false)
         );
     }
 
@@ -257,7 +_,7 @@
                 : this.defaultBlockState()
                     .setValue(TIP_DIRECTION, direction)
                     .setValue(THICKNESS, dripstoneThickness)
-                    .setValue(WATERLOGGED, Boolean.valueOf(level.getFluidState(clickedPos).getType() == Fluids.WATER));
+                    .setValue(WATERLOGGED, level.getFluidState(clickedPos).getType() == Fluids.WATER);
         }
     }
 
@@ -395,7 +_,7 @@
             .defaultBlockState()
             .setValue(TIP_DIRECTION, direction)
             .setValue(THICKNESS, thickness)
-            .setValue(WATERLOGGED, Boolean.valueOf(level.getFluidState(pos).getType() == Fluids.WATER));
+            .setValue(WATERLOGGED, level.getFluidState(pos).getType() == Fluids.WATER);
         org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockSpreadEvent(level, source, pos, blockState, 3); // CraftBukkit
     }
 
