--- a/net/minecraft/world/level/block/PoweredRailBlock.java
+++ b/net/minecraft/world/level/block/PoweredRailBlock.java
@@ -28,8 +_,8 @@
             this.stateDefinition
                 .any()
                 .setValue(SHAPE, RailShape.NORTH_SOUTH)
-                .setValue(POWERED, Boolean.valueOf(false))
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(POWERED, false)
+                .setValue(WATERLOGGED, false)
         );
     }
 
@@ -140,7 +_,7 @@
                 return;
             }
             // CraftBukkit end
-            level.setBlock(pos, state.setValue(POWERED, Boolean.valueOf(flag)), 3);
+            level.setBlock(pos, state.setValue(POWERED, flag), 3);
             level.updateNeighborsAt(pos.below(), this);
             if (state.getValue(SHAPE).isSlope()) {
                 level.updateNeighborsAt(pos.above(), this);
