--- a/net/minecraft/world/level/block/PressurePlateBlock.java
+++ b/net/minecraft/world/level/block/PressurePlateBlock.java
@@ -28,7 +_,7 @@
 
     protected PressurePlateBlock(BlockSetType type, BlockBehaviour.Properties properties) {
         super(properties, type);
-        this.registerDefaultState(this.stateDefinition.any().setValue(POWERED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(POWERED, false));
     }
 
     @Override
@@ -38,7 +_,7 @@
 
     @Override
     protected BlockState setSignalForState(BlockState state, int strength) {
-        return state.setValue(POWERED, Boolean.valueOf(strength > 0));
+        return state.setValue(POWERED, strength > 0);
     }
 
     @Override
