--- a/net/minecraft/world/level/block/RailBlock.java
+++ b/net/minecraft/world/level/block/RailBlock.java
@@ -22,7 +_,7 @@
 
     protected RailBlock(BlockBehaviour.Properties properties) {
         super(false, properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(SHAPE, RailShape.NORTH_SOUTH).setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(SHAPE, RailShape.NORTH_SOUTH).setValue(WATERLOGGED, false));
     }
 
     @Override
