--- a/net/minecraft/world/level/block/RedStoneOreBlock.java
+++ b/net/minecraft/world/level/block/RedStoneOreBlock.java
@@ -32,7 +_,7 @@
 
     public RedStoneOreBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.defaultBlockState().setValue(LIT, Boolean.valueOf(false)));
+        this.registerDefaultState(this.defaultBlockState().setValue(LIT, false));
     }
 
     @Override
@@ -86,7 +_,7 @@
                 return;
             }
             // CraftBukkit end
-            level.setBlock(pos, state.setValue(LIT, Boolean.valueOf(true)), 3);
+            level.setBlock(pos, state.setValue(LIT, true), 3);
         }
     }
 
@@ -103,7 +_,7 @@
                 return;
             }
             // CraftBukkit end
-            level.setBlock(pos, state.setValue(LIT, Boolean.valueOf(false)), 3);
+            level.setBlock(pos, state.setValue(LIT, false), 3);
         }
     }
 
@@ -131,8 +_,8 @@
     }
 
     private static void spawnParticles(Level level, BlockPos pos) {
-        double d = 0.5625;
-        RandomSource randomSource = level.random;
+        //double d = 0.5625; // Plazma - Remove unused variables
+        final RandomSource randomSource = level.random; // Plazma
 
         for (Direction direction : Direction.values()) {
             BlockPos blockPos = pos.relative(direction);
