--- a/net/minecraft/world/level/block/RedStoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedStoneWireBlock.java
@@ -107,7 +_,7 @@
                 .setValue(EAST, RedstoneSide.NONE)
                 .setValue(SOUTH, RedstoneSide.NONE)
                 .setValue(WEST, RedstoneSide.NONE)
-                .setValue(POWER, Integer.valueOf(0))
+                .setValue(POWER, 0)
         );
         this.crossState = this.defaultBlockState()
             .setValue(NORTH, RedstoneSide.SIDE)
@@ -139,7 +_,7 @@
 
     @Override
     protected VoxelShape getShape(BlockState state, BlockGetter level, BlockPos pos, CollisionContext context) {
-        return SHAPES_CACHE.get(state.setValue(POWER, Integer.valueOf(0)));
+        return SHAPES_CACHE.get(state.setValue(POWER, 0));
     }
 
     @Override
