--- a/net/minecraft/world/level/block/RedstoneTorchBlock.java
+++ b/net/minecraft/world/level/block/RedstoneTorchBlock.java
@@ -37,7 +_,7 @@
 
     protected RedstoneTorchBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(LIT, Boolean.valueOf(true)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(LIT, true));
     }
 
     @Override
@@ -85,7 +_,7 @@
         // CraftBukkit start
         org.bukkit.plugin.PluginManager manager = level.getCraftServer().getPluginManager();
         org.bukkit.block.Block block = level.getWorld().getBlockAt(pos.getX(), pos.getY(), pos.getZ());
-        int oldCurrent = ((Boolean) state.getValue(RedstoneTorchBlock.LIT)).booleanValue() ? 15 : 0;
+        int oldCurrent = (Boolean) state.getValue(RedstoneTorchBlock.LIT) ? 15 : 0;
 
         org.bukkit.event.block.BlockRedstoneEvent event = new org.bukkit.event.block.BlockRedstoneEvent(block, oldCurrent, oldCurrent);
         // CraftBukkit end
@@ -100,7 +_,7 @@
                     }
                 }
                 // CraftBukkit end
-                level.setBlock(pos, state.setValue(LIT, Boolean.valueOf(false)), 3);
+                level.setBlock(pos, state.setValue(LIT, false), 3);
                 if (isToggledTooFrequently(level, pos, true)) {
                     level.levelEvent(1502, pos, 0);
                     level.scheduleTick(pos, level.getBlockState(pos).getBlock(), 160);
@@ -116,7 +_,7 @@
                 }
             }
             // CraftBukkit end
-            level.setBlock(pos, state.setValue(LIT, Boolean.valueOf(true)), 3);
+            level.setBlock(pos, state.setValue(LIT, true), 3);
         }
     }
 
