--- a/net/minecraft/world/level/block/RepeaterBlock.java
+++ b/net/minecraft/world/level/block/RepeaterBlock.java
@@ -35,9 +_,9 @@
             this.stateDefinition
                 .any()
                 .setValue(FACING, Direction.NORTH)
-                .setValue(DELAY, Integer.valueOf(1))
-                .setValue(LOCKED, Boolean.valueOf(false))
-                .setValue(POWERED, Boolean.valueOf(false))
+                .setValue(DELAY, 1)
+                .setValue(LOCKED, false)
+                .setValue(POWERED, false)
         );
     }
 
@@ -59,7 +_,7 @@
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         BlockState blockState = super.getStateForPlacement(context);
-        return blockState.setValue(LOCKED, Boolean.valueOf(this.isLocked(context.getLevel(), context.getClickedPos(), blockState)));
+        return blockState.setValue(LOCKED, this.isLocked(context.getLevel(), context.getClickedPos(), blockState));
     }
 
     @Override
@@ -77,7 +_,7 @@
             return Blocks.AIR.defaultBlockState();
         } else {
             return !level.isClientSide() && direction.getAxis() != state.getValue(FACING).getAxis()
-                ? state.setValue(LOCKED, Boolean.valueOf(this.isLocked(level, pos, state)))
+                ? state.setValue(LOCKED, this.isLocked(level, pos, state))
                 : super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
         }
     }
