--- a/net/minecraft/world/level/block/RespawnAnchorBlock.java
+++ b/net/minecraft/world/level/block/RespawnAnchorBlock.java
@@ -68,7 +_,7 @@
 
     public RespawnAnchorBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(CHARGE, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(CHARGE, 0));
     }
 
     @Override
@@ -168,7 +_,7 @@
     }
 
     public static void charge(@Nullable Entity entity, Level level, BlockPos pos, BlockState state) {
-        BlockState blockState = state.setValue(CHARGE, Integer.valueOf(state.getValue(CHARGE) + 1));
+        BlockState blockState = state.setValue(CHARGE, state.getValue(CHARGE) + 1);
         level.setBlock(pos, blockState, 3);
         level.gameEvent(GameEvent.BLOCK_CHANGE, pos, GameEvent.Context.of(entity, blockState));
         level.playSound(null, pos.getX() + 0.5, pos.getY() + 0.5, pos.getZ() + 0.5, SoundEvents.RESPAWN_ANCHOR_CHARGE, SoundSource.BLOCKS, 1.0F, 1.0F);
