--- a/net/minecraft/world/level/block/SaplingBlock.java
+++ b/net/minecraft/world/level/block/SaplingBlock.java
@@ -36,7 +_,7 @@
     protected SaplingBlock(TreeGrower treeGrower, BlockBehaviour.Properties properties) {
         super(properties);
         this.treeGrower = treeGrower;
-        this.registerDefaultState(this.stateDefinition.any().setValue(STAGE, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(STAGE, 0));
     }
 
     @Override
