--- a/net/minecraft/world/level/block/SculkVeinBlock.java
+++ b/net/minecraft/world/level/block/SculkVeinBlock.java
@@ -49,7 +_,7 @@
 
         for (Direction direction : directions) {
             if (canAttachTo(level, pos, direction)) {
-                blockState = blockState.setValue(getFaceProperty(direction), Boolean.valueOf(true));
+                blockState = blockState.setValue(getFaceProperty(direction), true);
                 flag = true;
             }
         }
@@ -58,7 +_,7 @@
             return false;
         } else {
             if (!state.getFluidState().isEmpty()) {
-                blockState = blockState.setValue(MultifaceBlock.WATERLOGGED, Boolean.valueOf(true));
+                blockState = blockState.setValue(MultifaceBlock.WATERLOGGED, true);
             }
 
             level.setBlock(pos, blockState, 3);
@@ -72,7 +_,7 @@
             for (Direction direction : DIRECTIONS) {
                 BooleanProperty faceProperty = getFaceProperty(direction);
                 if (state.getValue(faceProperty) && level.getBlockState(pos.relative(direction)).is(Blocks.SCULK)) {
-                    state = state.setValue(faceProperty, Boolean.valueOf(false));
+                    state = state.setValue(faceProperty, false);
                 }
             }
 
