--- a/net/minecraft/world/level/block/SculkBlock.java
+++ b/net/minecraft/world/level/block/SculkBlock.java
@@ -62,13 +_,13 @@
     private BlockState getRandomGrowthState(LevelAccessor level, BlockPos pos, RandomSource random, boolean isWorldGeneration) {
         BlockState blockState;
         if (random.nextInt(11) == 0) {
-            blockState = Blocks.SCULK_SHRIEKER.defaultBlockState().setValue(SculkShriekerBlock.CAN_SUMMON, Boolean.valueOf(isWorldGeneration));
+            blockState = Blocks.SCULK_SHRIEKER.defaultBlockState().setValue(SculkShriekerBlock.CAN_SUMMON, isWorldGeneration);
         } else {
             blockState = Blocks.SCULK_SENSOR.defaultBlockState();
         }
 
         return blockState.hasProperty(BlockStateProperties.WATERLOGGED) && !level.getFluidState(pos).isEmpty()
-            ? blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true))
+            ? blockState.setValue(BlockStateProperties.WATERLOGGED, true)
             : blockState;
     }
 
