--- a/net/minecraft/world/level/block/SeaPickleBlock.java
+++ b/net/minecraft/world/level/block/SeaPickleBlock.java
@@ -41,7 +_,7 @@
 
     protected SeaPickleBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(PICKLES, Integer.valueOf(1)).setValue(WATERLOGGED, Boolean.valueOf(true)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(PICKLES, 1).setValue(WATERLOGGED, true));
     }
 
     @Nullable
@@ -49,11 +_,11 @@
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         BlockState blockState = context.getLevel().getBlockState(context.getClickedPos());
         if (blockState.is(this)) {
-            return blockState.setValue(PICKLES, Integer.valueOf(Math.min(4, blockState.getValue(PICKLES) + 1)));
+            return blockState.setValue(PICKLES, Math.min(4, blockState.getValue(PICKLES) + 1));
         } else {
             FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
             boolean flag = fluidState.getType() == Fluids.WATER;
-            return super.getStateForPlacement(context).setValue(WATERLOGGED, Boolean.valueOf(flag));
+            return super.getStateForPlacement(context).setValue(WATERLOGGED, flag);
         }
     }
 
@@ -153,7 +_,7 @@
                     if (blockPos != pos && random.nextInt(6) == 0 && level.getBlockState(blockPos).is(Blocks.WATER)) {
                         BlockState blockState = level.getBlockState(blockPos.below());
                         if (blockState.is(BlockTags.CORAL_BLOCKS)) {
-                            level.setBlock(blockPos, Blocks.SEA_PICKLE.defaultBlockState().setValue(PICKLES, Integer.valueOf(random.nextInt(4) + 1)), 3);
+                            level.setBlock(blockPos, Blocks.SEA_PICKLE.defaultBlockState().setValue(PICKLES, random.nextInt(4) + 1), 3);
                         }
                     }
                 }
@@ -170,7 +_,7 @@
             i3++;
         }
 
-        level.setBlock(pos, state.setValue(PICKLES, Integer.valueOf(4)), 2);
+        level.setBlock(pos, state.setValue(PICKLES, 4), 2);
     }
 
     @Override
