--- a/net/minecraft/world/level/block/SimpleWaterloggedBlock.java
+++ b/net/minecraft/world/level/block/SimpleWaterloggedBlock.java
@@ -25,7 +_,7 @@
     default boolean placeLiquid(LevelAccessor level, BlockPos pos, BlockState state, FluidState fluidState) {
         if (!state.getValue(BlockStateProperties.WATERLOGGED) && fluidState.getType() == Fluids.WATER) {
             if (!level.isClientSide()) {
-                level.setBlock(pos, state.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true)), 3);
+                level.setBlock(pos, state.setValue(BlockStateProperties.WATERLOGGED, true), 3);
                 level.scheduleTick(pos, fluidState.getType(), fluidState.getType().getTickDelay(level));
             }
 
@@ -38,7 +_,7 @@
     @Override
     default ItemStack pickupBlock(@Nullable Player player, LevelAccessor level, BlockPos pos, BlockState state) {
         if (state.getValue(BlockStateProperties.WATERLOGGED)) {
-            level.setBlock(pos, state.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(false)), 3);
+            level.setBlock(pos, state.setValue(BlockStateProperties.WATERLOGGED, false), 3);
             if (!state.canSurvive(level, pos)) {
                 level.destroyBlock(pos, true);
             }
