--- a/net/minecraft/world/level/block/SkullBlock.java
+++ b/net/minecraft/world/level/block/SkullBlock.java
@@ -37,7 +_,7 @@
 
     protected SkullBlock(SkullBlock.Type type, BlockBehaviour.Properties properties) {
         super(type, properties);
-        this.registerDefaultState(this.defaultBlockState().setValue(ROTATION, Integer.valueOf(0)));
+        this.registerDefaultState(this.defaultBlockState().setValue(ROTATION, 0));
     }
 
     @Override
@@ -52,17 +_,17 @@
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
-        return super.getStateForPlacement(context).setValue(ROTATION, Integer.valueOf(RotationSegment.convertToSegment(context.getRotation())));
+        return super.getStateForPlacement(context).setValue(ROTATION, RotationSegment.convertToSegment(context.getRotation()));
     }
 
     @Override
     protected BlockState rotate(BlockState state, Rotation rotation) {
-        return state.setValue(ROTATION, Integer.valueOf(rotation.rotate(state.getValue(ROTATION), ROTATIONS)));
+        return state.setValue(ROTATION, rotation.rotate(state.getValue(ROTATION), ROTATIONS));
     }
 
     @Override
     protected BlockState mirror(BlockState state, Mirror mirror) {
-        return state.setValue(ROTATION, Integer.valueOf(mirror.mirror(state.getValue(ROTATION), ROTATIONS)));
+        return state.setValue(ROTATION, mirror.mirror(state.getValue(ROTATION), ROTATIONS));
     }
 
     @Override
