--- a/net/minecraft/world/level/block/SlabBlock.java
+++ b/net/minecraft/world/level/block/SlabBlock.java
@@ -42,7 +_,7 @@
 
     public SlabBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.defaultBlockState().setValue(TYPE, SlabType.BOTTOM).setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.defaultBlockState().setValue(TYPE, SlabType.BOTTOM).setValue(WATERLOGGED, false));
     }
 
     @Override
@@ -74,12 +_,12 @@
         BlockPos clickedPos = context.getClickedPos();
         BlockState blockState = context.getLevel().getBlockState(clickedPos);
         if (blockState.is(this)) {
-            return blockState.setValue(TYPE, SlabType.DOUBLE).setValue(WATERLOGGED, Boolean.valueOf(false));
+            return blockState.setValue(TYPE, SlabType.DOUBLE).setValue(WATERLOGGED, false);
         } else {
             FluidState fluidState = context.getLevel().getFluidState(clickedPos);
             BlockState blockState1 = this.defaultBlockState()
                 .setValue(TYPE, SlabType.BOTTOM)
-                .setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+                .setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
             Direction clickedFace = context.getClickedFace();
             return clickedFace != Direction.DOWN && (clickedFace == Direction.UP || !(context.getClickLocation().y - clickedPos.getY() > 0.5))
                 ? blockState1
