--- a/net/minecraft/world/level/block/SnowLayerBlock.java
+++ b/net/minecraft/world/level/block/SnowLayerBlock.java
@@ -46,7 +_,7 @@
 
     protected SnowLayerBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(LAYERS, Integer.valueOf(1)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(LAYERS, 1));
     }
 
     @Override
@@ -148,7 +_,7 @@
         BlockState blockState = context.getLevel().getBlockState(context.getClickedPos());
         if (blockState.is(this)) {
             int layersValue = blockState.getValue(LAYERS);
-            return blockState.setValue(LAYERS, Integer.valueOf(Math.min(8, layersValue + 1)));
+            return blockState.setValue(LAYERS, Math.min(8, layersValue + 1));
         } else {
             return super.getStateForPlacement(context);
         }
