--- a/net/minecraft/world/level/block/SnowyDirtBlock.java
+++ b/net/minecraft/world/level/block/SnowyDirtBlock.java
@@ -25,7 +_,7 @@
 
     protected SnowyDirtBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(SNOWY, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(SNOWY, false));
     }
 
     @Override
@@ -40,14 +_,14 @@
         RandomSource random
     ) {
         return direction == Direction.UP
-            ? state.setValue(SNOWY, Boolean.valueOf(isSnowySetting(neighborState)))
+            ? state.setValue(SNOWY, isSnowySetting(neighborState))
             : super.updateShape(state, level, scheduledTickAccess, pos, direction, neighborPos, neighborState, random);
     }
 
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         BlockState blockState = context.getLevel().getBlockState(context.getClickedPos().above());
-        return this.defaultBlockState().setValue(SNOWY, Boolean.valueOf(isSnowySetting(blockState)));
+        return this.defaultBlockState().setValue(SNOWY, isSnowySetting(blockState));
     }
 
     protected static boolean isSnowySetting(BlockState state) {
