--- a/net/minecraft/world/level/block/StairBlock.java
+++ b/net/minecraft/world/level/block/StairBlock.java
@@ -90,7 +_,7 @@
                 .setValue(FACING, Direction.NORTH)
                 .setValue(HALF, Half.BOTTOM)
                 .setValue(SHAPE, StairsShape.STRAIGHT)
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(WATERLOGGED, false)
         );
         this.base = baseState.getBlock();
         this.baseState = baseState;
@@ -128,7 +_,7 @@
                     ? Half.BOTTOM
                     : Half.TOP
             )
-            .setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+            .setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
         return blockState.setValue(SHAPE, getStairsShape(blockState, context.getLevel(), clickedPos));
     }
 
