--- a/net/minecraft/world/level/block/StainedGlassPaneBlock.java
+++ b/net/minecraft/world/level/block/StainedGlassPaneBlock.java
@@ -23,11 +_,11 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(NORTH, Boolean.valueOf(false))
-                .setValue(EAST, Boolean.valueOf(false))
-                .setValue(SOUTH, Boolean.valueOf(false))
-                .setValue(WEST, Boolean.valueOf(false))
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(NORTH, false)
+                .setValue(EAST, false)
+                .setValue(SOUTH, false)
+                .setValue(WEST, false)
+                .setValue(WATERLOGGED, false)
         );
     }
 
