--- a/net/minecraft/world/level/block/StemBlock.java
+++ b/net/minecraft/world/level/block/StemBlock.java
@@ -63,7 +_,7 @@
         this.fruit = fruit;
         this.attachedStem = attachedStem;
         this.seed = seed;
-        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(AGE, 0));
     }
 
     @Override
@@ -83,7 +_,7 @@
             if (random.nextFloat() < ((this == Blocks.PUMPKIN_STEM ? level.spigotConfig.pumpkinModifier : level.spigotConfig.melonModifier) / (100.0f * (Math.floor((25.0F / growthSpeed) + 1))))) { // Spigot - SPIGOT-7159: Better modifier resolution
                 int ageValue = state.getValue(AGE);
                 if (ageValue < 7) {
-                    state = state.setValue(AGE, Integer.valueOf(ageValue + 1));
+                    state = state.setValue(AGE, ageValue + 1);
                     org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, state, 2); // CraftBukkit
                 } else {
                     Direction randomDirection = Direction.Plane.HORIZONTAL.getRandomDirection(random);
@@ -125,7 +_,7 @@
     @Override
     public void performBonemeal(ServerLevel level, RandomSource random, BlockPos pos, BlockState state) {
         int min = Math.min(7, state.getValue(AGE) + Mth.nextInt(level.random, 2, 5));
-        BlockState blockState = state.setValue(AGE, Integer.valueOf(min));
+        BlockState blockState = state.setValue(AGE, min);
         org.bukkit.craftbukkit.event.CraftEventFactory.handleBlockGrowEvent(level, pos, blockState, 2); // CraftBukkit
         if (min == 7) {
             blockState.randomTick(level, pos, level.random);
