--- a/net/minecraft/world/level/block/TrapDoorBlock.java
+++ b/net/minecraft/world/level/block/TrapDoorBlock.java
@@ -65,10 +_,10 @@
             this.stateDefinition
                 .any()
                 .setValue(FACING, Direction.NORTH)
-                .setValue(OPEN, Boolean.valueOf(false))
+                .setValue(OPEN, false)
                 .setValue(HALF, Half.BOTTOM)
-                .setValue(POWERED, Boolean.valueOf(false))
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(POWERED, false)
+                .setValue(WATERLOGGED, false)
         );
     }
 
@@ -206,10 +_,10 @@
         }
 
         if (context.getLevel().hasNeighborSignal(context.getClickedPos())) {
-            blockState = blockState.setValue(OPEN, Boolean.valueOf(true)).setValue(POWERED, Boolean.valueOf(true));
+            blockState = blockState.setValue(OPEN, true).setValue(POWERED, true);
         }
 
-        return blockState.setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+        return blockState.setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
     }
 
     @Override
