--- a/net/minecraft/world/level/block/TrialSpawnerBlock.java
+++ b/net/minecraft/world/level/block/TrialSpawnerBlock.java
@@ -35,7 +_,7 @@
 
     public TrialSpawnerBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(STATE, TrialSpawnerState.INACTIVE).setValue(OMINOUS, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(STATE, TrialSpawnerState.INACTIVE).setValue(OMINOUS, false));
     }
 
     @Override
