--- a/net/minecraft/world/level/block/TurtleEggBlock.java
+++ b/net/minecraft/world/level/block/TurtleEggBlock.java
@@ -48,7 +_,7 @@
 
     public TurtleEggBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.stateDefinition.any().setValue(HATCH, Integer.valueOf(0)).setValue(EGGS, Integer.valueOf(1)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(HATCH, 0).setValue(EGGS, 1));
     }
 
     @Override
@@ -97,7 +_,7 @@
         if (eggsValue <= 1) {
             level.destroyBlock(pos, false);
         } else {
-            level.setBlock(pos, state.setValue(EGGS, Integer.valueOf(eggsValue - 1)), 2);
+            level.setBlock(pos, state.setValue(EGGS, eggsValue - 1), 2);
             level.gameEvent(GameEvent.BLOCK_DESTROY, pos, GameEvent.Context.of(state));
             level.levelEvent(2001, pos, Block.getId(state));
         }
@@ -177,7 +_,7 @@
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         BlockState blockState = context.getLevel().getBlockState(context.getClickedPos());
         return blockState.is(this)
-            ? blockState.setValue(EGGS, Integer.valueOf(Math.min(4, blockState.getValue(EGGS) + 1)))
+            ? blockState.setValue(EGGS, Math.min(4, blockState.getValue(EGGS) + 1))
             : super.getStateForPlacement(context);
     }
 
