--- a/net/minecraft/world/level/block/VaultBlock.java
+++ b/net/minecraft/world/level/block/VaultBlock.java
@@ -39,7 +_,7 @@
     public VaultBlock(BlockBehaviour.Properties properties) {
         super(properties);
         this.registerDefaultState(
-            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(STATE, VaultState.INACTIVE).setValue(OMINOUS, Boolean.valueOf(false))
+            this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(STATE, VaultState.INACTIVE).setValue(OMINOUS, false)
         );
     }
 
