--- a/net/minecraft/world/level/block/WallBlock.java
+++ b/net/minecraft/world/level/block/WallBlock.java
@@ -60,12 +_,12 @@
         this.registerDefaultState(
             this.stateDefinition
                 .any()
-                .setValue(UP, Boolean.valueOf(true))
+                .setValue(UP, true)
                 .setValue(NORTH_WALL, WallSide.NONE)
                 .setValue(EAST_WALL, WallSide.NONE)
                 .setValue(SOUTH_WALL, WallSide.NONE)
                 .setValue(WEST_WALL, WallSide.NONE)
-                .setValue(WATERLOGGED, Boolean.valueOf(false))
+                .setValue(WATERLOGGED, false)
         );
         this.shapeByIndex = this.makeShapes(4.0F, 3.0F, 16.0F, 0.0F, 14.0F, 16.0F);
         this.collisionShapeByIndex = this.makeShapes(4.0F, 3.0F, 24.0F, 0.0F, 24.0F, 24.0F);
@@ -115,8 +_,8 @@
                                 .setValue(WEST_WALL, wallSide2)
                                 .setValue(NORTH_WALL, wallSide1)
                                 .setValue(SOUTH_WALL, wallSide3);
-                            builder.put(blockState.setValue(WATERLOGGED, Boolean.valueOf(false)), voxelShape9);
-                            builder.put(blockState.setValue(WATERLOGGED, Boolean.valueOf(true)), voxelShape9);
+                            builder.put(blockState.setValue(WATERLOGGED, false), voxelShape9);
+                            builder.put(blockState.setValue(WATERLOGGED, true), voxelShape9);
                         }
                     }
                 }
@@ -166,7 +_,7 @@
         boolean flag1 = this.connectsTo(blockState1, blockState1.isFaceSturdy(level, blockPos1, Direction.WEST), Direction.WEST);
         boolean flag2 = this.connectsTo(blockState2, blockState2.isFaceSturdy(level, blockPos2, Direction.NORTH), Direction.NORTH);
         boolean flag3 = this.connectsTo(blockState3, blockState3.isFaceSturdy(level, blockPos3, Direction.EAST), Direction.EAST);
-        BlockState blockState5 = this.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+        BlockState blockState5 = this.defaultBlockState().setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
         return this.updateShape(level, blockState5, blockPos4, blockState4, flag, flag1, flag2, flag3);
     }
 
@@ -241,7 +_,7 @@
     ) {
         VoxelShape faceShape = neighbour.getCollisionShape(level, pos).getFaceShape(Direction.DOWN);
         BlockState blockState = this.updateSides(state, northConnection, eastConnection, southConnection, westConnection, faceShape);
-        return blockState.setValue(UP, Boolean.valueOf(this.shouldRaisePost(blockState, neighbour, faceShape)));
+        return blockState.setValue(UP, this.shouldRaisePost(blockState, neighbour, faceShape));
     }
 
     private boolean shouldRaisePost(BlockState state, BlockState neighbour, VoxelShape shape) {
