--- a/net/minecraft/world/level/block/WallSignBlock.java
+++ b/net/minecraft/world/level/block/WallSignBlock.java
@@ -52,7 +_,7 @@
 
     public WallSignBlock(WoodType type, BlockBehaviour.Properties properties) {
         super(type, properties.sound(type.soundType()));
-        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(FACING, Direction.NORTH).setValue(WATERLOGGED, false));
     }
 
     @Override
@@ -79,7 +_,7 @@
                 Direction opposite = direction.getOpposite();
                 blockState = blockState.setValue(FACING, opposite);
                 if (blockState.canSurvive(level, clickedPos)) {
-                    return blockState.setValue(WATERLOGGED, Boolean.valueOf(fluidState.getType() == Fluids.WATER));
+                    return blockState.setValue(WATERLOGGED, fluidState.getType() == Fluids.WATER);
                 }
             }
         }
