--- a/net/minecraft/world/level/block/WaterloggedTransparentBlock.java
+++ b/net/minecraft/world/level/block/WaterloggedTransparentBlock.java
@@ -27,14 +_,14 @@
 
     protected WaterloggedTransparentBlock(BlockBehaviour.Properties properties) {
         super(properties);
-        this.registerDefaultState(this.defaultBlockState().setValue(WATERLOGGED, Boolean.valueOf(false)));
+        this.registerDefaultState(this.defaultBlockState().setValue(WATERLOGGED, false));
     }
 
     @Nullable
     @Override
     public BlockState getStateForPlacement(BlockPlaceContext context) {
         FluidState fluidState = context.getLevel().getFluidState(context.getClickedPos());
-        return super.getStateForPlacement(context).setValue(WATERLOGGED, Boolean.valueOf(fluidState.is(Fluids.WATER)));
+        return super.getStateForPlacement(context).setValue(WATERLOGGED, fluidState.is(Fluids.WATER));
     }
 
     @Override
