--- a/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
+++ b/net/minecraft/world/level/block/WeightedPressurePlateBlock.java
@@ -34,7 +_,7 @@
 
     protected WeightedPressurePlateBlock(int maxWeight, BlockSetType type, BlockBehaviour.Properties properties) {
         super(properties, type);
-        this.registerDefaultState(this.stateDefinition.any().setValue(POWER, Integer.valueOf(0)));
+        this.registerDefaultState(this.stateDefinition.any().setValue(POWER, 0));
         this.maxWeight = maxWeight;
     }
 
@@ -76,7 +_,7 @@
 
     @Override
     protected BlockState setSignalForState(BlockState state, int strength) {
-        return state.setValue(POWER, Integer.valueOf(strength));
+        return state.setValue(POWER, strength);
     }
 
     @Override
