--- a/net/minecraft/world/level/levelgen/feature/BambooFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/BambooFeature.java
@@ -15,11 +_,11 @@
 public class BambooFeature extends Feature<ProbabilityFeatureConfiguration> {
     private static final BlockState BAMBOO_TRUNK = Blocks.BAMBOO
         .defaultBlockState()
-        .setValue(BambooStalkBlock.AGE, Integer.valueOf(1))
+        .setValue(BambooStalkBlock.AGE, 1)
         .setValue(BambooStalkBlock.LEAVES, BambooLeaves.NONE)
-        .setValue(BambooStalkBlock.STAGE, Integer.valueOf(0));
+        .setValue(BambooStalkBlock.STAGE, 0);
     private static final BlockState BAMBOO_FINAL_LARGE = BAMBOO_TRUNK.setValue(BambooStalkBlock.LEAVES, BambooLeaves.LARGE)
-        .setValue(BambooStalkBlock.STAGE, Integer.valueOf(1));
+        .setValue(BambooStalkBlock.STAGE, 1);
     private static final BlockState BAMBOO_TOP_LARGE = BAMBOO_TRUNK.setValue(BambooStalkBlock.LEAVES, BambooLeaves.LARGE);
     private static final BlockState BAMBOO_TOP_SMALL = BAMBOO_TRUNK.setValue(BambooStalkBlock.LEAVES, BambooLeaves.SMALL);
 
