--- a/net/minecraft/world/level/levelgen/feature/CoralFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/CoralFeature.java
@@ -44,7 +_,7 @@
                     .map(Holder::value)
                     .ifPresent(block -> level.setBlock(blockPos, block.defaultBlockState(), 2));
             } else if (random.nextFloat() < 0.05F) {
-                level.setBlock(blockPos, Blocks.SEA_PICKLE.defaultBlockState().setValue(SeaPickleBlock.PICKLES, Integer.valueOf(random.nextInt(4) + 1)), 2);
+                level.setBlock(blockPos, Blocks.SEA_PICKLE.defaultBlockState().setValue(SeaPickleBlock.PICKLES, random.nextInt(4) + 1), 2);
             }
 
             for (Direction direction : Direction.Plane.HORIZONTAL) {
