--- a/net/minecraft/world/level/levelgen/feature/DripstoneUtils.java
+++ b/net/minecraft/world/level/levelgen/feature/DripstoneUtils.java
@@ -78,7 +_,7 @@
             BlockPos.MutableBlockPos mutableBlockPos = pos.mutable();
             buildBaseToTipColumn(direction, height, mergeTip, blockState -> {
                 if (blockState.is(Blocks.POINTED_DRIPSTONE)) {
-                    blockState = blockState.setValue(PointedDripstoneBlock.WATERLOGGED, Boolean.valueOf(level.isWaterAt(mutableBlockPos)));
+                    blockState = blockState.setValue(PointedDripstoneBlock.WATERLOGGED, level.isWaterAt(mutableBlockPos));
                 }
 
                 level.setBlock(mutableBlockPos, blockState, 2);
