--- a/net/minecraft/world/level/levelgen/feature/FeatureCountTracker.java
+++ b/net/minecraft/world/level/levelgen/feature/FeatureCountTracker.java
@@ -71,7 +_,7 @@
                                 string1
                                     + " "
                                     + String.format(Locale.ROOT, "%10d ", integer)
-                                    + String.format(Locale.ROOT, "%10f ", (double)integer.intValue() / value.intValue())
+                                    + String.format(Locale.ROOT, "%10f ", (double) integer / value)
                                     + featureData.topFeature().flatMap(registry::getResourceKey).<ResourceLocation>map(ResourceKey::location)
                                     + " "
                                     + featureData.feature().feature()
