--- a/net/minecraft/world/level/levelgen/feature/GeodeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/GeodeFeature.java
@@ -102,7 +_,7 @@
             double d8 = 0.0;
 
             for (Pair<BlockPos, Integer> pair : list) {
-                d7 += Mth.invSqrt(blockPos2.distSqr(pair.getFirst()) + pair.getSecond().intValue()) + d6;
+                d7 += Mth.invSqrt(blockPos2.distSqr(pair.getFirst()) + pair.getSecond()) + d6;
             }
 
             for (BlockPos blockPos3 : list1) {
@@ -155,7 +_,7 @@
                 BlockPos blockPos5 = blockPos1.relative(direction1);
                 BlockState blockState1 = worldGenLevel.getBlockState(blockPos5);
                 if (blockState.hasProperty(BlockStateProperties.WATERLOGGED)) {
-                    blockState = blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(blockState1.getFluidState().isSource()));
+                    blockState = blockState.setValue(BlockStateProperties.WATERLOGGED, blockState1.getFluidState().isSource());
                 }
 
                 if (BuddingAmethystBlock.canClusterGrowAtState(blockState1)) {
