--- a/net/minecraft/world/level/levelgen/feature/HugeRedMushroomFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/HugeRedMushroomFeature.java
@@ -38,11 +_,11 @@
                                 && state.hasProperty(HugeMushroomBlock.NORTH)
                                 && state.hasProperty(HugeMushroomBlock.SOUTH)
                                 && state.hasProperty(HugeMushroomBlock.UP)) {
-                                state = state.setValue(HugeMushroomBlock.UP, Boolean.valueOf(i >= treeHeight - 1))
-                                    .setValue(HugeMushroomBlock.WEST, Boolean.valueOf(i3 < -i2))
-                                    .setValue(HugeMushroomBlock.EAST, Boolean.valueOf(i3 > i2))
-                                    .setValue(HugeMushroomBlock.NORTH, Boolean.valueOf(i4 < -i2))
-                                    .setValue(HugeMushroomBlock.SOUTH, Boolean.valueOf(i4 > i2));
+                                state = state.setValue(HugeMushroomBlock.UP, i >= treeHeight - 1)
+                                    .setValue(HugeMushroomBlock.WEST, i3 < -i2)
+                                    .setValue(HugeMushroomBlock.EAST, i3 > i2)
+                                    .setValue(HugeMushroomBlock.NORTH, i4 < -i2)
+                                    .setValue(HugeMushroomBlock.SOUTH, i4 > i2);
                             }
 
                             this.setBlock(level, mutablePos, state);
