--- a/net/minecraft/world/level/levelgen/feature/HugeBrownMushroomFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/HugeBrownMushroomFeature.java
@@ -39,10 +_,10 @@
                             && state.hasProperty(HugeMushroomBlock.EAST)
                             && state.hasProperty(HugeMushroomBlock.NORTH)
                             && state.hasProperty(HugeMushroomBlock.SOUTH)) {
-                            state = state.setValue(HugeMushroomBlock.WEST, Boolean.valueOf(flag6))
-                                .setValue(HugeMushroomBlock.EAST, Boolean.valueOf(flag7))
-                                .setValue(HugeMushroomBlock.NORTH, Boolean.valueOf(flag8))
-                                .setValue(HugeMushroomBlock.SOUTH, Boolean.valueOf(flag9));
+                            state = state.setValue(HugeMushroomBlock.WEST, flag6)
+                                .setValue(HugeMushroomBlock.EAST, flag7)
+                                .setValue(HugeMushroomBlock.NORTH, flag8)
+                                .setValue(HugeMushroomBlock.SOUTH, flag9);
                         }
 
                         this.setBlock(level, mutablePos, state);
