--- a/net/minecraft/world/level/levelgen/feature/KelpFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/KelpFeature.java
@@ -33,7 +_,7 @@
                     && worldGenLevel.getBlockState(blockPos1.above()).is(Blocks.WATER)
                     && blockState1.canSurvive(worldGenLevel, blockPos1)) {
                     if (i2 == i1) {
-                        worldGenLevel.setBlock(blockPos1, blockState.setValue(KelpBlock.AGE, Integer.valueOf(randomSource.nextInt(4) + 20)), 2);
+                        worldGenLevel.setBlock(blockPos1, blockState.setValue(KelpBlock.AGE, randomSource.nextInt(4) + 20), 2);
                         i++;
                     } else {
                         worldGenLevel.setBlock(blockPos1, blockState1, 2);
@@ -41,7 +_,7 @@
                 } else if (i2 > 0) {
                     BlockPos blockPos2 = blockPos1.below();
                     if (blockState.canSurvive(worldGenLevel, blockPos2) && !worldGenLevel.getBlockState(blockPos2.below()).is(Blocks.KELP)) {
-                        worldGenLevel.setBlock(blockPos2, blockState.setValue(KelpBlock.AGE, Integer.valueOf(randomSource.nextInt(4) + 20)), 2);
+                        worldGenLevel.setBlock(blockPos2, blockState.setValue(KelpBlock.AGE, randomSource.nextInt(4) + 20), 2);
                         i++;
                     }
                     break;
