--- a/net/minecraft/world/level/levelgen/feature/SculkPatchFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/SculkPatchFeature.java
@@ -57,7 +_,7 @@
                 if (worldGenLevel.getBlockState(blockPos2).isAir()
                     && worldGenLevel.getBlockState(blockPos2.below()).isFaceSturdy(worldGenLevel, blockPos2.below(), Direction.UP)) {
                     worldGenLevel.setBlock(
-                        blockPos2, Blocks.SCULK_SHRIEKER.defaultBlockState().setValue(SculkShriekerBlock.CAN_SUMMON, Boolean.valueOf(true)), 3
+                        blockPos2, Blocks.SCULK_SHRIEKER.defaultBlockState().setValue(SculkShriekerBlock.CAN_SUMMON, true), 3
                     );
                 }
             }
