--- a/net/minecraft/world/level/levelgen/feature/SeaPickleFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/SeaPickleFeature.java
@@ -28,7 +_,7 @@
             int i4 = randomSource.nextInt(8) - randomSource.nextInt(8);
             int height = worldGenLevel.getHeight(Heightmap.Types.OCEAN_FLOOR, blockPos.getX() + i3, blockPos.getZ() + i4);
             BlockPos blockPos1 = new BlockPos(blockPos.getX() + i3, height, blockPos.getZ() + i4);
-            BlockState blockState = Blocks.SEA_PICKLE.defaultBlockState().setValue(SeaPickleBlock.PICKLES, Integer.valueOf(randomSource.nextInt(4) + 1));
+            BlockState blockState = Blocks.SEA_PICKLE.defaultBlockState().setValue(SeaPickleBlock.PICKLES, randomSource.nextInt(4) + 1);
             if (worldGenLevel.getBlockState(blockPos1).is(Blocks.WATER) && blockState.canSurvive(worldGenLevel, blockPos1)) {
                 worldGenLevel.setBlock(blockPos1, blockState, 2);
                 i++;
