--- a/net/minecraft/world/level/levelgen/feature/SnowAndFreezeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/SnowAndFreezeFeature.java
@@ -39,7 +_,7 @@
                     worldGenLevel.setBlock(mutableBlockPos, Blocks.SNOW.defaultBlockState(), 2);
                     BlockState blockState = worldGenLevel.getBlockState(mutableBlockPos1);
                     if (blockState.hasProperty(SnowyDirtBlock.SNOWY)) {
-                        worldGenLevel.setBlock(mutableBlockPos1, blockState.setValue(SnowyDirtBlock.SNOWY, Boolean.valueOf(true)), 2);
+                        worldGenLevel.setBlock(mutableBlockPos1, blockState.setValue(SnowyDirtBlock.SNOWY, true), 2);
                     }
                 }
             }
