--- a/net/minecraft/world/level/levelgen/feature/SpikeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/SpikeFeature.java
@@ -97,10 +_,10 @@
                             boolean flag4 = i4 == -2 || i4 == 2 || flag2;
                             BlockState blockState = Blocks.IRON_BARS
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.NORTH, Boolean.valueOf(flag3 && i4 != -2))
-                                .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(flag3 && i4 != 2))
-                                .setValue(IronBarsBlock.WEST, Boolean.valueOf(flag4 && i3 != -2))
-                                .setValue(IronBarsBlock.EAST, Boolean.valueOf(flag4 && i3 != 2));
+                                .setValue(IronBarsBlock.NORTH, flag3 && i4 != -2)
+                                .setValue(IronBarsBlock.SOUTH, flag3 && i4 != 2)
+                                .setValue(IronBarsBlock.WEST, flag4 && i3 != -2)
+                                .setValue(IronBarsBlock.EAST, flag4 && i3 != 2);
                             this.setBlock(level, mutableBlockPos.set(spike.getCenterX() + i3, spike.getHeight() + i5, spike.getCenterZ() + i4), blockState);
                         }
                     }
