--- a/net/minecraft/world/level/levelgen/feature/TreeFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TreeFeature.java
@@ -198,7 +_,7 @@
                 if (box.isInside(blockPos1)) {
                     if (i2 != 0) {
                         BlockState blockState = level.getBlockState(blockPos1);
-                        setBlockKnownShape(level, blockPos1, blockState.setValue(BlockStateProperties.DISTANCE, Integer.valueOf(i2)));
+                        setBlockKnownShape(level, blockPos1, blockState.setValue(BlockStateProperties.DISTANCE, i2));
                     }
 
                     discreteVoxelShape.fill(blockPos1.getX() - box.minX(), blockPos1.getY() - box.minY(), blockPos1.getZ() - box.minZ());
