--- a/net/minecraft/world/level/levelgen/feature/TwistingVinesFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/TwistingVinesFeature.java
@@ -76,7 +_,7 @@
                 if (i == length || !level.isEmptyBlock(pos.above())) {
                     level.setBlock(
                         pos,
-                        Blocks.TWISTING_VINES.defaultBlockState().setValue(GrowingPlantHeadBlock.AGE, Integer.valueOf(Mth.nextInt(random, minAge, maxAge))),
+                        Blocks.TWISTING_VINES.defaultBlockState().setValue(GrowingPlantHeadBlock.AGE, Mth.nextInt(random, minAge, maxAge)),
                         2
                     );
                     break;
