--- a/net/minecraft/world/level/levelgen/feature/WaterloggedVegetationPatchFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/WaterloggedVegetationPatchFeature.java
@@ -60,7 +_,7 @@
         if (super.placeVegetation(level, config, chunkGenerator, random, pos.below())) {
             BlockState blockState = level.getBlockState(pos);
             if (blockState.hasProperty(BlockStateProperties.WATERLOGGED) && !blockState.getValue(BlockStateProperties.WATERLOGGED)) {
-                level.setBlock(pos, blockState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true)), 2);
+                level.setBlock(pos, blockState.setValue(BlockStateProperties.WATERLOGGED, true), 2);
             }
 
             return true;
