--- a/net/minecraft/world/level/levelgen/feature/WeepingVinesFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/WeepingVinesFeature.java
@@ -101,7 +_,7 @@
                 if (i == height || !level.isEmptyBlock(pos.below())) {
                     level.setBlock(
                         pos,
-                        Blocks.WEEPING_VINES.defaultBlockState().setValue(GrowingPlantHeadBlock.AGE, Integer.valueOf(Mth.nextInt(random, minAge, maxAge))),
+                        Blocks.WEEPING_VINES.defaultBlockState().setValue(GrowingPlantHeadBlock.AGE, Mth.nextInt(random, minAge, maxAge)),
                         2
                     );
                     break;
