--- a/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/AbstractFurnaceBlockEntity.java
@@ -282,7 +_,7 @@
 
         if (isLit != furnace.isLit()) {
             flag = true;
-            state = state.setValue(AbstractFurnaceBlock.LIT, Boolean.valueOf(furnace.isLit()));
+            state = state.setValue(AbstractFurnaceBlock.LIT, furnace.isLit());
             level.setBlock(pos, state, 3);
         }
 
