--- a/net/minecraft/world/level/block/entity/BannerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BannerBlockEntity.java
@@ -80,12 +_,12 @@
     }
 
     // Paper start - always send patterns to client
-    ThreadLocal<Boolean> serialisingForNetwork = ThreadLocal.withInitial(() -> Boolean.FALSE);
+    ThreadLocal<Boolean> serialisingForNetwork = ThreadLocal.withInitial(() -> false);
     @Override
     public CompoundTag getUpdateTag(HolderLookup.Provider registries) {
         final Boolean wasSerialisingForNetwork = serialisingForNetwork.get();
         try {
-            serialisingForNetwork.set(Boolean.TRUE);
+            serialisingForNetwork.set(true);
         return this.saveWithoutMetadata(registries);
         } finally {
             serialisingForNetwork.set(wasSerialisingForNetwork);
