--- a/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BarrelBlockEntity.java
@@ -179,7 +_,7 @@
     }
 
     public void updateBlockState(BlockState state, boolean _open) {
-        this.level.setBlock(this.getBlockPos(), state.setValue(BarrelBlock.OPEN, Boolean.valueOf(_open)), 3);
+        this.level.setBlock(this.getBlockPos(), state.setValue(BarrelBlock.OPEN, _open), 3);
     }
 
     public void playSound(BlockState state, SoundEvent sound) {
