--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -316,7 +_,7 @@
     ) {
     // Paper emd - pass beacon block entity
         if (!level.isClientSide && primaryEffect != null) {
-            double d = computeBeaconRange(beaconLevel); // Paper - diff out applyEffects logic components - see below
+            //double d = computeBeaconRange(beaconLevel); // Paper - diff out applyEffects logic components - see below // Plazma - Remove unused variables
             int i = computeEffectAmplifier(beaconLevel, primaryEffect, secondaryEffect); // Paper - diff out applyEffects logic components - see below
 
             int i1 = computeEffectDuration(beaconLevel); // Paper - diff out applyEffects logic components - see below
