--- a/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrewingStandBlockEntity.java
@@ -184,7 +_,7 @@
             }
 
             for (int i = 0; i < BrewingStandBlock.HAS_BOTTLE.length; i++) {
-                blockState = blockState.setValue(BrewingStandBlock.HAS_BOTTLE[i], Boolean.valueOf(potionBits[i]));
+                blockState = blockState.setValue(BrewingStandBlock.HAS_BOTTLE[i], potionBits[i]);
             }
 
             level.setBlock(pos, blockState, 2);
