--- a/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BrushableBlockEntity.java
@@ -74,7 +_,7 @@
                 int completionState1 = this.getCompletionState();
                 if (completionState != completionState1) {
                     BlockState blockState = this.getBlockState();
-                    BlockState blockState1 = blockState.setValue(BlockStateProperties.DUSTED, Integer.valueOf(completionState1));
+                    BlockState blockState1 = blockState.setValue(BlockStateProperties.DUSTED, completionState1);
                     level.setBlock(this.getBlockPos(), blockState1, 3);
                 }
 
@@ -152,7 +_,7 @@
             this.brushCount = Math.max(0, this.brushCount - 2);
             int completionState1 = this.getCompletionState();
             if (completionState != completionState1) {
-                level.setBlock(this.getBlockPos(), this.getBlockState().setValue(BlockStateProperties.DUSTED, Integer.valueOf(completionState1)), 3);
+                level.setBlock(this.getBlockPos(), this.getBlockState().setValue(BlockStateProperties.DUSTED, completionState1), 3);
             }
 
             int i = 4;
