--- a/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ChiseledBookShelfBlockEntity.java
@@ -75,7 +_,7 @@
             for (int i = 0; i < ChiseledBookShelfBlock.SLOT_OCCUPIED_PROPERTIES.size(); i++) {
                 boolean flag = !this.getItem(i).isEmpty();
                 BooleanProperty booleanProperty = ChiseledBookShelfBlock.SLOT_OCCUPIED_PROPERTIES.get(i);
-                blockState = blockState.setValue(booleanProperty, Boolean.valueOf(flag));
+                blockState = blockState.setValue(booleanProperty, flag);
             }
 
             Objects.requireNonNull(this.level).setBlock(this.worldPosition, blockState, 3);
