--- a/net/minecraft/world/level/block/entity/CrafterBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/CrafterBlockEntity.java
@@ -274,7 +_,7 @@
         if (i >= 0) {
             crafter.craftingTicksRemaining = i;
             if (i == 0) {
-                level.setBlock(pos, state.setValue(CrafterBlock.CRAFTING, Boolean.valueOf(false)), 3);
+                level.setBlock(pos, state.setValue(CrafterBlock.CRAFTING, false), 3);
             }
         }
     }
