--- a/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/SculkShriekerBlockEntity.java
@@ -161,7 +_,7 @@
     private void shriek(ServerLevel level, @Nullable Entity sourceEntity) {
         BlockPos blockPos = this.getBlockPos();
         BlockState blockState = this.getBlockState();
-        level.setBlock(blockPos, blockState.setValue(SculkShriekerBlock.SHRIEKING, Boolean.valueOf(true)), 2);
+        level.setBlock(blockPos, blockState.setValue(SculkShriekerBlock.SHRIEKING, true), 2);
         level.scheduleTick(blockPos, blockState.getBlock(), 90);
         level.levelEvent(3007, blockPos, 0);
         level.gameEvent(GameEvent.SHRIEK, blockPos, GameEvent.Context.of(sourceEntity));
