--- a/ca/spottedleaf/dataconverter/types/MapType.java
+++ b/ca/spottedleaf/dataconverter/types/MapType.java
@@ -172,25 +_,25 @@
 
     public default void setGeneric(final K key, final Object value) {
         if (value instanceof Boolean) {
-            this.setBoolean(key, ((Boolean)value).booleanValue());
+            this.setBoolean(key, (Boolean) value);
         } else if (value instanceof Number) {
             if (value instanceof Byte) {
-                this.setByte(key, ((Byte)value).byteValue());
+                this.setByte(key, (Byte) value);
                 return;
             } else if (value instanceof Short) {
-                this.setShort(key, ((Short)value).shortValue());
+                this.setShort(key, (Short) value);
                 return;
             } else if (value instanceof Integer) {
-                this.setInt(key, ((Integer)value).intValue());
+                this.setInt(key, (Integer) value);
                 return;
             } else if (value instanceof Long) {
-                this.setLong(key, ((Long)value).longValue());
+                this.setLong(key, (Long) value);
                 return;
             } else if (value instanceof Float) {
-                this.setFloat(key, ((Float)value).floatValue());
+                this.setFloat(key, (Float) value);
                 return;
             } else if (value instanceof Double) {
-                this.setDouble(key, ((Double)value).doubleValue());
+                this.setDouble(key, (Double) value);
                 return;
             } // else fall through to throw
         } else if (value instanceof MapType) {
