--- a/net/minecraft/network/chat/Style.java
+++ b/net/minecraft/network/chat/Style.java
@@ -105,23 +_,23 @@
     }
 
     public boolean isBold() {
-        return this.bold == Boolean.TRUE;
+        return this.bold == true;
     }
 
     public boolean isItalic() {
-        return this.italic == Boolean.TRUE;
+        return this.italic == true;
     }
 
     public boolean isStrikethrough() {
-        return this.strikethrough == Boolean.TRUE;
+        return this.strikethrough == true;
     }
 
     public boolean isUnderlined() {
-        return this.underlined == Boolean.TRUE;
+        return this.underlined == true;
     }
 
     public boolean isObfuscated() {
-        return this.obfuscated == Boolean.TRUE;
+        return this.obfuscated == true;
     }
 
     public boolean isEmpty() {
