--- a/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
@@ -45,7 +_,7 @@
                                             blockPos1,
                                             Blocks.COCOA
                                                 .defaultBlockState()
-                                                .setValue(CocoaBlock.AGE, Integer.valueOf(randomSource.nextInt(3)))
+                                                .setValue(CocoaBlock.AGE, randomSource.nextInt(3))
                                                 .setValue(CocoaBlock.FACING, direction)
                                         );
                                     }
