--- a/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/CreakingHeartDecorator.java
@@ -50,8 +_,8 @@
                         optional.get(),
                         Blocks.CREAKING_HEART
                             .defaultBlockState()
-                            .setValue(CreakingHeartBlock.ACTIVE, Boolean.valueOf(true))
-                            .setValue(CreakingHeartBlock.NATURAL, Boolean.valueOf(true))
+                            .setValue(CreakingHeartBlock.ACTIVE, true)
+                            .setValue(CreakingHeartBlock.NATURAL, true)
                     );
                 }
             }
