--- a/net/minecraft/world/level/levelgen/feature/treedecorators/PaleMossDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/PaleMossDecorator.java
@@ -83,10 +_,10 @@
 
     private static void addMossHanger(BlockPos pos, TreeDecorator.Context context) {
         while (context.isAir(pos.below()) && !(context.random().nextFloat() < 0.5)) {
-            context.setBlock(pos, Blocks.PALE_HANGING_MOSS.defaultBlockState().setValue(HangingMossBlock.TIP, Boolean.valueOf(false)));
+            context.setBlock(pos, Blocks.PALE_HANGING_MOSS.defaultBlockState().setValue(HangingMossBlock.TIP, false));
             pos = pos.below();
         }
 
-        context.setBlock(pos, Blocks.PALE_HANGING_MOSS.defaultBlockState().setValue(HangingMossBlock.TIP, Boolean.valueOf(true)));
+        context.setBlock(pos, Blocks.PALE_HANGING_MOSS.defaultBlockState().setValue(HangingMossBlock.TIP, true));
     }
 }
