--- a/net/minecraft/world/level/levelgen/feature/treedecorators/TreeDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/TreeDecorator.java
@@ -53,7 +_,7 @@
         }
 
         public void placeVine(BlockPos pos, BooleanProperty sideProperty) {
-            this.setBlock(pos, Blocks.VINE.defaultBlockState().setValue(sideProperty, Boolean.valueOf(true)));
+            this.setBlock(pos, Blocks.VINE.defaultBlockState().setValue(sideProperty, true));
         }
 
         public void setBlock(BlockPos pos, BlockState state) {
