--- a/net/minecraft/advancements/AdvancementRewards.java
+++ b/net/minecraft/advancements/AdvancementRewards.java
@@ -26,7 +_,7 @@
 public record AdvancementRewards(int experience, List<ResourceKey<LootTable>> loot, List<ResourceKey<Recipe<?>>> recipes, Optional<CacheableFunction> function) {
     public static final Codec<AdvancementRewards> CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
-                Codec.INT.optionalFieldOf("experience", Integer.valueOf(0)).forGetter(AdvancementRewards::experience),
+                Codec.INT.optionalFieldOf("experience", 0).forGetter(AdvancementRewards::experience),
                 ResourceKey.codec(Registries.LOOT_TABLE).listOf().optionalFieldOf("loot", List.of()).forGetter(AdvancementRewards::loot),
                 ResourceKey.codec(Registries.RECIPE).listOf().optionalFieldOf("recipes", List.of()).forGetter(AdvancementRewards::recipes),
                 CacheableFunction.CODEC.optionalFieldOf("function").forGetter(AdvancementRewards::function)
