--- a/net/minecraft/advancements/DisplayInfo.java
+++ b/net/minecraft/advancements/DisplayInfo.java
@@ -18,9 +_,9 @@
                 ComponentSerialization.CODEC.fieldOf("description").forGetter(DisplayInfo::getDescription),
                 ResourceLocation.CODEC.optionalFieldOf("background").forGetter(DisplayInfo::getBackground),
                 AdvancementType.CODEC.optionalFieldOf("frame", AdvancementType.TASK).forGetter(DisplayInfo::getType),
-                Codec.BOOL.optionalFieldOf("show_toast", Boolean.valueOf(true)).forGetter(DisplayInfo::shouldShowToast),
-                Codec.BOOL.optionalFieldOf("announce_to_chat", Boolean.valueOf(true)).forGetter(DisplayInfo::shouldAnnounceChat),
-                Codec.BOOL.optionalFieldOf("hidden", Boolean.valueOf(false)).forGetter(DisplayInfo::isHidden)
+                Codec.BOOL.optionalFieldOf("show_toast", true).forGetter(DisplayInfo::shouldShowToast),
+                Codec.BOOL.optionalFieldOf("announce_to_chat", true).forGetter(DisplayInfo::shouldAnnounceChat),
+                Codec.BOOL.optionalFieldOf("hidden", false).forGetter(DisplayInfo::isHidden)
             )
             .apply(instance, DisplayInfo::new)
     );
