--- a/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/player/RegionizedPlayerChunkLoader.java
@@ -319,7 +_,7 @@
 
         private static final AtomicLong ID_GENERATOR = new AtomicLong();
         private final long id = ID_GENERATOR.incrementAndGet();
-        private final Long idBoxed = Long.valueOf(this.id);
+        private final Long idBoxed = this.id;
 
         private static final long MAX_RATE = 10_000L;
 
@@ -516,7 +_,7 @@
 
         private static int getClientViewDistance(final ServerPlayer player) {
             final Integer vd = player.requestedViewDistance();
-            return vd == null ? -1 : Math.max(0, vd.intValue());
+            return vd == null ? -1 : Math.max(0, vd);
         }
 
         private static int getTickDistance(final int playerTickViewDistance, final int worldTickViewDistance,
