--- a/net/minecraft/data/loot/packs/VanillaBlockLoot.java
+++ b/net/minecraft/data/loot/packs/VanillaBlockLoot.java
@@ -914,11 +_,11 @@
                                 LootItem.lootTableItem(block)
                                     .apply(
                                         List.of(2, 3, 4),
-                                        integer -> SetItemCountFunction.setCount(ConstantValue.exactly(integer.intValue()))
+                                        integer -> SetItemCountFunction.setCount(ConstantValue.exactly(integer))
                                             .when(
                                                 LootItemBlockStatePropertyCondition.hasBlockStateProperties(block)
                                                     .setProperties(
-                                                        StatePropertiesPredicate.Builder.properties().hasProperty(SeaPickleBlock.PICKLES, integer.intValue())
+                                                        StatePropertiesPredicate.Builder.properties().hasProperty(SeaPickleBlock.PICKLES, integer)
                                                     )
                                             )
                                     )
@@ -1324,10 +_,10 @@
                                             .when(
                                                 LootItemBlockStatePropertyCondition.hasBlockStateProperties(block)
                                                     .setProperties(
-                                                        StatePropertiesPredicate.Builder.properties().hasProperty(SnowLayerBlock.LAYERS, integer.intValue())
+                                                        StatePropertiesPredicate.Builder.properties().hasProperty(SnowLayerBlock.LAYERS, integer)
                                                     )
                                             )
-                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(integer.intValue())))
+                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(integer)))
                                     )
                                     .when(this.doesNotHaveSilkTouch()),
                                 AlternativesEntry.alternatives(
@@ -1335,11 +_,11 @@
                                     integer -> integer == 8
                                         ? LootItem.lootTableItem(Blocks.SNOW_BLOCK)
                                         : LootItem.lootTableItem(Blocks.SNOW)
-                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(integer.intValue())))
+                                            .apply(SetItemCountFunction.setCount(ConstantValue.exactly(integer)))
                                             .when(
                                                 LootItemBlockStatePropertyCondition.hasBlockStateProperties(block)
                                                     .setProperties(
-                                                        StatePropertiesPredicate.Builder.properties().hasProperty(SnowLayerBlock.LAYERS, integer.intValue())
+                                                        StatePropertiesPredicate.Builder.properties().hasProperty(SnowLayerBlock.LAYERS, integer)
                                                     )
                                             )
                                 )
@@ -1548,7 +_,7 @@
                                     LootItemBlockStatePropertyCondition.Builder builder1 = LootItemBlockStatePropertyCondition.hasBlockStateProperties(
                                             Blocks.PITCHER_CROP
                                         )
-                                        .setProperties(StatePropertiesPredicate.Builder.properties().hasProperty(PitcherCropBlock.AGE, integer.intValue()));
+                                        .setProperties(StatePropertiesPredicate.Builder.properties().hasProperty(PitcherCropBlock.AGE, integer));
                                     return integer == 4
                                         ? LootItem.lootTableItem(Items.PITCHER_PLANT)
                                             .when(builder1)
