--- a/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java
+++ b/ca/spottedleaf/moonrise/patches/starlight/light/StarLightInterface.java
@@ -688,7 +_,7 @@
         public synchronized ClientChunkTasks queueSectionChange(final SectionPos pos, final boolean newEmptyValue) {
             final ClientChunkTasks tasks = this.getOrCreate(CoordinateUtils.getChunkKey(pos));
 
-            tasks.setChangedSection(pos.getY(), Boolean.valueOf(newEmptyValue));
+            tasks.setChangedSection(pos.getY(), newEmptyValue);
 
             return tasks;
         }
@@ -807,7 +_,7 @@
                     );
                 }
 
-                valueInMap.setChangedSection(pos.getY(), Boolean.valueOf(newEmptyValue));
+                valueInMap.setChangedSection(pos.getY(), newEmptyValue);
 
                 return valueInMap;
             });
