From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 15 Sep 2025 02:03:42 +0900
Subject: [PATCH] Optimize default configurations


diff --git a/src/main/java/org/purpurmc/purpur/PurpurConfig.java b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
index 31bf765d2dfa75d854f7f8e55bab7792ace843ab..1813164e433849cc65ffdf62ffe06b76e47631cb 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurConfig.java
@@ -240,7 +240,7 @@ public class PurpurConfig {
         laggingThreshold = getDouble("settings.lagging-threshold", laggingThreshold);
     }
 
-    public static boolean useAlternateKeepAlive = false;
+    public static boolean useAlternateKeepAlive = org.plazmamc.plazma.Options.OPTIMIZE; // Plazma - Optimize default configuration
     private static void useAlternateKeepAlive() {
         useAlternateKeepAlive = getBoolean("settings.use-alternate-keepalive", useAlternateKeepAlive);
     }
diff --git a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
index a03e166b490952534e8050654c1afa975795f731..f82ed8ba3e491233480ae28fc83ca105139b5d46 100644
--- a/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
+++ b/src/main/java/org/purpurmc/purpur/PurpurWorldConfig.java
@@ -145,7 +145,7 @@ public class PurpurWorldConfig {
     public boolean disableDropsOnCrammingDeath = false;
     public boolean milkCuresBadOmen = true;
     public double tridentLoyaltyVoidReturnHeight = 0.0D;
-    public boolean entitiesCanUsePortals = true;
+    public boolean entitiesCanUsePortals = !org.plazmamc.plazma.Options.OPTIMIZE; // Plazma - Optimize default configuration
     public int raidCooldownSeconds = 0;
     public int animalBreedingCooldownSeconds = 0;
     public boolean persistentDroppableEntityDisplayNames = true;
@@ -451,7 +451,7 @@ public class PurpurWorldConfig {
     public boolean idleTimeoutTargetPlayer = true;
     public String playerDeathExpDropEquation = "expLevel * 7";
     public int playerDeathExpDropMax = 100;
-    public boolean teleportIfOutsideBorder = false;
+    public boolean teleportIfOutsideBorder = org.plazmamc.plazma.Options.OPTIMIZE; // Plazma - Optimize default configuration
     public boolean teleportOnNetherCeilingDamage = false;
     public boolean totemOfUndyingWorksInInventory = false;
     public boolean playerFixStuckPortal = false;
@@ -1555,7 +1555,7 @@ public class PurpurWorldConfig {
     public float dolphinSpitDamage = 2.0F;
     public double dolphinMaxHealth = 10.0D;
     public double dolphinScale = 1.0D;
-    public boolean dolphinDisableTreasureSearching = false;
+    public boolean dolphinDisableTreasureSearching = org.plazmamc.plazma.Options.OPTIMIZE; // Plazma - Optimize default configuration
     public boolean dolphinTakeDamageFromWater = false;
     public double dolphinNaturallyAggressiveToPlayersChance = 0.0D;
     public boolean dolphinAlwaysDropExp = false;
@@ -3158,14 +3158,14 @@ public class PurpurWorldConfig {
     public boolean villagerAllowTrading = true;
     public boolean villagerAlwaysDropExp = false;
     public int villagerMinimumDemand = 0;
-    public boolean villagerLobotomizeEnabled = false;
+    public boolean villagerLobotomizeEnabled = org.plazmamc.plazma.Options.OPTIMIZE; // Plazma - Optimize default configuration
     public int villagerLobotomizeCheckInterval = 100;
     public boolean villagerLobotomizeWaitUntilTradeLocked = false;
     public boolean villagerDisplayTradeItem = true;
     public int villagerSpawnIronGolemRadius = 0;
     public int villagerSpawnIronGolemLimit = 0;
-    public int villagerAcquirePoiSearchRadius = 48;
-    public int villagerNearestBedSensorSearchRadius = 48;
+    public int villagerAcquirePoiSearchRadius = org.plazmamc.plazma.Options.OPTIMIZE ? 16 : 48; // Plazma - Optimize default configuration
+    public int villagerNearestBedSensorSearchRadius = org.plazmamc.plazma.Options.OPTIMIZE ? 16 : 48; // Plazma - Optimize default configuration
     private void villagerSettings() {
         villagerRidable = getBoolean("mobs.villager.ridable", villagerRidable);
         villagerRidableInWater = getBoolean("mobs.villager.ridable-in-water", villagerRidableInWater);
@@ -3430,7 +3430,7 @@ public class PurpurWorldConfig {
     public boolean zombieJockeyOnlyBaby = true;
     public double zombieJockeyChance = 0.05D;
     public boolean zombieJockeyTryExistingChickens = true;
-    public boolean zombieAggressiveTowardsVillagerWhenLagging = true;
+    public boolean zombieAggressiveTowardsVillagerWhenLagging = !org.plazmamc.plazma.Options.OPTIMIZE; // Plazma - Optimize default configuration
     public Boolean zombieMobGriefingOverride = null;
     public boolean zombieTakeDamageFromWater = false;
     public boolean zombieAlwaysDropExp = false;
