--- a/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java
@@ -165,13 +_,13 @@
         LevelSimulatedReader level, FoliagePlacer.FoliageSetter foliageSetter, RandomSource random, TreeConfiguration treeConfiguration, BlockPos pos
     ) {
         boolean isStateAtPosition = level.isStateAtPosition(
-            pos, blockState -> blockState.getValueOrElse(BlockStateProperties.PERSISTENT, Boolean.valueOf(false))
+            pos, blockState -> blockState.getValueOrElse(BlockStateProperties.PERSISTENT, false)
         );
         if (!isStateAtPosition && TreeFeature.validTreePos(level, pos)) {
             BlockState state = treeConfiguration.foliageProvider.getState(random, pos);
             if (state.hasProperty(BlockStateProperties.WATERLOGGED)) {
                 state = state.setValue(
-                    BlockStateProperties.WATERLOGGED, Boolean.valueOf(level.isFluidAtPosition(pos, fluidState -> fluidState.isSourceOfType(Fluids.WATER)))
+                    BlockStateProperties.WATERLOGGED, level.isFluidAtPosition(pos, fluidState -> fluidState.isSourceOfType(Fluids.WATER))
                 );
             }
 
