--- a/io/papermc/paper/redstone/RedstoneWireTurbo.java
+++ b/io/papermc/paper/redstone/RedstoneWireTurbo.java
@@ -904,7 +_,7 @@
             // need to be known by some nearby non-redstone-wire block.
             BlockPos pos = new BlockPos(upd.self.getX(), upd.self.getY(), upd.self.getZ());
             if (wire.canSurvive(null, worldIn, pos)) {
-                state = state.setValue(RedStoneWireBlock.POWER, Integer.valueOf(j));
+                state = state.setValue(RedStoneWireBlock.POWER, j);
                 // [Space Walker] suppress shape updates and emit those manually to
                 // bypass the new neighbor update stack.
                 if (worldIn.setBlock(upd.self, state, Block.UPDATE_KNOWN_SHAPE | Block.UPDATE_CLIENTS))
