--- a/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/JungleTemplePiece.java
@@ -168,7 +_,7 @@
                 Blocks.TRIPWIRE_HOOK
                     .defaultBlockState()
                     .setValue(TripWireHookBlock.FACING, Direction.EAST)
-                    .setValue(TripWireHookBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireHookBlock.ATTACHED, true),
                 1,
                 -3,
                 8,
@@ -179,7 +_,7 @@
                 Blocks.TRIPWIRE_HOOK
                     .defaultBlockState()
                     .setValue(TripWireHookBlock.FACING, Direction.WEST)
-                    .setValue(TripWireHookBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireHookBlock.ATTACHED, true),
                 4,
                 -3,
                 8,
@@ -189,9 +_,9 @@
                 level,
                 Blocks.TRIPWIRE
                     .defaultBlockState()
-                    .setValue(TripWireBlock.EAST, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.WEST, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireBlock.EAST, true)
+                    .setValue(TripWireBlock.WEST, true)
+                    .setValue(TripWireBlock.ATTACHED, true),
                 2,
                 -3,
                 8,
@@ -201,9 +_,9 @@
                 level,
                 Blocks.TRIPWIRE
                     .defaultBlockState()
-                    .setValue(TripWireBlock.EAST, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.WEST, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireBlock.EAST, true)
+                    .setValue(TripWireBlock.WEST, true)
+                    .setValue(TripWireBlock.ATTACHED, true),
                 3,
                 -3,
                 8,
@@ -246,13 +_,13 @@
                 this.placedTrap1 = this.createDispenser(level, box, random, 3, -2, 1, Direction.NORTH, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
             }
 
-            this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.SOUTH, Boolean.valueOf(true)), 3, -2, 2, box);
+            this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.SOUTH, true), 3, -2, 2, box);
             this.placeBlock(
                 level,
                 Blocks.TRIPWIRE_HOOK
                     .defaultBlockState()
                     .setValue(TripWireHookBlock.FACING, Direction.NORTH)
-                    .setValue(TripWireHookBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireHookBlock.ATTACHED, true),
                 7,
                 -3,
                 1,
@@ -263,7 +_,7 @@
                 Blocks.TRIPWIRE_HOOK
                     .defaultBlockState()
                     .setValue(TripWireHookBlock.FACING, Direction.SOUTH)
-                    .setValue(TripWireHookBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireHookBlock.ATTACHED, true),
                 7,
                 -3,
                 5,
@@ -273,9 +_,9 @@
                 level,
                 Blocks.TRIPWIRE
                     .defaultBlockState()
-                    .setValue(TripWireBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.SOUTH, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireBlock.NORTH, true)
+                    .setValue(TripWireBlock.SOUTH, true)
+                    .setValue(TripWireBlock.ATTACHED, true),
                 7,
                 -3,
                 2,
@@ -285,9 +_,9 @@
                 level,
                 Blocks.TRIPWIRE
                     .defaultBlockState()
-                    .setValue(TripWireBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.SOUTH, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireBlock.NORTH, true)
+                    .setValue(TripWireBlock.SOUTH, true)
+                    .setValue(TripWireBlock.ATTACHED, true),
                 7,
                 -3,
                 3,
@@ -297,9 +_,9 @@
                 level,
                 Blocks.TRIPWIRE
                     .defaultBlockState()
-                    .setValue(TripWireBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.SOUTH, Boolean.valueOf(true))
-                    .setValue(TripWireBlock.ATTACHED, Boolean.valueOf(true)),
+                    .setValue(TripWireBlock.NORTH, true)
+                    .setValue(TripWireBlock.SOUTH, true)
+                    .setValue(TripWireBlock.ATTACHED, true),
                 7,
                 -3,
                 4,
@@ -344,8 +_,8 @@
                 this.placedTrap2 = this.createDispenser(level, box, random, 9, -2, 3, Direction.WEST, BuiltInLootTables.JUNGLE_TEMPLE_DISPENSER);
             }
 
-            this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, Boolean.valueOf(true)), 8, -1, 3, box);
-            this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, Boolean.valueOf(true)), 8, -2, 3, box);
+            this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, true), 8, -1, 3, box);
+            this.placeBlock(level, Blocks.VINE.defaultBlockState().setValue(VineBlock.EAST, true), 8, -2, 3, box);
             if (!this.placedMainChest) {
                 this.placedMainChest = this.createChest(level, box, random, 8, -3, 3, BuiltInLootTables.JUNGLE_TEMPLE);
             }
