--- a/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/MineshaftPieces.java
@@ -524,8 +_,8 @@
             if (this.isSupportingBox(level, box, minX, maxX, maxY, z)) {
                 BlockState planksState = this.type.getPlanksState();
                 BlockState fenceState = this.type.getFenceState();
-                this.generateBox(level, box, minX, minY, z, minX, maxY - 1, z, fenceState.setValue(FenceBlock.WEST, Boolean.valueOf(true)), CAVE_AIR, false);
-                this.generateBox(level, box, maxX, minY, z, maxX, maxY - 1, z, fenceState.setValue(FenceBlock.EAST, Boolean.valueOf(true)), CAVE_AIR, false);
+                this.generateBox(level, box, minX, minY, z, minX, maxY - 1, z, fenceState.setValue(FenceBlock.WEST, true), CAVE_AIR, false);
+                this.generateBox(level, box, maxX, minY, z, maxX, maxY - 1, z, fenceState.setValue(FenceBlock.EAST, true), CAVE_AIR, false);
                 if (random.nextInt(4) == 0) {
                     this.generateBox(level, box, minX, maxY, z, minX, maxY, z, planksState, CAVE_AIR, false);
                     this.generateBox(level, box, maxX, maxY, z, maxX, maxY, z, planksState, CAVE_AIR, false);
