--- a/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/NetherFortressPieces.java
@@ -298,10 +_,10 @@
 
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
-            BlockState blockState1 = blockState.setValue(FenceBlock.EAST, Boolean.valueOf(true));
-            BlockState blockState2 = blockState.setValue(FenceBlock.WEST, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
+            BlockState blockState1 = blockState.setValue(FenceBlock.EAST, true);
+            BlockState blockState2 = blockState.setValue(FenceBlock.WEST, true);
             this.generateBox(level, box, 0, 1, 1, 0, 4, 1, blockState1, blockState1, false);
             this.generateBox(level, box, 0, 3, 4, 0, 4, 4, blockState1, blockState1, false);
             this.generateBox(level, box, 0, 3, 14, 0, 4, 14, blockState1, blockState1, false);
@@ -354,8 +_,8 @@
             BlockState blockState = Blocks.NETHER_BRICK_STAIRS.defaultBlockState().setValue(StairBlock.FACING, Direction.SOUTH);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
 
             for (int i = 0; i <= 9; i++) {
                 int max = Math.max(1, 7 - i);
@@ -435,12 +_,12 @@
         ) {
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             this.generateBox(level, box, 0, 0, 0, 8, 1, 8, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             this.generateBox(level, box, 0, 2, 0, 8, 5, 8, Blocks.AIR.defaultBlockState(), Blocks.AIR.defaultBlockState(), false);
             this.generateBox(level, box, 0, 6, 0, 8, 6, 5, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
@@ -456,8 +_,8 @@
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true))
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.EAST, true)
+                    .setValue(FenceBlock.SOUTH, true),
                 0,
                 3,
                 8,
@@ -467,8 +_,8 @@
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.WEST, true)
+                    .setValue(FenceBlock.SOUTH, true),
                 8,
                 3,
                 8,
@@ -543,12 +_,12 @@
             this.generateBox(level, box, 5, 8, 0, 7, 8, 0, Blocks.NETHER_BRICK_FENCE.defaultBlockState(), Blocks.NETHER_BRICK_FENCE.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
 
             for (int i = 1; i <= 11; i += 2) {
                 this.generateBox(level, box, i, 10, 0, i, 11, 0, blockState, blockState, false);
@@ -571,41 +_,41 @@
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true)),
-                0,
-                13,
-                0,
-                box
-            );
-            this.placeBlock(
-                level,
-                Blocks.NETHER_BRICK_FENCE
-                    .defaultBlockState()
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true)),
-                0,
-                13,
-                12,
-                box
-            );
-            this.placeBlock(
-                level,
-                Blocks.NETHER_BRICK_FENCE
-                    .defaultBlockState()
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true)),
-                12,
-                13,
-                12,
-                box
-            );
-            this.placeBlock(
-                level,
-                Blocks.NETHER_BRICK_FENCE
-                    .defaultBlockState()
-                    .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.NORTH, true)
+                    .setValue(FenceBlock.EAST, true),
+                0,
+                13,
+                0,
+                box
+            );
+            this.placeBlock(
+                level,
+                Blocks.NETHER_BRICK_FENCE
+                    .defaultBlockState()
+                    .setValue(FenceBlock.SOUTH, true)
+                    .setValue(FenceBlock.EAST, true),
+                0,
+                13,
+                12,
+                box
+            );
+            this.placeBlock(
+                level,
+                Blocks.NETHER_BRICK_FENCE
+                    .defaultBlockState()
+                    .setValue(FenceBlock.SOUTH, true)
+                    .setValue(FenceBlock.WEST, true),
+                12,
+                13,
+                12,
+                box
+            );
+            this.placeBlock(
+                level,
+                Blocks.NETHER_BRICK_FENCE
+                    .defaultBlockState()
+                    .setValue(FenceBlock.NORTH, true)
+                    .setValue(FenceBlock.WEST, true),
                 12,
                 13,
                 0,
@@ -622,8 +_,8 @@
                     1,
                     8,
                     ix,
-                    blockState1.setValue(FenceBlock.WEST, Boolean.valueOf(true)),
-                    blockState1.setValue(FenceBlock.WEST, Boolean.valueOf(true)),
+                    blockState1.setValue(FenceBlock.WEST, true),
+                    blockState1.setValue(FenceBlock.WEST, true),
                     false
                 );
                 this.generateBox(
@@ -635,8 +_,8 @@
                     11,
                     8,
                     ix,
-                    blockState1.setValue(FenceBlock.EAST, Boolean.valueOf(true)),
-                    blockState1.setValue(FenceBlock.EAST, Boolean.valueOf(true)),
+                    blockState1.setValue(FenceBlock.EAST, true),
+                    blockState1.setValue(FenceBlock.EAST, true),
                     false
                 );
             }
@@ -780,12 +_,12 @@
             this.generateBox(level, box, 0, 2, 0, 4, 5, 4, Blocks.AIR.defaultBlockState(), Blocks.AIR.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
             this.generateBox(level, box, 4, 2, 0, 4, 5, 4, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             this.generateBox(level, box, 4, 3, 1, 4, 4, 1, blockState1, blockState1, false);
             this.generateBox(level, box, 4, 3, 3, 4, 4, 3, blockState1, blockState1, false);
@@ -850,8 +_,8 @@
             this.generateBox(level, box, 0, 2, 0, 4, 5, 4, Blocks.AIR.defaultBlockState(), Blocks.AIR.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
             this.generateBox(level, box, 0, 2, 0, 0, 5, 4, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             this.generateBox(level, box, 4, 2, 0, 4, 5, 4, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             this.generateBox(level, box, 0, 3, 1, 0, 4, 1, blockState, blockState, false);
@@ -919,12 +_,12 @@
             this.generateBox(level, box, 0, 2, 0, 4, 5, 4, Blocks.AIR.defaultBlockState(), Blocks.AIR.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
             this.generateBox(level, box, 0, 2, 0, 0, 5, 4, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             this.generateBox(level, box, 0, 3, 1, 0, 4, 1, blockState1, blockState1, false);
             this.generateBox(level, box, 0, 3, 3, 0, 4, 3, blockState1, blockState1, false);
@@ -997,14 +_,14 @@
             this.generateBox(level, box, 2, 11, 2, 10, 12, 10, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
-            BlockState blockState2 = blockState1.setValue(FenceBlock.WEST, Boolean.valueOf(true));
-            BlockState blockState3 = blockState1.setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
+            BlockState blockState2 = blockState1.setValue(FenceBlock.WEST, true);
+            BlockState blockState3 = blockState1.setValue(FenceBlock.EAST, true);
 
             for (int i = 1; i <= 11; i += 2) {
                 this.generateBox(level, box, i, 10, 0, i, 11, 0, blockState, blockState, false);
@@ -1027,41 +_,41 @@
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true)),
-                0,
-                13,
-                0,
-                box
-            );
-            this.placeBlock(
-                level,
-                Blocks.NETHER_BRICK_FENCE
-                    .defaultBlockState()
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true)),
-                0,
-                13,
-                12,
-                box
-            );
-            this.placeBlock(
-                level,
-                Blocks.NETHER_BRICK_FENCE
-                    .defaultBlockState()
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true)),
-                12,
-                13,
-                12,
-                box
-            );
-            this.placeBlock(
-                level,
-                Blocks.NETHER_BRICK_FENCE
-                    .defaultBlockState()
-                    .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.NORTH, true)
+                    .setValue(FenceBlock.EAST, true),
+                0,
+                13,
+                0,
+                box
+            );
+            this.placeBlock(
+                level,
+                Blocks.NETHER_BRICK_FENCE
+                    .defaultBlockState()
+                    .setValue(FenceBlock.SOUTH, true)
+                    .setValue(FenceBlock.EAST, true),
+                0,
+                13,
+                12,
+                box
+            );
+            this.placeBlock(
+                level,
+                Blocks.NETHER_BRICK_FENCE
+                    .defaultBlockState()
+                    .setValue(FenceBlock.SOUTH, true)
+                    .setValue(FenceBlock.WEST, true),
+                12,
+                13,
+                12,
+                box
+            );
+            this.placeBlock(
+                level,
+                Blocks.NETHER_BRICK_FENCE
+                    .defaultBlockState()
+                    .setValue(FenceBlock.NORTH, true)
+                    .setValue(FenceBlock.WEST, true),
                 12,
                 13,
                 0,
@@ -1200,20 +_,20 @@
             this.generateBox(level, box, 1, 5, 8, 5, 5, 8, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
-            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, Boolean.valueOf(true)), 1, 6, 3, box);
-            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, Boolean.valueOf(true)), 5, 6, 3, box);
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
+            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, true), 1, 6, 3, box);
+            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, true), 5, 6, 3, box);
             this.placeBlock(
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true))
-                    .setValue(FenceBlock.NORTH, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.EAST, true)
+                    .setValue(FenceBlock.NORTH, true),
                 0,
                 6,
                 3,
@@ -1223,8 +_,8 @@
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                    .setValue(FenceBlock.NORTH, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.WEST, true)
+                    .setValue(FenceBlock.NORTH, true),
                 6,
                 6,
                 3,
@@ -1236,8 +_,8 @@
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true))
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.EAST, true)
+                    .setValue(FenceBlock.SOUTH, true),
                 0,
                 6,
                 8,
@@ -1247,20 +_,20 @@
                 level,
                 Blocks.NETHER_BRICK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true)),
+                    .setValue(FenceBlock.WEST, true)
+                    .setValue(FenceBlock.SOUTH, true),
                 6,
                 6,
                 8,
                 box
             );
             this.generateBox(level, box, 1, 6, 8, 5, 6, 8, blockState, blockState, false);
-            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, Boolean.valueOf(true)), 1, 7, 8, box);
+            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, true), 1, 7, 8, box);
             this.generateBox(level, box, 2, 7, 8, 4, 7, 8, blockState, blockState, false);
-            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, Boolean.valueOf(true)), 5, 7, 8, box);
-            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, Boolean.valueOf(true)), 2, 8, 8, box);
+            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, true), 5, 7, 8, box);
+            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, true), 2, 8, 8, box);
             this.placeBlock(level, blockState, 3, 8, 8, box);
-            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, Boolean.valueOf(true)), 4, 8, 8, box);
+            this.placeBlock(level, Blocks.NETHER_BRICK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, true), 4, 8, 8, box);
             if (!this.hasPlacedSpawner) {
                 BlockPos worldPos = this.getWorldPos(3, 5, 5);
                 if (box.isInside(worldPos)) {
@@ -1650,12 +_,12 @@
             this.generateBox(level, box, 6, 2, 5, 6, 6, 6, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
             this.generateBox(level, box, 2, 6, 0, 4, 6, 0, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             this.generateBox(level, box, 2, 5, 0, 4, 5, 0, blockState, blockState, false);
             this.generateBox(level, box, 2, 6, 6, 4, 6, 6, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
@@ -1718,12 +_,12 @@
             this.generateBox(level, box, 1, 2, 6, 5, 8, 6, Blocks.NETHER_BRICKS.defaultBlockState(), Blocks.NETHER_BRICKS.defaultBlockState(), false);
             BlockState blockState = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                .setValue(FenceBlock.WEST, true)
+                .setValue(FenceBlock.EAST, true);
             BlockState blockState1 = Blocks.NETHER_BRICK_FENCE
                 .defaultBlockState()
-                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(FenceBlock.NORTH, true)
+                .setValue(FenceBlock.SOUTH, true);
             this.generateBox(level, box, 0, 3, 2, 0, 5, 4, blockState1, blockState1, false);
             this.generateBox(level, box, 6, 3, 2, 6, 5, 2, blockState1, blockState1, false);
             this.generateBox(level, box, 6, 3, 4, 6, 5, 4, blockState1, blockState1, false);
