--- a/net/minecraft/world/level/levelgen/structure/structures/RuinedPortalPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/RuinedPortalPiece.java
@@ -184,7 +_,7 @@
             if (blockState1.isAir()) {
                 if (Block.isFaceFull(blockState.getCollisionShape(level, pos), randomHorizontalDirection)) {
                     BooleanProperty propertyForFace = VineBlock.getPropertyForFace(randomHorizontalDirection.getOpposite());
-                    level.setBlock(blockPos, Blocks.VINE.defaultBlockState().setValue(propertyForFace, Boolean.valueOf(true)), 3);
+                    level.setBlock(blockPos, Blocks.VINE.defaultBlockState().setValue(propertyForFace, true), 3);
                 }
             }
         }
@@ -192,7 +_,7 @@
 
     private void maybeAddLeavesAbove(RandomSource random, LevelAccessor level, BlockPos pos) {
         if (random.nextFloat() < 0.5F && level.getBlockState(pos).is(Blocks.NETHERRACK) && level.getBlockState(pos.above()).isAir()) {
-            level.setBlock(pos.above(), Blocks.JUNGLE_LEAVES.defaultBlockState().setValue(LeavesBlock.PERSISTENT, Boolean.valueOf(true)), 3);
+            level.setBlock(pos.above(), Blocks.JUNGLE_LEAVES.defaultBlockState().setValue(LeavesBlock.PERSISTENT, true), 3);
         }
     }
 
