--- a/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/StrongholdPieces.java
@@ -619,19 +_,19 @@
                 this.placeBlock(level, Blocks.OAK_PLANKS.defaultBlockState(), 9, 5, 10, box);
                 BlockState blockState = Blocks.OAK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.WEST, Boolean.valueOf(true))
-                    .setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                    .setValue(FenceBlock.WEST, true)
+                    .setValue(FenceBlock.EAST, true);
                 BlockState blockState1 = Blocks.OAK_FENCE
                     .defaultBlockState()
-                    .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                    .setValue(FenceBlock.SOUTH, Boolean.valueOf(true));
+                    .setValue(FenceBlock.NORTH, true)
+                    .setValue(FenceBlock.SOUTH, true);
                 this.generateBox(level, box, 3, 6, 3, 3, 6, 11, blockState1, blockState1, false);
                 this.generateBox(level, box, 10, 6, 3, 10, 6, 9, blockState1, blockState1, false);
                 this.generateBox(level, box, 4, 6, 2, 9, 6, 2, blockState, blockState, false);
                 this.generateBox(level, box, 4, 6, 12, 7, 6, 12, blockState, blockState, false);
                 this.placeBlock(
                     level,
-                    Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.NORTH, Boolean.valueOf(true)).setValue(FenceBlock.EAST, Boolean.valueOf(true)),
+                    Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.NORTH, true).setValue(FenceBlock.EAST, true),
                     3,
                     6,
                     2,
@@ -639,7 +_,7 @@
                 );
                 this.placeBlock(
                     level,
-                    Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.SOUTH, Boolean.valueOf(true)).setValue(FenceBlock.EAST, Boolean.valueOf(true)),
+                    Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.SOUTH, true).setValue(FenceBlock.EAST, true),
                     3,
                     6,
                     12,
@@ -647,7 +_,7 @@
                 );
                 this.placeBlock(
                     level,
-                    Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.NORTH, Boolean.valueOf(true)).setValue(FenceBlock.WEST, Boolean.valueOf(true)),
+                    Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.NORTH, true).setValue(FenceBlock.WEST, true),
                     10,
                     6,
                     2,
@@ -657,7 +_,7 @@
                 for (int i4 = 0; i4 <= 2; i4++) {
                     this.placeBlock(
                         level,
-                        Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.SOUTH, Boolean.valueOf(true)).setValue(FenceBlock.WEST, Boolean.valueOf(true)),
+                        Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.SOUTH, true).setValue(FenceBlock.WEST, true),
                         8 + i4,
                         6,
                         12 - i4,
@@ -668,8 +_,8 @@
                             level,
                             Blocks.OAK_FENCE
                                 .defaultBlockState()
-                                .setValue(FenceBlock.NORTH, Boolean.valueOf(true))
-                                .setValue(FenceBlock.EAST, Boolean.valueOf(true)),
+                                .setValue(FenceBlock.NORTH, true)
+                                .setValue(FenceBlock.EAST, true),
                             8 + i4,
                             6,
                             11 - i4,
@@ -688,21 +_,21 @@
                 this.placeBlock(level, blockState2, 10, 7, 13, box);
                 int i5 = 7;
                 int i6 = 7;
-                BlockState blockState3 = Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                BlockState blockState3 = Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.EAST, true);
                 this.placeBlock(level, blockState3, 6, 9, 7, box);
-                BlockState blockState4 = Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, Boolean.valueOf(true));
+                BlockState blockState4 = Blocks.OAK_FENCE.defaultBlockState().setValue(FenceBlock.WEST, true);
                 this.placeBlock(level, blockState4, 7, 9, 7, box);
                 this.placeBlock(level, blockState3, 6, 8, 7, box);
                 this.placeBlock(level, blockState4, 7, 8, 7, box);
-                BlockState blockState5 = blockState1.setValue(FenceBlock.WEST, Boolean.valueOf(true)).setValue(FenceBlock.EAST, Boolean.valueOf(true));
+                BlockState blockState5 = blockState1.setValue(FenceBlock.WEST, true).setValue(FenceBlock.EAST, true);
                 this.placeBlock(level, blockState5, 6, 7, 7, box);
                 this.placeBlock(level, blockState5, 7, 7, 7, box);
                 this.placeBlock(level, blockState3, 5, 7, 7, box);
                 this.placeBlock(level, blockState4, 8, 7, 7, box);
-                this.placeBlock(level, blockState3.setValue(FenceBlock.NORTH, Boolean.valueOf(true)), 6, 7, 6, box);
-                this.placeBlock(level, blockState3.setValue(FenceBlock.SOUTH, Boolean.valueOf(true)), 6, 7, 8, box);
-                this.placeBlock(level, blockState4.setValue(FenceBlock.NORTH, Boolean.valueOf(true)), 7, 7, 6, box);
-                this.placeBlock(level, blockState4.setValue(FenceBlock.SOUTH, Boolean.valueOf(true)), 7, 7, 8, box);
+                this.placeBlock(level, blockState3.setValue(FenceBlock.NORTH, true), 6, 7, 6, box);
+                this.placeBlock(level, blockState3.setValue(FenceBlock.SOUTH, true), 6, 7, 8, box);
+                this.placeBlock(level, blockState4.setValue(FenceBlock.NORTH, true), 7, 7, 6, box);
+                this.placeBlock(level, blockState4.setValue(FenceBlock.SOUTH, true), 7, 7, 8, box);
                 BlockState blockState6 = Blocks.TORCH.defaultBlockState();
                 this.placeBlock(level, blockState6, 5, 8, 7, box);
                 this.placeBlock(level, blockState6, 8, 8, 7, box);
@@ -802,12 +_,12 @@
             this.generateBox(level, box, 4, 1, 9, 6, 1, 11, Blocks.LAVA.defaultBlockState(), Blocks.LAVA.defaultBlockState(), false);
             BlockState blockState = Blocks.IRON_BARS
                 .defaultBlockState()
-                .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true));
+                .setValue(IronBarsBlock.NORTH, true)
+                .setValue(IronBarsBlock.SOUTH, true);
             BlockState blockState1 = Blocks.IRON_BARS
                 .defaultBlockState()
-                .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
-                .setValue(IronBarsBlock.EAST, Boolean.valueOf(true));
+                .setValue(IronBarsBlock.WEST, true)
+                .setValue(IronBarsBlock.EAST, true);
 
             for (int i1 = 3; i1 < 14; i1 += 2) {
                 this.generateBox(level, box, 0, 3, i1, 0, 4, i1, blockState, blockState, false);
@@ -841,18 +_,18 @@
                 flag &= flags[i3];
             }
 
-            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[0])), 4, 3, 8, box);
-            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[1])), 5, 3, 8, box);
-            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[2])), 6, 3, 8, box);
-            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[3])), 4, 3, 12, box);
-            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[4])), 5, 3, 12, box);
-            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[5])), 6, 3, 12, box);
-            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[6])), 3, 3, 9, box);
-            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[7])), 3, 3, 10, box);
-            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[8])), 3, 3, 11, box);
-            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[9])), 7, 3, 9, box);
-            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[10])), 7, 3, 10, box);
-            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, Boolean.valueOf(flags[11])), 7, 3, 11, box);
+            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, flags[0]), 4, 3, 8, box);
+            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, flags[1]), 5, 3, 8, box);
+            this.placeBlock(level, blockState3.setValue(EndPortalFrameBlock.HAS_EYE, flags[2]), 6, 3, 8, box);
+            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, flags[3]), 4, 3, 12, box);
+            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, flags[4]), 5, 3, 12, box);
+            this.placeBlock(level, blockState4.setValue(EndPortalFrameBlock.HAS_EYE, flags[5]), 6, 3, 12, box);
+            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, flags[6]), 3, 3, 9, box);
+            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, flags[7]), 3, 3, 10, box);
+            this.placeBlock(level, blockState5.setValue(EndPortalFrameBlock.HAS_EYE, flags[8]), 3, 3, 11, box);
+            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, flags[9]), 7, 3, 9, box);
+            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, flags[10]), 7, 3, 10, box);
+            this.placeBlock(level, blockState6.setValue(EndPortalFrameBlock.HAS_EYE, flags[11]), 7, 3, 11, box);
             if (flag) {
                 BlockState blockState7 = Blocks.END_PORTAL.defaultBlockState();
                 this.placeBlock(level, blockState7, 4, 3, 9, box);
@@ -934,64 +_,64 @@
                     level,
                     Blocks.IRON_BARS
                         .defaultBlockState()
-                        .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                        .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true)),
-                    4,
-                    i,
-                    4,
-                    box
-                );
-                this.placeBlock(
-                    level,
-                    Blocks.IRON_BARS
-                        .defaultBlockState()
-                        .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                        .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
-                        .setValue(IronBarsBlock.EAST, Boolean.valueOf(true)),
-                    4,
-                    i,
-                    5,
-                    box
-                );
-                this.placeBlock(
-                    level,
-                    Blocks.IRON_BARS
-                        .defaultBlockState()
-                        .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                        .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true)),
-                    4,
-                    i,
-                    6,
-                    box
-                );
-                this.placeBlock(
-                    level,
-                    Blocks.IRON_BARS
-                        .defaultBlockState()
-                        .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
-                        .setValue(IronBarsBlock.EAST, Boolean.valueOf(true)),
-                    5,
-                    i,
-                    5,
-                    box
-                );
-                this.placeBlock(
-                    level,
-                    Blocks.IRON_BARS
-                        .defaultBlockState()
-                        .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
-                        .setValue(IronBarsBlock.EAST, Boolean.valueOf(true)),
-                    6,
-                    i,
-                    5,
-                    box
-                );
-                this.placeBlock(
-                    level,
-                    Blocks.IRON_BARS
-                        .defaultBlockState()
-                        .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
-                        .setValue(IronBarsBlock.EAST, Boolean.valueOf(true)),
+                        .setValue(IronBarsBlock.NORTH, true)
+                        .setValue(IronBarsBlock.SOUTH, true),
+                    4,
+                    i,
+                    4,
+                    box
+                );
+                this.placeBlock(
+                    level,
+                    Blocks.IRON_BARS
+                        .defaultBlockState()
+                        .setValue(IronBarsBlock.NORTH, true)
+                        .setValue(IronBarsBlock.SOUTH, true)
+                        .setValue(IronBarsBlock.EAST, true),
+                    4,
+                    i,
+                    5,
+                    box
+                );
+                this.placeBlock(
+                    level,
+                    Blocks.IRON_BARS
+                        .defaultBlockState()
+                        .setValue(IronBarsBlock.NORTH, true)
+                        .setValue(IronBarsBlock.SOUTH, true),
+                    4,
+                    i,
+                    6,
+                    box
+                );
+                this.placeBlock(
+                    level,
+                    Blocks.IRON_BARS
+                        .defaultBlockState()
+                        .setValue(IronBarsBlock.WEST, true)
+                        .setValue(IronBarsBlock.EAST, true),
+                    5,
+                    i,
+                    5,
+                    box
+                );
+                this.placeBlock(
+                    level,
+                    Blocks.IRON_BARS
+                        .defaultBlockState()
+                        .setValue(IronBarsBlock.WEST, true)
+                        .setValue(IronBarsBlock.EAST, true),
+                    6,
+                    i,
+                    5,
+                    box
+                );
+                this.placeBlock(
+                    level,
+                    Blocks.IRON_BARS
+                        .defaultBlockState()
+                        .setValue(IronBarsBlock.WEST, true)
+                        .setValue(IronBarsBlock.EAST, true),
                     7,
                     i,
                     5,
@@ -1001,7 +_,7 @@
 
             this.placeBlock(
                 level,
-                Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.NORTH, Boolean.valueOf(true)).setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true)),
+                Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.NORTH, true).setValue(IronBarsBlock.SOUTH, true),
                 4,
                 3,
                 2,
@@ -1009,7 +_,7 @@
             );
             this.placeBlock(
                 level,
-                Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.NORTH, Boolean.valueOf(true)).setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true)),
+                Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.NORTH, true).setValue(IronBarsBlock.SOUTH, true),
                 4,
                 3,
                 8,
@@ -1514,14 +_,14 @@
                 case GRATES:
                     this.placeBlock(level, Blocks.CAVE_AIR.defaultBlockState(), x + 1, y, z, box);
                     this.placeBlock(level, Blocks.CAVE_AIR.defaultBlockState(), x + 1, y + 1, z, box);
-                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.WEST, Boolean.valueOf(true)), x, y, z, box);
-                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.WEST, Boolean.valueOf(true)), x, y + 1, z, box);
+                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.WEST, true), x, y, z, box);
+                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.WEST, true), x, y + 1, z, box);
                     this.placeBlock(
                         level,
                         Blocks.IRON_BARS
                             .defaultBlockState()
-                            .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                            .setValue(IronBarsBlock.WEST, Boolean.valueOf(true)),
+                            .setValue(IronBarsBlock.EAST, true)
+                            .setValue(IronBarsBlock.WEST, true),
                         x,
                         y + 2,
                         z,
@@ -1531,8 +_,8 @@
                         level,
                         Blocks.IRON_BARS
                             .defaultBlockState()
-                            .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                            .setValue(IronBarsBlock.WEST, Boolean.valueOf(true)),
+                            .setValue(IronBarsBlock.EAST, true)
+                            .setValue(IronBarsBlock.WEST, true),
                         x + 1,
                         y + 2,
                         z,
@@ -1542,15 +_,15 @@
                         level,
                         Blocks.IRON_BARS
                             .defaultBlockState()
-                            .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                            .setValue(IronBarsBlock.WEST, Boolean.valueOf(true)),
+                            .setValue(IronBarsBlock.EAST, true)
+                            .setValue(IronBarsBlock.WEST, true),
                         x + 2,
                         y + 2,
                         z,
                         box
                     );
-                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.EAST, Boolean.valueOf(true)), x + 2, y + 1, z, box);
-                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.EAST, Boolean.valueOf(true)), x + 2, y, z, box);
+                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.EAST, true), x + 2, y + 1, z, box);
+                    this.placeBlock(level, Blocks.IRON_BARS.defaultBlockState().setValue(IronBarsBlock.EAST, true), x + 2, y, z, box);
                     break;
                 case IRON_DOOR:
                     this.placeBlock(level, Blocks.STONE_BRICKS.defaultBlockState(), x, y, z, box);
