--- a/net/minecraft/world/level/levelgen/carver/CarverDebugSettings.java
+++ b/net/minecraft/world/level/levelgen/carver/CarverDebugSettings.java
@@ -15,7 +_,7 @@
     );
     public static final Codec<CarverDebugSettings> CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
-                Codec.BOOL.optionalFieldOf("debug_mode", Boolean.valueOf(false)).forGetter(CarverDebugSettings::isDebugMode),
+                Codec.BOOL.optionalFieldOf("debug_mode", false).forGetter(CarverDebugSettings::isDebugMode),
                 BlockState.CODEC.optionalFieldOf("air_state", DEFAULT.getAirState()).forGetter(CarverDebugSettings::getAirState),
                 BlockState.CODEC.optionalFieldOf("water_state", DEFAULT.getAirState()).forGetter(CarverDebugSettings::getWaterState),
                 BlockState.CODEC.optionalFieldOf("lava_state", DEFAULT.getAirState()).forGetter(CarverDebugSettings::getLavaState),
