--- a/net/minecraft/world/level/levelgen/carver/WorldCarver.java
+++ b/net/minecraft/world/level/levelgen/carver/WorldCarver.java
@@ -185,7 +_,7 @@
         } else if (state.is(Blocks.WATER)) {
             BlockState waterState = config.debugSettings.getWaterState();
             return waterState.hasProperty(BlockStateProperties.WATERLOGGED)
-                ? waterState.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(true))
+                ? waterState.setValue(BlockStateProperties.WATERLOGGED, true)
                 : waterState;
         } else {
             return state.is(Blocks.LAVA) ? config.debugSettings.getLavaState() : state;
