--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -19,12 +_,6 @@
      */
     Key BRAND_PAPER_ID = Key.key("papermc", "paper");
 
-    // Purpur start
-    /**
-     * The brand id for Purpur.
-     */
-    Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
-    // Purpur end
     /**
      * Gets the {@code ServerBuildInfo}.
      *
