--- a/src/main/java/io/papermc/paper/ServerBuildInfo.java
+++ b/src/main/java/io/papermc/paper/ServerBuildInfo.java
@@ -25,6 +_,14 @@
      */
     Key BRAND_PURPUR_ID = Key.key("purpurmc", "purpur");
     // Purpur end
+
+    // Plazma start
+    /**
+     * The brand id for Plazma.
+     */
+    Key BRAND_PLAZMA_ID = Key.key("plazmamc", "plazma");
+    // Plazma end
+
     /**
      * Gets the {@code ServerBuildInfo}.
      *
