--- a/net/minecraft/world/level/levelgen/feature/stateproviders/DualNoiseProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/DualNoiseProvider.java
@@ -54,7 +_,7 @@
     @Override
     public BlockState getState(RandomSource random, BlockPos pos) {
         double slowNoiseValue = this.getSlowNoiseValue(pos);
-        int i = (int)Mth.clampedMap(slowNoiseValue, -1.0, 1.0, (double)this.variety.minInclusive().intValue(), (double)(this.variety.maxInclusive() + 1));
+        int i = (int)Mth.clampedMap(slowNoiseValue, -1.0, 1.0, (double) this.variety.minInclusive(), (double)(this.variety.maxInclusive() + 1));
         List<BlockState> list = Lists.newArrayListWithCapacity(i);
 
         for (int i1 = 0; i1 < i; i1++) {
