--- a/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
+++ b/net/minecraft/world/level/levelgen/feature/stateproviders/RandomizedIntStateProvider.java
@@ -65,7 +_,7 @@
             this.property = integerProperty;
         }
 
-        return state.setValue(this.property, Integer.valueOf(this.values.sample(random)));
+        return state.setValue(this.property, this.values.sample(random));
     }
 
     @Nullable
