--- a/net/minecraft/world/entity/vehicle/ContainerEntity.java
+++ b/net/minecraft/world/entity/vehicle/ContainerEntity.java
@@ -104,6 +_,8 @@
     }
 
     default void unpackChestVehicleLootTable(@Nullable Player player) {
+        if (this.level().plazmaConfig().entity.populateLootTableOnlyForPlayerInteraction && player == null) return; // Plazma - Populate LootTable only for player interaction
+
         MinecraftServer server = this.level().getServer();
         if (server != null && this.lootableData().shouldReplenish(this, com.destroystokyo.paper.loottable.PaperLootableInventoryData.ENTITY, player)) { // Paper - LootTable API
             LootTable lootTable = server.reloadableRegistries().getLootTable(this.getContainerLootTable());
