--- a/net/minecraft/world/entity/vehicle/MinecartFurnace.java
+++ b/net/minecraft/world/entity/vehicle/MinecartFurnace.java
@@ -144,6 +_,6 @@
 
     @Override
     public BlockState getDefaultDisplayBlockState() {
-        return Blocks.FURNACE.defaultBlockState().setValue(FurnaceBlock.FACING, Direction.NORTH).setValue(FurnaceBlock.LIT, Boolean.valueOf(this.hasFuel()));
+        return Blocks.FURNACE.defaultBlockState().setValue(FurnaceBlock.FACING, Direction.NORTH).setValue(FurnaceBlock.LIT, this.hasFuel());
     }
 }
