--- a/net/minecraft/world/level/chunk/UpgradeData.java
+++ b/net/minecraft/world/level/chunk/UpgradeData.java
@@ -434,7 +_,7 @@
                     for (BlockPos blockPos : set) {
                         BlockState blockState = level.getBlockState(blockPos);
                         if (blockState.getValue(BlockStateProperties.DISTANCE) >= i1) {
-                            level.setBlock(blockPos, blockState.setValue(BlockStateProperties.DISTANCE, Integer.valueOf(i1)), 18);
+                            level.setBlock(blockPos, blockState.setValue(BlockStateProperties.DISTANCE, i1), 18);
                             if (i != 7) {
                                 for (Direction direction : DIRECTIONS) {
                                     mutableBlockPos.setWithOffset(blockPos, direction);
