--- a/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java
@@ -322,10 +_,12 @@
         return this.handle.getUncachedNoiseBiome(biomeX, biomeY, biomeZ);
     }
 
+    /* // Plazma - Remove isClientSide on DedicatedServer
     @Override
     public boolean isClientSide() {
         return this.handle.isClientSide();
     }
+     */ // Plazma - Remove isClientSide on DedicatedServer
 
     @Override
     public int getSeaLevel() {
@@ -812,4 +_,3 @@
     }
     // Paper end
 }
-
