--- a/src/main/java/org/bukkit/craftbukkit/util/DummyGeneratorAccess.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/DummyGeneratorAccess.java
@@ -163,10 +_,12 @@
         throw new UnsupportedOperationException("Not supported yet.");
     }
 
+    /* // Plazma - Remove isClientSide on DedicatedServer
     @Override
     public boolean isClientSide() {
         return false;
     }
+     */ // Plazma - Remove isClientSide on DedicatedServer
 
     @Override
     public int getSeaLevel() {
