--- a/purpur-api/build.gradle.kts
+++ b/purpur-api/build.gradle.kts
@@ -101,21 +_,17 @@
 }
 sourceSets {
     main {
-        java {
-            srcDir(generatedApiPath)
-            srcDir(file("../paper-api/src/main/java"))
-        }
-        resources {
-            srcDir(file("../paper-api/src/main/resources"))
-        }
+        java { srcDir(generatedApiPath) }
+        java { srcDir(file("../paper-api/src/main/java")) }
+        resources { srcDir(file("../paper-api/src/main/resources")) }
+        java { srcDir(file("../purpur-api/src/main/java")) }
+        resources { srcDir(file("../purpur-api/src/main/resources")) }
     }
     test {
-        java {
-            srcDir(file("../paper-api/src/test/java"))
-        }
-        resources {
-            srcDir(file("../paper-api/src/test/resources"))
-        }
+        java { srcDir(file("../paper-api/src/test/java")) }
+        resources { srcDir(file("../paper-api/src/test/resources")) }
+        java { srcDir(file("../purpur-api/src/test/java")) }
+        resources { srcDir(file("../purpur-api/src/test/resources")) }
     }
 }
 
