--- a/purpur-api/build.gradle.kts
+++ b/purpur-api/build.gradle.kts
@@ -101,17 +_,21 @@
         java {
             srcDir(generatedDir)
             srcDir(file("../paper-api/src/main/java"))
+            srcDir(file("../purpur-api/src/main/java"))
         }
         resources {
             srcDir(file("../paper-api/src/main/resources"))
+            srcDir(file("../purpur-api/src/main/resources"))
         }
     }
     test {
         java {
             srcDir(file("../paper-api/src/test/java"))
+            srcDir(file("../purpur-api/src/test/java"))
         }
         resources {
             srcDir(file("../paper-api/src/test/resources"))
+            srcDir(file("../purpur-api/src/test/resources"))
         }
     }
 }
@@ -158,6 +_,13 @@
         pomProps.get().asFile.writeText("version=$projectVersion")
     }
 }
+
+// Plazma Start - SIMD support
+tasks.withType<JavaCompile>().configureEach {
+    val compilerArgs = options.compilerArgs
+    compilerArgs.add("--add-modules=jdk.incubator.vector")
+}
+// Plazma End - SIMD support
 
 tasks.jar {
     from(generateApiVersioningFile.map { it.outputs.files.singleFile }) {
