--- a/ca/spottedleaf/dataconverter/types/json/JsonListType.java
+++ b/ca/spottedleaf/dataconverter/types/json/JsonListType.java
@@ -91,7 +_,7 @@
                 return primitive.getAsBoolean() ? Byte.valueOf((byte)1) : Byte.valueOf((byte)0);
             } else if (this.compressed && primitive.isString()) {
                 try {
-                    return Integer.valueOf(Integer.parseInt(primitive.getAsString()));
+                    return Integer.parseInt(primitive.getAsString());
                 } catch (final NumberFormatException ex) {
                     return null;
                 }
@@ -110,7 +_,7 @@
 
     @Override
     public void setByte(final int index, final byte to) {
-        this.array.set(index, new JsonPrimitive(Byte.valueOf(to)));
+        this.array.set(index, new JsonPrimitive(to));
     }
 
     @Override
@@ -122,7 +_,7 @@
 
     @Override
     public void setShort(final int index, final short to) {
-        this.array.set(index, new JsonPrimitive(Short.valueOf(to)));
+        this.array.set(index, new JsonPrimitive(to));
     }
 
     @Override
@@ -134,7 +_,7 @@
 
     @Override
     public void setInt(final int index, final int to) {
-        this.array.set(index, new JsonPrimitive(Integer.valueOf(to)));
+        this.array.set(index, new JsonPrimitive(to));
     }
 
     @Override
@@ -146,7 +_,7 @@
 
     @Override
     public void setLong(final int index, final long to) {
-        this.array.set(index, new JsonPrimitive(Long.valueOf(to)));
+        this.array.set(index, new JsonPrimitive(to));
     }
 
     @Override
@@ -158,7 +_,7 @@
 
     @Override
     public void setFloat(final int index, final float to) {
-        this.array.set(index, new JsonPrimitive(Float.valueOf(to)));
+        this.array.set(index, new JsonPrimitive(to));
     }
 
     @Override
@@ -170,7 +_,7 @@
 
     @Override
     public void setDouble(final int index, final double to) {
-        this.array.set(index, new JsonPrimitive(Double.valueOf(to)));
+        this.array.set(index, new JsonPrimitive(to));
     }
 
     @Override
@@ -268,7 +_,7 @@
 
     @Override
     public void addByte(final byte b) {
-        this.array.add(Byte.valueOf(b));
+        this.array.add(b);
     }
 
     @Override
@@ -279,7 +_,7 @@
 
     @Override
     public void addShort(final short s) {
-        this.array.add(Short.valueOf(s));
+        this.array.add(s);
     }
 
     @Override
@@ -290,7 +_,7 @@
 
     @Override
     public void addInt(final int i) {
-        this.array.add(Integer.valueOf(i));
+        this.array.add(i);
     }
 
     @Override
@@ -301,7 +_,7 @@
 
     @Override
     public void addLong(final long l) {
-        this.array.add(Long.valueOf(l));
+        this.array.add(l);
     }
 
     @Override
@@ -312,7 +_,7 @@
 
     @Override
     public void addFloat(final float f) {
-        this.array.add(Float.valueOf(f));
+        this.array.add(f);
     }
 
     @Override
@@ -323,7 +_,7 @@
 
     @Override
     public void addDouble(final double d) {
-        this.array.add(Double.valueOf(d));
+        this.array.add(d);
     }
 
     @Override
