--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMerchantCustom.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMerchantCustom.java
@@ -138,10 +_,12 @@
         public void overrideOffers(MerchantOffers offers) {
         }
 
+        /* // Plazma - Remove isClientSide on DedicatedServer
         @Override
         public boolean isClientSide() {
             return false;
         }
+         */ // Plazma - Remove isClientSide on DedicatedServer
 
         @Override
         public boolean stillValid(Player player) {
