--- a/net/minecraft/data/loot/BlockLootSubProvider.java
+++ b/net/minecraft/data/loot/BlockLootSubProvider.java
@@ -423,7 +_,7 @@
                                     age -> SetItemCountFunction.setCount(BinomialDistributionGenerator.binomial(3, (age + 1) / 15.0F))
                                         .when(
                                             LootItemBlockStatePropertyCondition.hasBlockStateProperties(block)
-                                                .setProperties(StatePropertiesPredicate.Builder.properties().hasProperty(StemBlock.AGE, age.intValue()))
+                                                .setProperties(StatePropertiesPredicate.Builder.properties().hasProperty(StemBlock.AGE, age))
                                         )
                                 )
                         )
@@ -657,11 +_,11 @@
                             LootItem.lootTableItem(candleBlock)
                                 .apply(
                                     List.of(2, 3, 4),
-                                    candles -> SetItemCountFunction.setCount(ConstantValue.exactly(candles.intValue()))
+                                    candles -> SetItemCountFunction.setCount(ConstantValue.exactly(candles))
                                         .when(
                                             LootItemBlockStatePropertyCondition.hasBlockStateProperties(candleBlock)
                                                 .setProperties(
-                                                    StatePropertiesPredicate.Builder.properties().hasProperty(CandleBlock.CANDLES, candles.intValue())
+                                                    StatePropertiesPredicate.Builder.properties().hasProperty(CandleBlock.CANDLES, candles)
                                                 )
                                         )
                                 )
@@ -681,11 +_,11 @@
                             LootItem.lootTableItem(petalBlock)
                                 .apply(
                                     IntStream.rangeClosed(1, 4).boxed().toList(),
-                                    amount -> SetItemCountFunction.setCount(ConstantValue.exactly(amount.intValue()))
+                                    amount -> SetItemCountFunction.setCount(ConstantValue.exactly(amount))
                                         .when(
                                             LootItemBlockStatePropertyCondition.hasBlockStateProperties(petalBlock)
                                                 .setProperties(
-                                                    StatePropertiesPredicate.Builder.properties().hasProperty(PinkPetalsBlock.AMOUNT, amount.intValue())
+                                                    StatePropertiesPredicate.Builder.properties().hasProperty(PinkPetalsBlock.AMOUNT, amount)
                                                 )
                                         )
                                 )
