--- a/net/minecraft/world/level/levelgen/feature/rootplacers/RootPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/rootplacers/RootPlacer.java
@@ -75,7 +_,7 @@
     protected BlockState getPotentiallyWaterloggedState(LevelSimulatedReader level, BlockPos pos, BlockState state) {
         if (state.hasProperty(BlockStateProperties.WATERLOGGED)) {
             boolean isFluidAtPosition = level.isFluidAtPosition(pos, fluidState -> fluidState.is(FluidTags.WATER));
-            return state.setValue(BlockStateProperties.WATERLOGGED, Boolean.valueOf(isFluidAtPosition));
+            return state.setValue(BlockStateProperties.WATERLOGGED, isFluidAtPosition);
         } else {
             return state;
         }
