--- a/net/minecraft/util/datafix/fixes/CustomModelDataExpandFix.java
+++ b/net/minecraft/util/datafix/fixes/CustomModelDataExpandFix.java
@@ -20,7 +_,7 @@
             "Custom Model Data expansion",
             type,
             typed -> typed.update(DSL.remainderFinder(), dynamic -> dynamic.update("minecraft:custom_model_data", dynamic1 -> {
-                float f = dynamic1.asNumber(Float.valueOf(0.0F)).floatValue();
+                float f = dynamic1.asNumber(0.0F).floatValue();
                 return dynamic1.createMap(Map.of(dynamic1.createString("floats"), dynamic1.createList(Stream.of(dynamic1.createFloat(f)))));
             }))
         );
