--- a/net/minecraft/world/item/component/DyedItemColor.java
+++ b/net/minecraft/world/item/component/DyedItemColor.java
@@ -22,7 +_,7 @@
     private static final Codec<DyedItemColor> FULL_CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
                 Codec.INT.fieldOf("rgb").forGetter(DyedItemColor::rgb),
-                Codec.BOOL.optionalFieldOf("show_in_tooltip", Boolean.valueOf(true)).forGetter(DyedItemColor::showInTooltip)
+                Codec.BOOL.optionalFieldOf("show_in_tooltip", true).forGetter(DyedItemColor::showInTooltip)
             )
             .apply(instance, DyedItemColor::new)
     );
