--- a/net/minecraft/world/item/component/LodestoneTracker.java
+++ b/net/minecraft/world/item/component/LodestoneTracker.java
@@ -15,7 +_,7 @@
     public static final Codec<LodestoneTracker> CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
                 GlobalPos.CODEC.optionalFieldOf("target").forGetter(LodestoneTracker::target),
-                Codec.BOOL.optionalFieldOf("tracked", Boolean.valueOf(true)).forGetter(LodestoneTracker::tracked)
+                Codec.BOOL.optionalFieldOf("tracked", true).forGetter(LodestoneTracker::tracked)
             )
             .apply(instance, LodestoneTracker::new)
     );
