--- a/net/minecraft/world/item/component/SeededContainerLoot.java
+++ b/net/minecraft/world/item/component/SeededContainerLoot.java
@@ -10,7 +_,7 @@
     public static final Codec<SeededContainerLoot> CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
                 ResourceKey.codec(Registries.LOOT_TABLE).fieldOf("loot_table").forGetter(SeededContainerLoot::lootTable),
-                Codec.LONG.optionalFieldOf("seed", Long.valueOf(0L)).forGetter(SeededContainerLoot::seed)
+                Codec.LONG.optionalFieldOf("seed", 0L).forGetter(SeededContainerLoot::seed)
             )
             .apply(instance, SeededContainerLoot::new)
     );
