--- a/net/minecraft/world/item/component/SuspiciousStewEffects.java
+++ b/net/minecraft/world/item/component/SuspiciousStewEffects.java
@@ -67,7 +_,7 @@
         public static final Codec<SuspiciousStewEffects.Entry> CODEC = RecordCodecBuilder.create(
             instance -> instance.group(
                     MobEffect.CODEC.fieldOf("id").forGetter(SuspiciousStewEffects.Entry::effect),
-                    Codec.INT.lenientOptionalFieldOf("duration", Integer.valueOf(160)).forGetter(SuspiciousStewEffects.Entry::duration)
+                    Codec.INT.lenientOptionalFieldOf("duration", 160).forGetter(SuspiciousStewEffects.Entry::duration)
                 )
                 .apply(instance, SuspiciousStewEffects.Entry::new)
         );
