--- a/net/minecraft/world/item/component/Tool.java
+++ b/net/minecraft/world/item/component/Tool.java
@@ -18,7 +_,7 @@
     public static final Codec<Tool> CODEC = RecordCodecBuilder.create(
         instance -> instance.group(
                 Tool.Rule.CODEC.listOf().fieldOf("rules").forGetter(Tool::rules),
-                Codec.FLOAT.optionalFieldOf("default_mining_speed", Float.valueOf(1.0F)).forGetter(Tool::defaultMiningSpeed),
+                Codec.FLOAT.optionalFieldOf("default_mining_speed", 1.0F).forGetter(Tool::defaultMiningSpeed),
                 ExtraCodecs.NON_NEGATIVE_INT.optionalFieldOf("damage_per_block", 1).forGetter(Tool::damagePerBlock)
             )
             .apply(instance, Tool::new)
