--- a/net/minecraft/world/item/component/Unbreakable.java
+++ b/net/minecraft/world/item/component/Unbreakable.java
@@ -13,7 +_,7 @@
 
 public record Unbreakable(boolean showInTooltip) implements TooltipProvider {
     public static final Codec<Unbreakable> CODEC = RecordCodecBuilder.create(
-        instance -> instance.group(Codec.BOOL.optionalFieldOf("show_in_tooltip", Boolean.valueOf(true)).forGetter(Unbreakable::showInTooltip))
+        instance -> instance.group(Codec.BOOL.optionalFieldOf("show_in_tooltip", true).forGetter(Unbreakable::showInTooltip))
             .apply(instance, Unbreakable::new)
     );
     public static final StreamCodec<ByteBuf, Unbreakable> STREAM_CODEC = ByteBufCodecs.BOOL.map(Unbreakable::new, Unbreakable::showInTooltip);
