--- a/net/minecraft/world/item/component/WrittenBookContent.java
+++ b/net/minecraft/world/item/component/WrittenBookContent.java
@@ -36,7 +_,7 @@
                 Codec.STRING.fieldOf("author").forGetter(WrittenBookContent::author),
                 ExtraCodecs.intRange(0, 3).optionalFieldOf("generation", 0).forGetter(WrittenBookContent::generation),
                 PAGES_CODEC.optionalFieldOf("pages", List.of()).forGetter(WrittenBookContent::pages),
-                Codec.BOOL.optionalFieldOf("resolved", Boolean.valueOf(false)).forGetter(WrittenBookContent::resolved)
+                Codec.BOOL.optionalFieldOf("resolved", false).forGetter(WrittenBookContent::resolved)
             )
             .apply(instance, WrittenBookContent::new)
     );
