--- a/src/main/java/com/destroystokyo/paper/io/SyncLoadFinder.java
+++ b/src/main/java/com/destroystokyo/paper/io/SyncLoadFinder.java
@@ -51,7 +_,7 @@
                 ++valueInMap.times;
 
                 valueInMap.coordinateTimes.compute(ChunkPos.asLong(chunkX, chunkZ), (Long keyInMap1, Integer valueInMap1) -> {
-                    return valueInMap1 == null ? Integer.valueOf(1) : Integer.valueOf(valueInMap1.intValue() + 1);
+                    return valueInMap1 == null ? Integer.valueOf(1) : Integer.valueOf(valueInMap1 + 1);
                 });
 
                 return valueInMap;
