--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawner.java
@@ -133,14 +_,14 @@
     }
 
     public void applyOminous(ServerLevel level, BlockPos pos) {
-        level.setBlock(pos, level.getBlockState(pos).setValue(TrialSpawnerBlock.OMINOUS, Boolean.valueOf(true)), 3);
+        level.setBlock(pos, level.getBlockState(pos).setValue(TrialSpawnerBlock.OMINOUS, true), 3);
         level.levelEvent(3020, pos, 1);
         this.isOminous = true;
         this.data.resetAfterBecomingOminous(this, level);
     }
 
     public void removeOminous(ServerLevel level, BlockPos pos) {
-        level.setBlock(pos, level.getBlockState(pos).setValue(TrialSpawnerBlock.OMINOUS, Boolean.valueOf(false)), 3);
+        level.setBlock(pos, level.getBlockState(pos).setValue(TrialSpawnerBlock.OMINOUS, false), 3);
         this.isOminous = false;
     }
 
