--- a/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
+++ b/net/minecraft/world/level/block/entity/trialspawner/TrialSpawnerData.java
@@ -53,8 +_,8 @@
                     .lenientOptionalFieldOf("registered_players", Sets.newHashSet())
                     .forGetter(trialSpawnerData -> trialSpawnerData.detectedPlayers),
                 UUIDUtil.CODEC_SET.lenientOptionalFieldOf("current_mobs", Sets.newHashSet()).forGetter(trialSpawnerData -> trialSpawnerData.currentMobs),
-                Codec.LONG.lenientOptionalFieldOf("cooldown_ends_at", Long.valueOf(0L)).forGetter(trialSpawnerData -> trialSpawnerData.cooldownEndsAt),
-                Codec.LONG.lenientOptionalFieldOf("next_mob_spawns_at", Long.valueOf(0L)).forGetter(trialSpawnerData -> trialSpawnerData.nextMobSpawnsAt),
+                Codec.LONG.lenientOptionalFieldOf("cooldown_ends_at", 0L).forGetter(trialSpawnerData -> trialSpawnerData.cooldownEndsAt),
+                Codec.LONG.lenientOptionalFieldOf("next_mob_spawns_at", 0L).forGetter(trialSpawnerData -> trialSpawnerData.nextMobSpawnsAt),
                 Codec.intRange(0, Integer.MAX_VALUE)
                     .lenientOptionalFieldOf("total_mobs_spawned", 0)
                     .forGetter(trialSpawnerData -> trialSpawnerData.totalMobsSpawned),
