--- a/src/main/java/io/papermc/paper/configuration/Configuration.java
+++ b/src/main/java/io/papermc/paper/configuration/Configuration.java
@@ -1,11 +_,15 @@
 package io.papermc.paper.configuration;
 
+@Deprecated(forRemoval = true)
 public final class Configuration {
+
+    @Deprecated(forRemoval = true)
     public static final String VERSION_FIELD = "_version";
-    @Deprecated
+
+    @Deprecated(forRemoval = true)
     public static final String LEGACY_CONFIG_VERSION_FIELD = "config-version";
 
-    @Deprecated
+    @Deprecated(forRemoval = true)
     public static final int FINAL_LEGACY_VERSION = 27;
 
     private Configuration() {
