--- a/src/main/java/io/papermc/paper/configuration/mapping/InnerClassFieldDiscoverer.java
+++ b/src/main/java/io/papermc/paper/configuration/mapping/InnerClassFieldDiscoverer.java
@@ -39,11 +_,8 @@
         return null;
     }
 
-    public static FieldDiscoverer<?> worldConfig(WorldConfiguration worldConfiguration) {
-        final Map<Class<?>, Object> overrides = Map.of(
-            WorldConfiguration.class, worldConfiguration
-        );
-        return new InnerClassFieldDiscoverer(overrides);
+    public static <W extends ConfigurationPart> FieldDiscoverer<?> worldConfig(final Class<W> worldConfigClass, final W worldConfiguration) {
+        return new InnerClassFieldDiscoverer(Map.of(worldConfigClass, worldConfiguration));
     }
 
     public static FieldDiscoverer<?> globalConfig() {
