From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alpha <dev@alpha93.kr>
Date: Tue, 27 Feb 2024 18:36:43 +0900
Subject: [PATCH] Rebrand


diff --git a/build.gradle.kts b/build.gradle.kts
index a278ee93d7001e8dc7fd2f5409c2f528c637cb2f..a00cf1659f1fd9dff3ff34561d78732645b51dfb 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -9,7 +9,7 @@ java {
 }
 
 dependencies {
-    implementation(project(":paper-api"))
+    implementation(project(":plazma-api")) // Plazma - Rebrand
     api("com.mojang:brigadier:1.0.18")
 
     compileOnly("it.unimi.dsi:fastutil:8.5.6")
