From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Alpha <dev@alpha93.kr>
Date: Tue, 27 Feb 2024 18:36:43 +0900
Subject: [PATCH] Rebrand


diff --git a/build.gradle.kts b/build.gradle.kts
index e60be45e2513ea49a8605682c3d80835df4c58ee..4ca37af5e5c68c8e8dbd0cdf9cc92285deecf269 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -9,7 +9,7 @@ java {
 }
 
 dependencies {
-    implementation(project(":paper-api"))
+    implementation(project(":${property("projectName").toString().lowercase()}-api")) // Plazma - Rebrand
     api("com.mojang:brigadier:1.0.18")
 
     compileOnly("it.unimi.dsi:fastutil:8.5.6")
