From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Tue, 7 May 2024 23:07:33 +0900
Subject: [PATCH] Add .gitignore file


diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000000000000000000000000000000000000..97e78e27ee0eea2c8b24886eeb19164d552323fe
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,40 @@
+.gradle/
+
+# Eclipse stuff
+/.classpath
+/.project
+/.settings
+/.checkstyle
+/.factorypath
+
+# netbeans
+/nbproject
+/nb-configuration.xml
+
+# we use maven!
+/build.xml
+
+# maven
+/target
+
+# vim
+.*.sw[a-p]
+
+# various other potential build files
+/build
+/bin
+/dist
+/manifest.mf
+
+# Mac filesystem dust
+.DS_Store
+
+# intellij
+*.iml
+*.ipr
+*.iws
+.idea/
+
+# vs code
+/.vscode
+/.factorypath
