--- a/net/minecraft/world/level/levelgen/heightproviders/TrapezoidHeight.java
+++ b/net/minecraft/world/level/levelgen/heightproviders/TrapezoidHeight.java
@@ -15,7 +_,7 @@
         instance -> instance.group(
                 VerticalAnchor.CODEC.fieldOf("min_inclusive").forGetter(provider -> provider.minInclusive),
                 VerticalAnchor.CODEC.fieldOf("max_inclusive").forGetter(provider -> provider.maxInclusive),
-                Codec.INT.optionalFieldOf("plateau", Integer.valueOf(0)).forGetter(provider -> provider.plateau)
+                Codec.INT.optionalFieldOf("plateau", 0).forGetter(provider -> provider.plateau)
             )
             .apply(instance, TrapezoidHeight::new)
     );
