--- a/net/minecraft/world/RandomizableContainer.java
+++ b/net/minecraft/world/RandomizableContainer.java
@@ -88,6 +_,8 @@
     default void unpackLootTable(@Nullable final Player player, final boolean forceClearLootTable) {
         // Paper end - LootTable API
         Level level = this.getLevel();
+        if (level.plazmaConfig().entity.populateLootTableOnlyForPlayerInteraction && player == null) return; // Plazma - Populate LootTable only for player interaction
+
         BlockPos blockPos = this.getBlockPos();
         ResourceKey<LootTable> lootTable = this.getLootTable();
         // Paper start - LootTable API
