--- a/src/main/java/org/spigotmc/AsyncCatcher.java
+++ b/src/main/java/org/spigotmc/AsyncCatcher.java
@@ -4,7 +_,7 @@
 
 public class AsyncCatcher {
 
-    public static boolean enabled = true;
+    public static boolean enabled = !org.plazmamc.plazma.Options.ASYNC; // Plazma - Add option to allow async access
 
     public static void catchOp(String reason) {
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThread()) { // Paper - chunk system
