--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1777,7 +_,7 @@
 
             profilerFiller.pop();
             profilerFiller.pop();
-            serverLevel.explosionDensityCache.clear(); // Paper - Optimize explosions
+            serverLevel.clearExplosionDensityCache(); // Paper - Optimize explosions // Plazma - package-private
         }
         this.isIteratingOverLevels = false; // Paper - Throw exception on world create while being ticked
 
