--- a/net/minecraft/world/level/block/state/properties/BooleanProperty.java
+++ b/net/minecraft/world/level/block/state/properties/BooleanProperty.java
@@ -9,11 +_,11 @@
     private static final int FALSE_INDEX = 1;
 
     // Paper start - optimise blockstate property access
-    private static final Boolean[] BY_ID = new Boolean[]{ Boolean.FALSE, Boolean.TRUE };
+    private static final Boolean[] BY_ID = new Boolean[]{ false, true };
 
     @Override
     public final int moonrise$getIdFor(final Boolean value) {
-        return value.booleanValue() ? 1 : 0;
+        return value ? 1 : 0;
     }
     // Paper end - optimise blockstate property access
 
