--- a/net/minecraft/world/level/block/state/properties/IntegerProperty.java
+++ b/net/minecraft/world/level/block/state/properties/IntegerProperty.java
@@ -13,7 +_,7 @@
     // Paper start - optimise blockstate property access
     @Override
     public final int moonrise$getIdFor(final Integer value) {
-        final int val = value.intValue();
+        final int val = value;
         final int ret = val - this.min;
 
         return ret | ((this.max - ret) >> 31);
@@ -25,7 +_,7 @@
 
         final Integer[] byId = new Integer[max - min + 1];
         for (int i = min; i <= max; ++i) {
-            byId[i - min] = Integer.valueOf(i);
+            byId[i - min] = i;
         }
 
         this.moonrise$setById(byId);
