--- a/net/minecraft/world/level/levelgen/placement/SurfaceRelativeThresholdFilter.java
+++ b/net/minecraft/world/level/levelgen/placement/SurfaceRelativeThresholdFilter.java
@@ -11,8 +_,8 @@
     public static final MapCodec<SurfaceRelativeThresholdFilter> CODEC = RecordCodecBuilder.mapCodec(
         instance -> instance.group(
                 Heightmap.Types.CODEC.fieldOf("heightmap").forGetter(filter -> filter.heightmap),
-                Codec.INT.optionalFieldOf("min_inclusive", Integer.valueOf(Integer.MIN_VALUE)).forGetter(filter -> filter.minInclusive),
-                Codec.INT.optionalFieldOf("max_inclusive", Integer.valueOf(Integer.MAX_VALUE)).forGetter(filter -> filter.maxInclusive)
+                Codec.INT.optionalFieldOf("min_inclusive", Integer.MIN_VALUE).forGetter(filter -> filter.minInclusive),
+                Codec.INT.optionalFieldOf("max_inclusive", Integer.MAX_VALUE).forGetter(filter -> filter.maxInclusive)
             )
             .apply(instance, SurfaceRelativeThresholdFilter::new)
     );
