--- a/src/main/java/org/purpurmc/purpur/command/RamCommand.java
+++ b/src/main/java/org/purpurmc/purpur/command/RamCommand.java
@@ -9,7 +_,7 @@
 import org.purpurmc.purpur.PurpurConfig;
 import org.purpurmc.purpur.task.RamBarTask;
 
-public class RamCommand {
+public interface RamCommand { // Plazma
     public static void register(CommandDispatcher<CommandSourceStack> dispatcher) {
         dispatcher.register(Commands.literal("ram")
                 .requires(listener -> listener.hasPermission(2, "bukkit.command.ram"))
