--- a/net/minecraft/core/dispenser/DispenseItemBehavior.java
+++ b/net/minecraft/core/dispenser/DispenseItemBehavior.java
@@ -455,7 +_,7 @@
                     }
                     // CraftBukkit end
                 } else if (CampfireBlock.canLight(blockState) || CandleBlock.canLight(blockState) || CandleCakeBlock.canLight(blockState)) {
-                    serverLevel.setBlockAndUpdate(blockPos, blockState.setValue(BlockStateProperties.LIT, Boolean.valueOf(true)));
+                    serverLevel.setBlockAndUpdate(blockPos, blockState.setValue(BlockStateProperties.LIT, true));
                     serverLevel.gameEvent(null, GameEvent.BLOCK_CHANGE, blockPos);
                 } else if (blockState.getBlock() instanceof TntBlock && org.bukkit.craftbukkit.event.CraftEventFactory.callTNTPrimeEvent(serverLevel, blockPos, org.bukkit.event.block.TNTPrimeEvent.PrimeCause.DISPENSER, null, blockSource.pos())) { // CraftBukkit - TNTPrimeEvent
                     TntBlock.explode(serverLevel, blockPos);
@@ -614,7 +_,7 @@
                             blockPos,
                             Blocks.WITHER_SKELETON_SKULL
                                 .defaultBlockState()
-                                .setValue(SkullBlock.ROTATION, Integer.valueOf(RotationSegment.convertToSegment(direction))),
+                                .setValue(SkullBlock.ROTATION, RotationSegment.convertToSegment(direction)),
                             3
                         );
                         level.gameEvent(null, GameEvent.BLOCK_PLACE, blockPos);
