--- a/net/minecraft/world/level/levelgen/feature/configurations/SimpleBlockConfiguration.java
+++ b/net/minecraft/world/level/levelgen/feature/configurations/SimpleBlockConfiguration.java
@@ -9,7 +_,7 @@
         instance -> instance.group(
                 BlockStateProvider.CODEC.fieldOf("to_place").forGetter(config -> config.toPlace),
                 Codec.BOOL
-                    .optionalFieldOf("schedule_tick", Boolean.valueOf(false))
+                    .optionalFieldOf("schedule_tick", false)
                     .forGetter(simpleBlockConfiguration -> simpleBlockConfiguration.scheduleTick)
             )
             .apply(instance, SimpleBlockConfiguration::new)
