From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 6 Sep 2023 13:31:43 +0900
Subject: [PATCH] Bump-Dependencies


diff --git a/build.gradle.kts b/build.gradle.kts
index d8b27e3f2fa1ede41125ec27e129b1e1ad42caa7..2c59ed16ea909b6535df3d1d40d83b065c03d041 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -20,7 +20,7 @@ dependencies {
     }
     // Purpur end
     // Paper start
-    implementation("org.jline:jline-terminal-jansi:3.21.0")
+    implementation("org.jline:jline-terminal-jansi:3.23.0") // Plazma - Bump Dependencies
     implementation("net.minecrell:terminalconsoleappender:1.3.0")
     implementation("net.kyori:adventure-text-serializer-ansi:4.14.0") // Keep in sync with adventureVersion from Paper-API build file
     implementation("net.kyori:ansi:1.0.3") // Manually bump beyond above transitive dep
@@ -30,48 +30,48 @@ dependencies {
           all its classes to check if they are plugins.
           Scanning takes about 1-2 seconds so adding this speeds up the server start.
      */
-    implementation("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - implementation
-    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.19.0") // Paper - Needed to generate meta for our Log4j plugins
+    implementation("org.apache.logging.log4j:log4j-core:2.20.0") // Paper - implementation
+    log4jPlugins.annotationProcessorConfigurationName("org.apache.logging.log4j:log4j-core:2.20.0") // Paper - Needed to generate meta for our Log4j plugins
     runtimeOnly(log4jPlugins.output)
     alsoShade(log4jPlugins.output)
-    implementation("io.netty:netty-codec-haproxy:4.1.87.Final") // Paper - Add support for proxy protocol
+    implementation("io.netty:netty-codec-haproxy:4.1.97.Final") // Paper - Add support for proxy protocol
     // Paper end
-    implementation("org.apache.logging.log4j:log4j-iostreams:2.19.0") // Paper - remove exclusion
-    implementation("org.ow2.asm:asm:9.4")
-    implementation("org.ow2.asm:asm-commons:9.4") // Paper - ASM event executor generation
-    testImplementation("org.mockito:mockito-core:4.9.0") // Paper - switch to mockito
+    implementation("org.apache.logging.log4j:log4j-iostreams:2.20.0") // Paper - remove exclusion
+    implementation("org.ow2.asm:asm:9.5")
+    implementation("org.ow2.asm:asm-commons:9.5") // Paper - ASM event executor generation
+    testImplementation("org.mockito:mockito-core:5.5.0") // Paper - switch to mockito
     implementation("org.spongepowered:configurate-yaml:4.1.2") // Paper - config files
     implementation("commons-lang:commons-lang:2.6")
-    implementation("net.fabricmc:mapping-io:0.3.0") // Paper - needed to read mappings for stacktrace deobfuscation
-    runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.0")
+    implementation("net.fabricmc:mapping-io:0.4.2") // Paper - needed to read mappings for stacktrace deobfuscation
+    runtimeOnly("org.xerial:sqlite-jdbc:3.42.0.0")
     runtimeOnly("com.mysql:mysql-connector-j:8.0.33")
     runtimeOnly("com.lmax:disruptor:3.4.4") // Paper
     // Paper start - Use Velocity cipher
-    implementation("com.velocitypowered:velocity-native:3.1.2-SNAPSHOT") {
+    implementation("com.velocitypowered:velocity-native:3.2.0-SNAPSHOT") {
         isTransitive = false
     }
     // Paper end

-    runtimeOnly("org.apache.maven:maven-resolver-provider:3.8.5")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.7.3")
-    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.7.3")
+    runtimeOnly("org.apache.maven:maven-resolver-provider:3.9.4")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-connector-basic:1.9.15")
+    runtimeOnly("org.apache.maven.resolver:maven-resolver-transport-http:1.9.15")

     implementation("org.mozilla:rhino-runtime:1.7.14") // Purpur
     implementation("org.mozilla:rhino-engine:1.7.14") // Purpur
     implementation("dev.omega24:upnp4j:1.0") // Purpur

     // Pufferfish start
-    implementation("org.yaml:snakeyaml:1.32")
+    implementation("org.yaml:snakeyaml:1.32")
     implementation ("com.github.carleslc.Simple-YAML:Simple-Yaml:1.8.4") {
         exclude(group="org.yaml", module="snakeyaml")
     }
     // Pufferfish end

-    testImplementation("io.github.classgraph:classgraph:4.8.47") // Paper - mob goal test
+    testImplementation("io.github.classgraph:classgraph:4.8.162") // Paper - mob goal test
     testImplementation("junit:junit:4.13.2")
-    testImplementation("org.hamcrest:hamcrest-library:1.3")
+    testImplementation("org.hamcrest:hamcrest-library:2.2")

-    implementation("io.netty:netty-all:4.1.87.Final"); // Paper - Bump netty
+    implementation("io.netty:netty-all:4.1.97.Final"); // Paper - Bump netty
 }

 val craftbukkitPackageVersion = "1_20_R1" // Paper
@@ -225,3 +225,6 @@ tasks.registerRunTask("runDev") {
     description = "Spin up a non-relocated Mojang-mapped test server"
     classpath(sourceSets.main.map { it.runtimeClasspath })
 }
+repositories {
+    mavenCentral()
+}
