--- a/net/minecraft/world/entity/ai/village/poi/PoiSection.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiSection.java
@@ -158,7 +_,7 @@
     public record Packed(boolean isValid, List<PoiRecord.Packed> records) {
         public static final Codec<PoiSection.Packed> CODEC = RecordCodecBuilder.create(
             instance -> instance.group(
-                    Codec.BOOL.lenientOptionalFieldOf("Valid", Boolean.valueOf(false)).forGetter(PoiSection.Packed::isValid),
+                    Codec.BOOL.lenientOptionalFieldOf("Valid", false).forGetter(PoiSection.Packed::isValid),
                     PoiRecord.Packed.CODEC.listOf().fieldOf("Records").forGetter(PoiSection.Packed::records)
                 )
                 .apply(instance, PoiSection.Packed::new)
