--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2309,6 +_,11 @@
     // Spigot start
     public class Spigot {
 
+        /**
+         * @deprecated Use {@link #getSpigotConfig()} instead.
+         * @return The spigot configuration
+         */
+        @Deprecated(since = "Plazma 1.21.4")
         @NotNull
         public org.bukkit.configuration.file.YamlConfiguration getConfig() {
             throw new UnsupportedOperationException("Not supported yet.");
@@ -2334,17 +_,27 @@
         }
         // Paper end
 
+        // Plazma start
         // Purpur start
         @NotNull
-        public org.bukkit.configuration.file.YamlConfiguration getPurpurConfig() {
+        public org.bukkit.configuration.file.YamlConfiguration getPurpurConfig()
+        {
             throw new UnsupportedOperationException("Not supported yet.");
         }
 
         @NotNull
-        public java.util.Properties getServerProperties() {
+        public java.util.Properties getServerProperties()
+        {
             throw new UnsupportedOperationException("Not supported yet.");
         }
         // Purpur end
+
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getPlazmaConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Plazma end
 
         /**
          * Sends the component to the player
