--- a/net/minecraft/util/FormattedCharSequence.java
+++ b/net/minecraft/util/FormattedCharSequence.java
@@ -32,7 +_,7 @@
     }
 
     static FormattedCharSink decorateOutput(FormattedCharSink sink, Int2IntFunction codePointMapper) {
-        return (index, style, codePoint) -> sink.accept(index, style, codePointMapper.apply(Integer.valueOf(codePoint)));
+        return (index, style, codePoint) -> sink.accept(index, style, codePointMapper.apply(codePoint));
     }
 
     static FormattedCharSequence composite() {
