From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 22 Dec 2022 20:20:03 +0900
Subject: [PATCH] Plazma Configurations


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index a15d0ed710ff261f203d7e355e7d532f2e68abc9..a1d870c6523b0b382d01ca27501a4c091eb795af 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -2266,6 +2266,13 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
         }
         // Purpur end
 
+        // Plazma start - Plazma Configuration API
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getPlazmaConfig() {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Plazma end
+
         /**
          * Sends the component to the player
          *
