From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 19 Jun 2023 18:19:50 +0900
Subject: [PATCH] Publish-Packages


diff --git a/build.gradle.kts b/build.gradle.kts
index f22c0403729b3554ff769b06f7455516b15fd974..73d5655b7217382b7d3a73a24c61a97b40230e7e 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -178,3 +178,22 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+// Plazma start
+publishing {
+    repositories {
+        maven {
+            name = "githubPackage"
+            url = uri("https://maven.pkg.github.com/PlazmaMC/Plazma")
+
+            credentials.username = System.getenv("GITHUB_USERNAME")
+            credentials.password = System.getenv("GITHUB_TOKEN")
+        }
+
+        publications {
+            register<MavenPublication>("gpr") {
+                from(components["java"])
+            }
+        }
+    }
+}
+// Plazma end
\ No newline at end of file
