From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Thu, 22 Dec 2022 20:29:45 +0900
Subject: [PATCH] Bump Bungeecord Chat API to 1.19-R0.1-SNAPSHOT


diff --git a/build.gradle.kts b/build.gradle.kts
index 5c8dd4d3313a791d1fee00ec5d4bc595b76b7d6d..a85f1db096555a6077dd77cbd7572442d3adc848 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -25,7 +25,7 @@ dependencies {
     // api dependencies are listed transitively to API consumers
     api("com.google.guava:guava:31.1-jre")
     api("com.google.code.gson:gson:2.10")
-    api("net.md-5:bungeecord-chat:1.16-R0.4-deprecated+build.6") // Paper
+    api("net.md-5:bungeecord-chat:1.19-R0.1-SNAPSHOT") // Paper // Plazma
     api("org.yaml:snakeyaml:1.33")
     // Paper start
     api("com.googlecode.json-simple:json-simple:1.1.1") {
