From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 19 Jun 2023 18:19:50 +0900
Subject: [PATCH] Publish-Packages


diff --git a/build.gradle.kts b/build.gradle.kts
index 1261dbb7a1c22860922afd0ac8c341aea3569df3..299ce875ee51294d87b31571b624a7991e0fdb1c 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -163,3 +163,22 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+// Plazma start
+publishing {
+    repositories {
+        maven {
+            name = "githubPackage"
+            url = uri("https://maven.pkg.github.com/PlazmaMC/Plazma")
+
+            credentials.username = System.getenv("GITHUB_USERNAME")
+            credentials.password = System.getenv("GITHUB_TOKEN")
+        }
+
+        publications {
+            register<MavenPublication>("gpr") {
+                from(components["java"])
+            }
+        }
+    }
+}
+// Plazma end
\ No newline at end of file
