From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Mon, 19 Jun 2023 18:19:50 +0900
Subject: [PATCH] Publish-Packages


diff --git a/build.gradle.kts b/build.gradle.kts
index 6c0df825238ca037abeb2ba619983b6f554180ea..54da0cab8681daac2dd9fbf0eb94de471819d4c8 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -171,3 +171,22 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+// Plazma start
+publishing {
+    repositories {
+        maven {
+            name = "githubPackage"
+            url = uri("https://maven.pkg.github.com/PlazmaMC/Plazma")
+
+            credentials.username = System.getenv("GITHUB_USERNAME")
+            credentials.password = System.getenv("GITHUB_TOKEN")
+        }
+
+        publications {
+            register<MavenPublication>("gpr") {
+                from(components["java"])
+            }
+        }
+    }
+}
+// Plazma end
\ No newline at end of file
