From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <alphakr93@outlook.com>
Date: Fri, 6 Jan 2023 17:11:31 +0900
Subject: [PATCH] Publish Packages


diff --git a/build.gradle.kts b/build.gradle.kts
index 5c8dd4d3313a791d1fee00ec5d4bc595b76b7d6d..47343ec14550f85ed542c15f57988798e7c3d79a 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -153,3 +153,23 @@ tasks.check {
     dependsOn(scanJar)
 }
 // Paper end
+
+// Plazma start
+publishing {
+    repositories {
+        maven {
+            name = "githubPackage"
+            url = uri("https://maven.pkg.github.com/PlazmaMC/Plazma")
+
+            credentials.username = System.getenv("GITHUB_USERNAME")
+            credentials.password = System.getenv("GITHUB_TOKEN")
+        }
+
+        publications {
+            register<MavenPublication>("gpr") {
+                from(components["java"])
+            }
+        }
+    }
+}
+// Plazma end
