From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Thu, 28 Sep 2023 12:21:36 +0900
Subject: [PATCH] Implement No Chat Reports


diff --git a/src/main/java/org/bukkit/event/player/PlayerKickEvent.java b/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
index b8bf61bea73086c61dce6230686e627dc324ebe0..4b91f99f2ca9c68378ea23cbd78908c6f8d46bc7 100644
--- a/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerKickEvent.java
@@ -218,6 +218,8 @@ public class PlayerKickEvent extends PlayerEvent implements Cancellable {
          * Fallback cause
          */
         UNKNOWN,
+
+        NCR_NOT_INSTALLED, // Plazma - Implement No Chat Reports
     }
     // Paper end
 }
