From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Mon, 6 May 2024 20:43:37 +0900
Subject: [PATCH] Various API Tweaks


diff --git a/src/main/java/com/destroystokyo/paper/MaterialTags.java b/src/main/java/com/destroystokyo/paper/MaterialTags.java
index 679f78e07a3a2de745fa237165d0a8db5e086f29..ac24a288fda957546e2832f375bd1c958db8852b 100644
--- a/src/main/java/com/destroystokyo/paper/MaterialTags.java
+++ b/src/main/java/com/destroystokyo/paper/MaterialTags.java
@@ -591,6 +591,15 @@ public class MaterialTags {
         .add(material -> material.name().startsWith("DEEPSLATE_") && material.name().endsWith("_ORE"))
         .ensureSize("DEEPSLATE_ORES", 8).lock();
 
+    // Plazma start - Various API Tweaks
+    /**
+     * Covers the variants of stone ores.
+     */
+    public static final MaterialSetTag STONE_ORES = new MaterialSetTag(keyFor("stone_ores"))
+        .add(material -> !material.name().startsWith("DEEPSLATE_") && !material.name().startsWith("NETHER_") && material.name().endsWith("_ORE"))
+        .ensureSize("STONE_ORES", 8).lock();
+    // Plazma end - Various API Tweaks
+
     /**
      * Covers the variants of raw ore blocks.
      */
