From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 15 Dec 2024 13:34:45 +0900
Subject: [PATCH] Temporary javadoc build fix


diff --git a/build.gradle.kts b/build.gradle.kts
index 6963a76b9040c01dc60d0f24c9eef4738e93f753..1f3d951a69cd6a9ea5fa7ddbe60aaa5e8d9f0974 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -165,7 +165,7 @@ tasks.withType<Javadoc> {
         "https://jd.advntr.dev/text-logger-slf4j/${common.adventure.api.orNull?.version}/",
         "https://javadoc.io/doc/org.yaml/snakeyaml/${common.snakeyaml.orNull?.version}/",
         "https://javadoc.io/doc/org.slf4j/slf4j-api/${api.slf4j.orNull?.version}/",
-        "https://javadoc.io/doc/org.apache.logging.log4j/log4j-api/${common.log4j.api.orNull?.version}/",
+        "https://javadoc.io/doc/org.apache.logging.log4j/log4j-api/2.20.0/", // Plazma - TODO: log4j 2.24.2 javadocs not found
         "https://javadoc.io/doc/org.apache.maven.resolver/maven-resolver-api/${common.mvn.connector.orNull?.version}",
         // Plazma end
     )
