From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 5 Nov 2023 10:47:05 +0900
Subject: [PATCH] Setup basic configuration sections


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index d8006ae086c18a4ef43906d516eba0d5fd397f5f..aff5c68bc715c6ea9b5b808e5a94cbac9a3725ba 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -3,6 +3,7 @@ package org.plazmamc.plazma.configurations;
 import io.papermc.paper.configuration.Configuration;
 import io.papermc.paper.configuration.ConfigurationPart;
 import org.jetbrains.annotations.NotNull;
+import org.spongepowered.configurate.objectmapping.meta.PostProcess;
 import org.spongepowered.configurate.objectmapping.meta.Setting;
 
 @SuppressWarnings({"CanBeFinal", "FieldCanBeLocal", "FieldMayBeFinal", "InnerClassMayBeStatic"})
@@ -23,4 +24,49 @@ public class GlobalConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+
+    }
+
+    @Setting("world-generation")
+    public WorldGeneration worldgen;
+    public class WorldGeneration extends ConfigurationPart {
+
+
+    }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+
+        @PostProcess
+        public void post() {
+        }
+
+    }
+
+    public World world;
+    public class World extends ConfigurationPart {
+
+
+        @PostProcess
+        public void post() {
+        }
+
+    }
+
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+
+    }
+
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
 }
diff --git a/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java
index 25c0f5d28107b45677aa7b19bc0d5238512d9826..ac0f038de0ce5cf6df0b730af69d3229c3119eff 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java
@@ -2,9 +2,12 @@ package org.plazmamc.plazma.configurations;
 
 import org.spongepowered.configurate.NodePath;
 
+import static org.spongepowered.configurate.NodePath.path;
+
 interface RemovedConfigurations {
 
     NodePath[] WORLD_PATHS = {
+        path("structure", "nether-portal")
     };
 
     NodePath[] GLOBAL_PATHS = {
diff --git a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
index f2d3d51cb4b8fc7a5fd6db1a63289fff6d32a1ea..8dce68cf7769fcd5ea03be32621ccb6bab174697 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
@@ -19,4 +19,38 @@ public class WorldConfigurations extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+        public Phantom phantom;
+        public class Phantom extends ConfigurationPart {
+
+
+        }
+
+    }
+
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+        public NetherPortal netherPortal;
+        public class NetherPortal extends ConfigurationPart {
+
+
+        }
+
+    }
+
+    public Block block;
+    public class Block extends ConfigurationPart {
+        
+
+    }
+
 }
