From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 5 Nov 2023 10:47:05 +0900
Subject: [PATCH] Setup basic configuration sections


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 69c4d9cb0532621018f6cd99916c409fc150ab7e..5875aae8d7d0146ef9903d8b4bc2b673a615fd98 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -23,4 +23,23 @@ public class GlobalConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+
+    }
+
+    @Setting("world-generation")
+    public WorldGeneration worldgen;
+    public class WorldGeneration extends ConfigurationPart {
+
+
+    }
+
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
 }
diff --git a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
index 842fb520a2d638aaa5bd0a7198190dbe3ecbe14c..a372b5be3b1da5868d3766a8ba58997a48118581 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
@@ -19,4 +19,32 @@ public class WorldConfigurations extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+        public Phantom phantom;
+        public class Phantom extends ConfigurationPart {
+
+
+        }
+
+    }
+
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+        public NetherPortal netherPortal;
+        public class NetherPortal extends ConfigurationPart {
+
+
+        }
+
+    }
+
 }
