From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 5 Nov 2023 10:47:05 +0900
Subject: [PATCH] Setup basic configuration sections


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 261734da470664d55d831c801e1e297fa057a7f4..7d0d595a87308316a3ae57c0fafc3faa7d96aeea 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -26,4 +26,23 @@ public class GlobalConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+
+    }
+
+    @Setting("world-generation")
+    public WorldGeneration worldgen;
+    public class WorldGeneration extends ConfigurationPart {
+
+
+    }
+
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
 }
diff --git a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
index 9e76fd2548a57bbc7c884a7a152a88ca61d01285..5e706773d2f64c81a5d3bbf41b8217a343a5d77e 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
@@ -18,4 +18,32 @@ public class WorldConfigurations extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+        public Phantom phantom;
+        public class Phantom extends ConfigurationPart {
+
+
+        }
+
+    }
+
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+        public NetherPortal netherPortal;
+        public class NetherPortal extends ConfigurationPart {
+
+
+        }
+
+    }
+
 }
