From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 5 Nov 2023 10:47:05 +0900
Subject: [PATCH] Setup basic configuration sections


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 37dfa58c0fa2b3c3a419cfdb96c4d1c84bdbde25..127e2aaed186b88fae42d25c91aba6af5e9ea636 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -22,4 +22,23 @@ public class GlobalConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+
+    }
+
+    @Setting("world-generation")
+    public WorldGeneration worldgen;
+    public class WorldGeneration extends ConfigurationPart {
+
+
+    }
+
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
 }
diff --git a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
index fcdd21f96240baa1dc19ea1279c1dd02fc64c468..f189af938e8c083197d38e8abe7313f3bc2badbe 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
@@ -18,4 +18,32 @@ public class WorldConfigurations extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+        public Phantom phantom;
+        public class Phantom extends ConfigurationPart {
+
+
+        }
+
+    }
+
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+        public NetherPortal netherPortal;
+        public class NetherPortal extends ConfigurationPart {
+
+
+        }
+
+    }
+
 }
