From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Thu, 23 Mar 2023 14:25:09 +0900
Subject: [PATCH] ChunkSending-Configuration


diff --git a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
index d1b5b269173c43452960bbe38d5d940f2089c0fa..06795ed5990b1b6e8aa3b40648daffec7bcc4ab5 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
@@ -3,9 +3,13 @@ package org.plazmamc.plazma.configurations;
 import io.papermc.paper.configuration.Configuration;
 import io.papermc.paper.configuration.ConfigurationPart;
 import io.papermc.paper.configuration.PaperConfigurations;
+import io.papermc.paper.configuration.WorldConfiguration;
+import io.papermc.paper.configuration.type.EngineMode;
 import net.minecraft.resources.ResourceLocation;
 import org.spongepowered.configurate.objectmapping.meta.Setting;
 
+import java.util.List;
+
 @SuppressWarnings({"FieldCanBeLocal", "FieldMayBeFinal", "NotNullFieldNotInitialized", "InnerClassMayBeStatic"})
 public class LevelConfigurations extends ConfigurationPart {
     public static final int CURRENT_VERSION = 1;
@@ -21,4 +25,11 @@ public class LevelConfigurations extends ConfigurationPart {
 
     @Setting(Configuration.VERSION_FIELD)
     public int version = CURRENT_VERSION;
+
+    public ChunkSending chunksending;
+
+    public class ChunkSending extends ConfigurationPart {
+        public boolean enable = true;
+        public int maxChunksPerTick = 5;
+    }
 }
