From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Tue, 7 May 2024 22:42:25 +0900
Subject: [PATCH] Development Build


diff --git a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
index 0a846802f053c545992763e0447e15565204e764..31edd38bbcd117dcde867e4cf7d25f657ea16468 100644
--- a/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
+++ b/src/main/java/com/destroystokyo/paper/PaperVersionFetcher.java
@@ -27,7 +27,7 @@ public class PaperVersionFetcher implements VersionFetcher {
     private static final String PROJECT_NAME = "Plazma";
     private static final String PROJECT_REPO = "PlazmaMC/PlazmaBukkit";
     private static final String DOWNLOAD_PAGE = "https://docs.plazmamc.org/plazma/about/downloads";
-    public static final boolean DEVELOPMENT = false;
+    public static final boolean DEVELOPMENT = true;
     // Plazma end
 
     @Override
