From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Tue, 7 May 2024 22:42:25 +0900
Subject: [PATCH] Development Build


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index d89fa400c036de89b9027faaee6b7ca7616e9365..17e9ed0ed774ebd156a7159943312d44af85dd6e 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -18,7 +18,7 @@ public final class Versioning {
     public static final String REPOSITORY = "PlazmaMC/PlazmaBukkit";
     public static final String DOCUMENT = "https://docs.plazmamc.org/plazma/administration/getting-started";
     public static final String STEP_UP = "https://docs.plazmamc.org/plazma/administration/getting-started/next-step";
-    public static final boolean DEVELOPMENT = false;
+    public static final boolean DEVELOPMENT = true; // Plazma - Development build
     static {
         InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/org.plazmamc.plazma/plazma-api/pom.properties");
         Properties properties = new Properties();
