From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 5 Nov 2023 10:47:05 +0900
Subject: [PATCH] Setup basic configuration sections


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index b638690d5cd2360444e6c525f97f634d8e1dc60e..e89534b6e2b91483d7eddfcb3441e9addf9bccc7 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -6,6 +6,7 @@ import it.unimi.dsi.fastutil.objects.Reference2ObjectMap;
 import it.unimi.dsi.fastutil.objects.Reference2ObjectOpenHashMap;
 import net.minecraft.Util;
 import org.jetbrains.annotations.NotNull;
+import org.spongepowered.configurate.objectmapping.meta.PostProcess;
 import org.spongepowered.configurate.objectmapping.meta.Setting;
 import java.util.List;
 
@@ -27,4 +28,61 @@ public class GlobalConfiguration extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Messages messages;
+    public class Messages extends ConfigurationPart {
+
+
+    }
+
+    public Command command;
+    public class Command extends ConfigurationPart {
+
+
+    }
+
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+
+    }
+
+    @Setting("world-generation")
+    public WorldGeneration worldgen;
+    public class WorldGeneration extends ConfigurationPart {
+
+
+    }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+
+        @PostProcess
+        public void post() {
+        }
+
+    }
+
+    public World world;
+    public class World extends ConfigurationPart {
+
+
+        @PostProcess
+        public void post() {
+        }
+
+    }
+
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+
+    }
+
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
 }
diff --git a/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java
index 25c0f5d28107b45677aa7b19bc0d5238512d9826..ac0f038de0ce5cf6df0b730af69d3229c3119eff 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/RemovedConfigurations.java
@@ -2,9 +2,12 @@ package org.plazmamc.plazma.configurations;
 
 import org.spongepowered.configurate.NodePath;
 
+import static org.spongepowered.configurate.NodePath.path;
+
 interface RemovedConfigurations {
 
     NodePath[] WORLD_PATHS = {
+        path("structure", "nether-portal")
     };
 
     NodePath[] GLOBAL_PATHS = {
diff --git a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
index f2d3d51cb4b8fc7a5fd6db1a63289fff6d32a1ea..c222d58520b655b54b8975ec226a4ba948691cd4 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
@@ -19,4 +19,44 @@ public class WorldConfigurations extends ConfigurationPart {
     @Setting(Configuration.VERSION_FIELD)
     int version = VERSION;
 
+    public Miscellaneous misc;
+    public class Miscellaneous extends ConfigurationPart {
+
+
+    }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+        public Phantom phantom;
+        public class Phantom extends ConfigurationPart {
+
+
+        }
+
+    }
+
+    public Structure structure;
+    public class Structure extends ConfigurationPart {
+
+        public NetherPortal netherPortal;
+        public class NetherPortal extends ConfigurationPart {
+
+
+        }
+
+    }
+
+    public Block block;
+    public class Block extends ConfigurationPart {
+        
+
+    }
+
+    public Item item;
+    public class Item extends ConfigurationPart {
+
+
+    }
+
 }
