From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Thu, 23 Mar 2023 14:25:09 +0900
Subject: [PATCH] Implement ChunkSending - Configuration


diff --git a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
index d1b5b269173c43452960bbe38d5d940f2089c0fa..1f0b3cc1bad949147f632e6ace40ac61235acbc1 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
@@ -21,4 +21,12 @@ public class LevelConfigurations extends ConfigurationPart {
 
     @Setting(Configuration.VERSION_FIELD)
     public int version = CURRENT_VERSION;
+
+    public ChunkSending chunkSending;
+    public class ChunkSending extends ConfigurationPart {
+
+        public boolean enabled = true;
+        public int maxChunksPerTick = 5;
+
+    }
 }
