From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Tue, 7 May 2024 22:42:25 +0900
Subject: [PATCH] Development Build


diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index 317ac7c5f1c2d542be260f7e86f9a7281506ebe0..2612df0680800e8c4644599b1885230778c812b7 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -18,7 +18,7 @@ public final class Versioning {
     public static final String DOWNLOAD = "https://plazmamc.org/downloads";
     public static final String DOCUMENT = "https://docs.plazmamc.org/plazma/administration/getting-started";
     public static final String STEP_UP = "https://docs.plazmamc.org/plazma/administration/getting-started/next-step";
-    public static final boolean DEVELOPMENT = false;
+    public static final boolean DEVELOPMENT = true; // Plazma - Development build
     static {
         InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/org.plazmamc.plazma/plazma-api/pom.properties");
         Properties properties = new Properties();
