From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Thu, 23 Mar 2023 21:18:47 +0900
Subject: [PATCH] Player configuration section


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 5dcc234f189dfc86b1585c1eaa6335949d3b53d6..1d6c9e45ddb8711ed253bb8daa1de1f72e27dc89 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -27,5 +27,12 @@ public class GlobalConfiguration extends ConfigurationPart {
         public boolean enableRootUserWarnings = true;
         public boolean enableSecureChatMarker = Boolean.getBoolean("Plazma.disableConfigOptimization");
 
+    }
+
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+
+
     }
 }
