From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Thu, 23 Mar 2023 21:18:47 +0900
Subject: [PATCH] Player configuration section


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index c408e9ccee77022c5d2172252e62c5c75c853610..6c85924291a09252031ec404c7a8d2f12e175d0f 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -27,5 +27,12 @@ public class GlobalConfiguration extends ConfigurationPart {
         public boolean enableRootUserWarnings = true;
         public boolean enableSecureChatMarker = Boolean.getBoolean("plazma.disableAutoOptimizeConfigurations");
 
+    }
+
+    public Player player;
+    public class Player extends ConfigurationPart {
+
+
+
     }
 }
