From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Wed, 27 Sep 2023 22:11:15 +0900
Subject: [PATCH] Apply various optimizations

[REFERENCE PATCHES]
Akarin - Swaps the predicate order of collision

diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
index 067302709a2291274eeecd1e4e04662b6914da12..aafe8924d1293af0bbc3f264b83e1fd618dbb7ad 100644
--- a/src/main/java/net/minecraft/world/entity/Entity.java
+++ b/src/main/java/net/minecraft/world/entity/Entity.java
@@ -2106,8 +2106,9 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
     public void playerTouch(Player player) {}
 
     public void push(Entity entity) {
+        if (entity.noPhysics || this.noPhysics) return; // Plazma
         if (!this.isPassengerOfSameVehicle(entity)) {
-            if (!entity.noPhysics && !this.noPhysics) {
+            //if (!entity.noPhysics && !this.noPhysics) { // Plazma - moved up
                 if (this.level.paperConfig().collisions.onlyPlayersCollide && !(entity instanceof ServerPlayer || this instanceof ServerPlayer)) return; // Paper
                 double d0 = entity.getX() - this.getX();
                 double d1 = entity.getZ() - this.getZ();
@@ -2136,7 +2137,7 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource {
                     }
                 }
 
-            }
+            //} // Plazma
         }
     }
 
