From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sat, 25 Mar 2023 20:38:21 +0900
Subject: [PATCH] Entity Configuration


diff --git a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
index 14987ceec693397602b5c8e1fe460a6fc212a0ee..5dcabaf3fa198a05afc25753f6d062fe7d86972e 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
@@ -90,4 +90,21 @@ public class LevelConfigurations extends ConfigurationPart {
         public boolean reduceCreateRandomInstance = DO_OPTIMIZE;
 
     }
+
+    public Entity entity;
+    public class Entity extends ConfigurationPart {
+
+        public Monster monster;
+        public class Monster extends ConfigurationPart {
+
+            public Phantom phantom;
+            public class Phantom extends ConfigurationPart {
+
+
+
+            }
+
+        }
+
+    }
 }
