From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Mon, 6 Nov 2023 11:36:08 +0900
Subject: [PATCH] Configurable entity sensor tick


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 2f5250d3baba15e914290ce30f813ad63f6bd943..8b9d790351cd416e5d6051e9ad27d66a7097f558 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -997,10 +997,11 @@ public abstract class Mob extends LivingEntity implements EquipmentUser, Targeti
         //ProfilerFiller gameprofilerfiller = this.level().getProfiler(); // Purpur
 
         //gameprofilerfiller.push("sensing"); // Purpur
-        this.sensing.tick();
+        //this.sensing.tick(); // Plazma - Moved down
         //gameprofilerfiller.pop(); // Purpur
         int i = this.tickCount + this.getId();
 
+        if (i % this.level().plazmaConfig().entity.sensorTick == 0) this.sensing.tick(); // Plazma - Configurable entity sensor tick
         if (i % 2 != 0 && this.tickCount > 1) {
             //gameprofilerfiller.push("targetSelector"); // Purpur
             if (this.targetSelector.inactiveTick(this.activatedPriority, false)) // Pufferfish - use this to alternate ticking
diff --git a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
index 9debf554deb5b7aa021502ff7d600905f56fa25c..dceb5b2cec5f0c20be74d892e8cab98823d2f59a 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
@@ -30,6 +30,7 @@ public class WorldConfigurations extends ConfigurationPart {
     public class Entity extends ConfigurationPart {
 
         public boolean ignoreUselessPackets = OPTIMIZE;
+        public int sensorTick = 1;
 
         public Phantom phantom;
         public class Phantom extends ConfigurationPart {
