From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Wed, 27 Sep 2023 22:35:19 +0900
Subject: [PATCH] Ignore useless entity packets


diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index c6ef510d335b8baea58c4491853414a52a06b66b..ccc2f5f71181c1d5c36dcb85de3e48982db0aa8f 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -200,6 +200,8 @@ public class ServerEntity {
                         flag4 = true;
                         flag5 = true;
                     }
+
+                    if (this.entity.level().plazmaConfig().entity.ignoreUselessPackets && isUselessPacket(packet1)) packet1 = null; // Plazma
                 }
 
                 if ((this.trackDelta || this.entity.hasImpulse || this.entity instanceof LivingEntity && ((LivingEntity) this.entity).isFallFlying()) && this.tickCount > 0) {
@@ -272,6 +274,19 @@ public class ServerEntity {
         });
     }
 
+    // Plazma start
+    private boolean isUselessPacket(@Nullable Packet<?> packet) {
+        if (!(packet instanceof ClientboundMoveEntityPacket p)) return false;
+        if (p instanceof ClientboundMoveEntityPacket.Pos)
+            return p.getXa() == 0 && p.getYa() == 0 && p.getZa() == 0;
+        if (p instanceof ClientboundMoveEntityPacket.Rot)
+            return p.getxRot() == 0 && p.getyRot() == 0;
+        if (p instanceof ClientboundMoveEntityPacket.PosRot)
+            return p.getXa() == 0 && p.getYa() == 0 && p.getZa() == 0 && p.getxRot() == 0 && p.getyRot() == 0;
+        return false;
+    }
+    // Plazma end
+
     public void removePairing(ServerPlayer player) {
         this.entity.stopSeenByPlayer(player);
         player.connection.send(new ClientboundRemoveEntitiesPacket(new int[]{this.entity.getId()}));
diff --git a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
index 50c78bb76c130e228f9160f09c09b0fb6877a040..fa578a0e72eff9721c7e3b2f21e3043b9b39c35a 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/WorldConfigurations.java
@@ -28,6 +28,8 @@ public class WorldConfigurations extends ConfigurationPart {
     public Entity entity;
     public class Entity extends ConfigurationPart {
 
+        public boolean ignoreUselessPackets;
+
         public Phantom phantom;
         public class Phantom extends ConfigurationPart {
 
