From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sat, 25 Mar 2023 22:07:39 +0900
Subject: [PATCH] Do not send useless entity packets


diff --git a/src/main/java/net/minecraft/server/level/ServerEntity.java b/src/main/java/net/minecraft/server/level/ServerEntity.java
index fb11b020d52988360562db23d8568e5ef37e21c8..80ab684a33876d62186dcf86f555e46db1594b09 100644
--- a/src/main/java/net/minecraft/server/level/ServerEntity.java
+++ b/src/main/java/net/minecraft/server/level/ServerEntity.java
@@ -203,6 +203,7 @@ public class ServerEntity {
                         flag4 = true;
                         flag5 = true;
                     }
+                    if (org.plazmamc.plazma.configurations.GlobalConfiguration.get().misc.doNotSendUselessEntityPackets && isUselessEntityPacket(packet1)) packet1 = null; // Plazma
                 }
 
                 if ((this.trackDelta || this.entity.hasImpulse || this.entity instanceof LivingEntity && ((LivingEntity) this.entity).isFallFlying()) && this.tickCount > 0) {
@@ -275,6 +276,21 @@ public class ServerEntity {
         });
     }
 
+    // Plazma start
+    private boolean isUselessEntityPacket(@Nullable Packet<?> packet) {
+        if (packet == null) return false;
+        if (packet instanceof ClientboundMoveEntityPacket p) {
+           if (p instanceof ClientboundMoveEntityPacket.Pos)
+                return p.getXa() == 0 && p.getYa() == 0 && p.getZa() == 0;
+            else if (p instanceof ClientboundMoveEntityPacket.Rot)
+                return p.getxRot() == 0 && p.getyRot() == 0;
+            else if (p instanceof ClientboundMoveEntityPacket.PosRot)
+                return p.getXa() == 0 && p.getYa() == 0 && p.getZa() == 0 && p.getxRot() == 0 && p.getyRot() == 0;
+        }
+        return false;
+    }
+    // Plazma end
+
     public void removePairing(ServerPlayer player) {
         this.entity.stopSeenByPlayer(player);
         player.connection.send(new ClientboundRemoveEntitiesPacket(new int[]{this.entity.getId()}));
diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index e91b404fca2823748c2ed343cbd83301e55b2f0a..d21be63b98f026fa3ac7dcdc73f9f488a7000643 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -42,6 +42,7 @@ public class GlobalConfiguration extends ConfigurationPart {
 
         public boolean reduceCreateRandomInstance = DO_OPTIMIZE;
         public boolean doNotTriggerLootTableRefreshForNonPlayerInteraction = DO_OPTIMIZE;
+        public boolean doNotSendUselessEntityPackets = DO_OPTIMIZE;
 
     }
 }
