From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Sun, 26 Mar 2023 13:21:35 +0900
Subject: [PATCH] Implement-FixMySpawnR-Configuration


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 6f27085885aa274062fc2c659132fc7321bab0a1..8713a94bcd7a4f8cab0efa9779c9d54833afec72 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -44,4 +44,12 @@ public class GlobalConfiguration extends ConfigurationPart {
         public boolean doNotSendUselessEntityPackets = !Boolean.getBoolean("Plazma.disableConfigOptimization");
 
     }
+
+    public FixMySpawnR fixMySpawnR;
+    public class FixMySpawnR extends ConfigurationPart {
+
+        public boolean enabled = !Boolean.getBoolean("Plazma.disableConfigOptimization");
+        public int timerTimeOut = 0;
+
+    }
 }
