From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Mon, 4 Dec 2023 23:19:46 +0900
Subject: [PATCH] Optimize spigot event bus


diff --git a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
index 23594fb7eb4b2f33146592866608c2858ef23937..94eecf374f1e6b00556a380fd28376f720d61e8e 100644
--- a/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
+++ b/src/main/java/io/papermc/paper/plugin/manager/PaperEventManager.java
@@ -39,12 +39,19 @@ class PaperEventManager {
         // Plazma start - Skip event if no listeners
         RegisteredListener[] listeners = event.getHandlers().getRegisteredListeners();
         if (listeners.length == 0) return;
-        // Plazma end
-        if (event.isAsynchronous() && this.server.isPrimaryThread()) {
-            throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
-        } else if (!event.isAsynchronous() && !this.server.isPrimaryThread() && !this.server.isStopping()) {
-            throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
+
+        // Optimize Spigot event bus
+        if (event.asynchronous() == net.kyori.adventure.util.TriState.NOT_SET) {
+            boolean async = event.isAsynchronous();
+            boolean primary = this.server.isPrimaryThread();
+
+            if (async && primary)
+                throw new IllegalStateException(event.getEventName() + " may only be triggered asynchronously.");
+
+            if (!async && !primary && !this.server.isStopping())
+                throw new IllegalStateException(event.getEventName() + " may only be triggered synchronously.");
         }
+        // Plazma end
 
         for (RegisteredListener registration : listeners) {
             if (!registration.getPlugin().isEnabled()) {
