From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Thu, 11 Jan 2024 00:03:20 +0900
Subject: [PATCH] Fix build


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
index 935dac757280731bfeb0a8f033cbe315ecac46da..038f370ac2cb768e14fe7605b32b2ac811c33b8f 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -60,20 +60,20 @@ public class DedicatedPlayerList extends PlayerList {
     }
 
     private void saveIpBanList() {
-        try {
+        //try { // Plazma - Build fix
             this.getIpBans().save();
-        } catch (IOException var2) {
+        /*} catch (IOException var2) { // Plazma - Build fix
             LOGGER.warn("Failed to save ip banlist: ", (Throwable)var2);
-        }
+        }*/ // Plazma - Build fix
 
     }
 
     private void saveUserBanList() {
-        try {
+        //try { // Plazma - Build fix
             this.getBans().save();
-        } catch (IOException var2) {
+        /*} catch (IOException var2) { // Plazma - Build fix
             LOGGER.warn("Failed to save user banlist: ", (Throwable)var2);
-        }
+        }*/ // Plazma - Build fix
 
     }
 
