From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Thu, 11 Jan 2024 00:03:20 +0900
Subject: [PATCH] Fix build


diff --git a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
index 935dac757280731bfeb0a8f033cbe315ecac46da..038f370ac2cb768e14fe7605b32b2ac811c33b8f 100644
--- a/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
+++ b/src/main/java/net/minecraft/server/dedicated/DedicatedPlayerList.java
@@ -60,20 +60,20 @@ public class DedicatedPlayerList extends PlayerList {
     }
 
     private void saveIpBanList() {
-        try {
+        //try { // Plazma - Build fix
             this.getIpBans().save();
-        } catch (IOException var2) {
+        /*} catch (IOException var2) { // Plazma - Build fix
             LOGGER.warn("Failed to save ip banlist: ", (Throwable)var2);
-        }
+        }*/ // Plazma - Build fix
 
     }
 
     private void saveUserBanList() {
-        try {
+        //try { // Plazma - Build fix
             this.getBans().save();
-        } catch (IOException var2) {
+        /*} catch (IOException var2) { // Plazma - Build fix
             LOGGER.warn("Failed to save user banlist: ", (Throwable)var2);
-        }
+        }*/ // Plazma - Build fix
 
     }
 
diff --git a/src/main/java/net/minecraft/server/players/StoredUserList.java b/src/main/java/net/minecraft/server/players/StoredUserList.java
index b3d4f579fab96f572811ed0e33ecbb5204475391..a35340b9231bc3009ca9449fa6ea8f8ae6929146 100644
--- a/src/main/java/net/minecraft/server/players/StoredUserList.java
+++ b/src/main/java/net/minecraft/server/players/StoredUserList.java
@@ -61,11 +61,11 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
     public void add(V entry) {
         this.map.put(this.getKeyForUser(entry.getUser()), entry);
 
-        try {
+        //try { // Plazma - Fix build
             this.save();
-        } catch (IOException ioexception) {
+        /*} catch (IOException ioexception) { // Plazma - Fix build
             StoredUserList.LOGGER.warn("Could not save the list after adding a user.", ioexception);
-        }
+        }*/ // Plazma - Fix build
 
     }
 
@@ -83,11 +83,11 @@ public abstract class StoredUserList<K, V extends StoredUserEntry<K>> {
     public void remove(K key) {
         this.map.remove(this.getKeyForUser(key));
 
-        try {
+        //try { // Plazma - Fix build
             this.save();
-        } catch (IOException ioexception) {
+        /*} catch (IOException ioexception) { // Plazma - Fix build
             StoredUserList.LOGGER.warn("Could not save the list after removing a user.", ioexception);
-        }
+        }*/ // Plazma - Fix build
 
     }
 
