From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Thu, 18 May 2023 16:21:37 +0900
Subject: [PATCH] Let Me Despawn Configuration


diff --git a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
index a1b55bb5431d1712fc06da0cf06857a986efeea0..116384c1fe27a868bcbe1d21f0b2723c73b9170c 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
@@ -131,4 +131,11 @@ public class LevelConfigurations extends ConfigurationPart {
         }
 
     }
+
+    public LetMeDespawn letMeDespawn;
+    public class LetMeDespawn extends ConfigurationPart {
+
+        public boolean enabled = false;
+
+    }
 }
