From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ipecter <ipectert@gmail.com>
Date: Sat, 9 Mar 2024 16:55:06 +0900
Subject: [PATCH] CarpetFixes-Configuration


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 3cfb9357d0c61e84488c2c60f073fa12df942b8e..05a601ff9dd68c08d257e76594bb9e3f5f9a5e95 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -93,4 +93,12 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     }
 
+    public CarpetFixes carpetFixes;
+
+    public class CarpetFixes extends ConfigurationPart {
+
+        public boolean enabled = OPTIMIZE;
+
+    }
+
 }
