From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: AlphaKR93 <dev@alpha93.kr>
Date: Sun, 27 Oct 2024 00:16:25 +0900
Subject: [PATCH] Optimize advancement criteria triggering


diff --git a/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java b/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
index ebbad97920df3d1645637e646a98e16cb04d361e..a149c949a042e1d1e504d9f5ce0bfa6df92db2fc 100644
--- a/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
+++ b/src/main/java/net/minecraft/advancements/critereon/InventoryChangeTrigger.java
@@ -22,7 +22,21 @@ public class InventoryChangeTrigger extends SimpleCriterionTrigger<InventoryChan
         return InventoryChangeTrigger.TriggerInstance.CODEC;
     }
 
+    // Plazma start - Optimize advancement criteria triggering
+    public static int idleTick;
+    private int ticksSkipped;
+
+    private boolean shouldTick() {
+        if (++this.ticksSkipped <= idleTick) return false;
+
+        this.ticksSkipped = 0;
+        return true;
+    }
+    // Plazma end - Optimize advancement criteria triggering
+
     public void trigger(ServerPlayer player, Inventory inventory, ItemStack stack) {
+        if (idleTick > 0 && !this.shouldTick()) return; // Plazma - Optimize advancement criteria triggering
+
         int i = 0;
         int j = 0;
         int k = 0;
diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 9de9b3f2148216df41258d503fa1de896cca4771..49bce6e28e12f3729cab5628cf3e0f508a56d0d7 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -134,4 +134,16 @@ public class GlobalConfiguration extends ConfigurationPart {
 
     }
 
+    public Advancements advancements;
+    public class Advancements extends ConfigurationPart {
+
+        int criteriaTriggerIdleTick = OPTIMIZE ? 5 : 0;
+
+        @PostProcess
+        void post() {
+            net.minecraft.advancements.critereon.InventoryChangeTrigger.idleTick = this.criteriaTriggerIdleTick;
+        }
+
+    }
+
 }
