From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Wed, 6 Sep 2023 15:40:34 +0900
Subject: [PATCH] CarpetFixes-Configuration


diff --git a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
index 754ed13a04631c69e3fb2421a12b17d0a6f732c6..b552cb547fef8b4d87cb68e410d1b8fd669a9d4a 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/GlobalConfiguration.java
@@ -70,4 +70,12 @@ public class GlobalConfiguration extends ConfigurationPart {
         public int timerTimeOut = 0;
 
     }
+
+    public CarpetFixes carpetFixes;
+
+    public class CarpetFixes extends ConfigurationPart {
+
+        public boolean enabled = DO_OPTIMIZE;
+
+    }
 }
diff --git a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
index a1b55bb5431d1712fc06da0cf06857a986efeea0..4c4cb2a6a4730b934931c78090ff31b9e8d40202 100644
--- a/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
+++ b/src/main/java/org/plazmamc/plazma/configurations/LevelConfigurations.java
@@ -131,4 +131,12 @@ public class LevelConfigurations extends ConfigurationPart {
         }
 
     }
+
+    public CarpetFixes carpetFixes;
+
+    public class CarpetFixes extends ConfigurationPart {
+
+        public boolean enabled = DO_OPTIMIZE;
+
+    }
 }
