--- a/net/minecraft/world/level/storage/loot/entries/LootPoolSingletonContainer.java
+++ b/net/minecraft/world/level/storage/loot/entries/LootPoolSingletonContainer.java
@@ -49,8 +_,8 @@
         Instance<T> instance
     ) {
         return instance.group(
-                Codec.INT.optionalFieldOf("weight", Integer.valueOf(1)).forGetter(container -> container.weight),
-                Codec.INT.optionalFieldOf("quality", Integer.valueOf(0)).forGetter(lootPoolSingletonContainer -> lootPoolSingletonContainer.quality)
+                Codec.INT.optionalFieldOf("weight", 1).forGetter(container -> container.weight),
+                Codec.INT.optionalFieldOf("quality", 0).forGetter(lootPoolSingletonContainer -> lootPoolSingletonContainer.quality)
             )
             .and(commonFields(instance).t1())
             .and(LootItemFunctions.ROOT_CODEC.listOf().optionalFieldOf("functions", List.of()).forGetter(container -> container.functions));
