--- a/net/minecraft/data/worldgen/ProcessorLists.java
+++ b/net/minecraft/data/worldgen/ProcessorLists.java
@@ -115,27 +_,27 @@
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.NORTH, true)
+                                    .setValue(IronBarsBlock.SOUTH, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.NORTH, true)
+                                .setValue(IronBarsBlock.SOUTH, true)
                         ),
                         new ProcessorRule(
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.EAST, true)
+                                    .setValue(IronBarsBlock.WEST, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.EAST, true)
+                                .setValue(IronBarsBlock.WEST, true)
                         ),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.WHEAT, 0.3F), AlwaysTrueTest.INSTANCE, Blocks.CARROTS.defaultBlockState()),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.WHEAT, 0.2F), AlwaysTrueTest.INSTANCE, Blocks.POTATOES.defaultBlockState()),
@@ -165,27 +_,27 @@
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.NORTH, true)
+                                    .setValue(IronBarsBlock.SOUTH, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.NORTH, true)
+                                .setValue(IronBarsBlock.SOUTH, true)
                         ),
                         new ProcessorRule(
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.EAST, true)
+                                    .setValue(IronBarsBlock.WEST, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.EAST, true)
+                                .setValue(IronBarsBlock.WEST, true)
                         ),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.WHEAT, 0.1F), AlwaysTrueTest.INSTANCE, Blocks.MELON_STEM.defaultBlockState())
                     )
@@ -215,27 +_,27 @@
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.NORTH, true)
+                                    .setValue(IronBarsBlock.SOUTH, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.NORTH, true)
+                                .setValue(IronBarsBlock.SOUTH, true)
                         ),
                         new ProcessorRule(
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.EAST, true)
+                                    .setValue(IronBarsBlock.WEST, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.EAST, true)
+                                .setValue(IronBarsBlock.WEST, true)
                         ),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.WHEAT, 0.1F), AlwaysTrueTest.INSTANCE, Blocks.CARROTS.defaultBlockState()),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.WHEAT, 0.8F), AlwaysTrueTest.INSTANCE, Blocks.POTATOES.defaultBlockState())
@@ -258,7 +_,7 @@
                         new ProcessorRule(
                             new BlockMatchTest(Blocks.CAMPFIRE),
                             AlwaysTrueTest.INSTANCE,
-                            Blocks.CAMPFIRE.defaultBlockState().setValue(CampfireBlock.LIT, Boolean.valueOf(false))
+                            Blocks.CAMPFIRE.defaultBlockState().setValue(CampfireBlock.LIT, false)
                         ),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.COBBLESTONE, 0.08F), AlwaysTrueTest.INSTANCE, Blocks.COBWEB.defaultBlockState()),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.SPRUCE_LOG, 0.08F), AlwaysTrueTest.INSTANCE, Blocks.COBWEB.defaultBlockState()),
@@ -267,27 +_,27 @@
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.NORTH, true)
+                                    .setValue(IronBarsBlock.SOUTH, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.NORTH, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.SOUTH, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.NORTH, true)
+                                .setValue(IronBarsBlock.SOUTH, true)
                         ),
                         new ProcessorRule(
                             new BlockStateMatchTest(
                                 Blocks.GLASS_PANE
                                     .defaultBlockState()
-                                    .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                    .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                    .setValue(IronBarsBlock.EAST, true)
+                                    .setValue(IronBarsBlock.WEST, true)
                             ),
                             AlwaysTrueTest.INSTANCE,
                             Blocks.BROWN_STAINED_GLASS_PANE
                                 .defaultBlockState()
-                                .setValue(IronBarsBlock.EAST, Boolean.valueOf(true))
-                                .setValue(IronBarsBlock.WEST, Boolean.valueOf(true))
+                                .setValue(IronBarsBlock.EAST, true)
+                                .setValue(IronBarsBlock.WEST, true)
                         ),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.WHEAT, 0.3F), AlwaysTrueTest.INSTANCE, Blocks.PUMPKIN_STEM.defaultBlockState()),
                         new ProcessorRule(new RandomBlockMatchTest(Blocks.WHEAT, 0.2F), AlwaysTrueTest.INSTANCE, Blocks.POTATOES.defaultBlockState())
@@ -834,17 +_,17 @@
                         new ProcessorRule(
                             new RandomBlockMatchTest(Blocks.WAXED_COPPER_BULB, 0.1F),
                             AlwaysTrueTest.INSTANCE,
-                            Blocks.WAXED_OXIDIZED_COPPER_BULB.defaultBlockState().setValue(CopperBulbBlock.LIT, Boolean.valueOf(true))
+                            Blocks.WAXED_OXIDIZED_COPPER_BULB.defaultBlockState().setValue(CopperBulbBlock.LIT, true)
                         ),
                         new ProcessorRule(
                             new RandomBlockMatchTest(Blocks.WAXED_COPPER_BULB, 0.33333334F),
                             AlwaysTrueTest.INSTANCE,
-                            Blocks.WAXED_WEATHERED_COPPER_BULB.defaultBlockState().setValue(CopperBulbBlock.LIT, Boolean.valueOf(true))
+                            Blocks.WAXED_WEATHERED_COPPER_BULB.defaultBlockState().setValue(CopperBulbBlock.LIT, true)
                         ),
                         new ProcessorRule(
                             new RandomBlockMatchTest(Blocks.WAXED_COPPER_BULB, 0.5F),
                             AlwaysTrueTest.INSTANCE,
-                            Blocks.WAXED_EXPOSED_COPPER_BULB.defaultBlockState().setValue(CopperBulbBlock.LIT, Boolean.valueOf(true))
+                            Blocks.WAXED_EXPOSED_COPPER_BULB.defaultBlockState().setValue(CopperBulbBlock.LIT, true)
                         )
                     )
                 ),
