From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: IPECTER <ipectert@gmail.com>
Date: Sat, 13 Sep 2025 18:17:51 +0900
Subject: [PATCH] Rebrand


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index b06c2c4aa77edafb374f7cf0406cf4d29c6e7f9f..2a1720cf10006893ca7d8b672422460292743b81 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -98,6 +98,11 @@ public class Main {
             }
             */ // CraftBukkit end
         try {
+            // Plazma start - Fork-friendly Rebranding
+            String logo = org.plazmamc.plazma.ServerBrandConstants.ASCII_LOGO;
+            //noinspection ConstantValue
+            if (logo != null) System.out.println(logo);
+            // Plazma end - Fork-friendly Rebranding
 
             Path path = (Path) optionSet.valueOf("pidFile"); // CraftBukkit
             if (path != null) {
diff --git a/net/minecraft/server/dedicated/DedicatedServerProperties.java b/net/minecraft/server/dedicated/DedicatedServerProperties.java
index b286dc17cda16fca3af9374d2a4a8bd137c24450..22547894240501dab0bc478be957df7f926dbfd5 100644
--- a/net/minecraft/server/dedicated/DedicatedServerProperties.java
+++ b/net/minecraft/server/dedicated/DedicatedServerProperties.java
@@ -54,7 +54,7 @@ public class DedicatedServerProperties extends Settings<DedicatedServerPropertie
     public final String serverName = this.get("server-name", "Unknown Server"); // Purpur - Bring back server name
     public final boolean pvp = this.get("pvp", true);
     public final boolean allowFlight = this.get("allow-flight", false);
-    public final String motd = this.get("motd", "A Minecraft Server");
+    public final String motd = this.get("motd", "A " + org.plazmamc.plazma.ServerBrandConstants.BRAND_NAME + " Server");
     public final String bugReportLink = this.get("bug-report-link", "");
     public final boolean forceGameMode = this.get("force-gamemode", false);
     public final boolean enforceWhitelist = this.get("enforce-whitelist", false);
diff --git a/net/minecraft/server/gui/MinecraftServerGui.java b/net/minecraft/server/gui/MinecraftServerGui.java
index 614c7d9f673c926562acc8fa3b3788623900db41..cd4947a0ac3eb6396a3921d20e41c288e345f6a0 100644
--- a/net/minecraft/server/gui/MinecraftServerGui.java
+++ b/net/minecraft/server/gui/MinecraftServerGui.java
@@ -51,7 +51,7 @@ public class MinecraftServerGui extends JComponent {
         } catch (Exception var3) {
         }
 
-        final JFrame jFrame = new JFrame("Purpur Minecraft server"); // Purpur - Improve GUI
+        final JFrame jFrame = new JFrame(org.plazmamc.plazma.ServerBrandConstants.BRAND_NAME + " Minecraft server"); // Purpur - Improve GUI // Plazma - Rebrand
         final MinecraftServerGui minecraftServerGui = new MinecraftServerGui(server);
         jFrame.setDefaultCloseOperation(2);
         jFrame.add(minecraftServerGui);
@@ -59,7 +59,7 @@ public class MinecraftServerGui extends JComponent {
         jFrame.setLocationRelativeTo(null);
         jFrame.setVisible(true);
         // Paper start - Improve ServerGUI
-        jFrame.setName("Purpur Minecraft server"); // Purpur - Improve GUI
+        jFrame.setName(org.plazmamc.plazma.ServerBrandConstants.BRAND_NAME + " Minecraft server"); // Purpur - Improve GUI // Plazma - Rebrand
         try {
             jFrame.setIconImage(javax.imageio.ImageIO.read(java.util.Objects.requireNonNull(MinecraftServerGui.class.getClassLoader().getResourceAsStream("logo.png"))));
         } catch (java.io.IOException ignore) {
@@ -69,7 +69,7 @@ public class MinecraftServerGui extends JComponent {
             @Override
             public void windowClosing(WindowEvent event) {
                 if (!minecraftServerGui.isClosing.getAndSet(true)) {
-                    jFrame.setTitle("Purpur Minecraft server - shutting down!"); // Purpur - Improve GUI
+                    jFrame.setTitle(org.plazmamc.plazma.ServerBrandConstants.BRAND_NAME + " Minecraft server - shutting down!"); // Purpur - Improve GUI // Rebrand
                     server.halt(true);
                     minecraftServerGui.runFinalizers();
                 }
@@ -234,7 +234,7 @@ public class MinecraftServerGui extends JComponent {
 
     // Paper start - Add onboarding message for initial server start
     private JComponent buildOnboardingPanel() {
-        String onboardingLink = "https://docs.papermc.io/paper/next-steps";
+        String onboardingLink = org.plazmamc.plazma.ServerBrandConstants.USAGE_GUIDE; // Plazma - Rebrand
         JPanel jPanel = new JPanel();
 
         javax.swing.JLabel jLabel = new javax.swing.JLabel("If you need help setting up your server you can visit:");
