--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -172,7 +_,7 @@
         // Paper - replace random
     }
     // Paper end - Share random for entities to make them more random
-    public org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason; // Paper - Entity#getEntitySpawnReason
+    public @Nullable org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason spawnReason; // Paper - Entity#getEntitySpawnReason // Plazma - Null safety
 
     public boolean collisionLoadChunks = false; // Paper
     private @Nullable org.bukkit.craftbukkit.entity.CraftEntity bukkitEntity;
