--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -4421,7 +_,7 @@
 
         BlockState blockState = this.level().getBlockState(pos);
         if (blockState.getBlock() instanceof BedBlock) {
-            this.level().setBlock(pos, blockState.setValue(BedBlock.OCCUPIED, Boolean.valueOf(true)), 3);
+            this.level().setBlock(pos, blockState.setValue(BedBlock.OCCUPIED, true), 3);
         }
 
         this.setPose(Pose.SLEEPING);
@@ -4444,7 +_,7 @@
             BlockState blockState = this.level().getBlockState(blockPos);
             if (blockState.getBlock() instanceof BedBlock) {
                 Direction direction = blockState.getValue(BedBlock.FACING);
-                this.level().setBlock(blockPos, blockState.setValue(BedBlock.OCCUPIED, Boolean.valueOf(false)), 3);
+                this.level().setBlock(blockPos, blockState.setValue(BedBlock.OCCUPIED, false), 3);
                 Vec3 vec31 = BedBlock.findStandUpPosition(this.getType(), this.level(), blockPos, direction, this.getYRot()).orElseGet(() -> {
                     BlockPos blockPos1 = blockPos.above();
                     return new Vec3(blockPos1.getX() + 0.5, blockPos1.getY() + 0.1, blockPos1.getZ() + 0.5);
