--- a/net/minecraft/world/entity/Mob.java
+++ b/net/minecraft/world/entity/Mob.java
@@ -884,9 +_,9 @@
         // Paper end - Allow nerfed mobs to jump and float
         ProfilerFiller profilerFiller = Profiler.get();
         profilerFiller.push("sensing");
-        this.sensing.tick();
         profilerFiller.pop();
         int i = this.tickCount + this.getId();
+        if (i % this.level().plazmaConfig().entity.sensorTick == 0) this.sensing.tick(); // Plazma - Moved down; Configurable entity sensor tick
         if (i % 2 != 0 && this.tickCount > 1) {
             profilerFiller.push("targetSelector");
             this.targetSelector.tickRunningGoals(false);
