--- a/src/test/java/io/papermc/paper/permissions/MinecraftCommandPermissionsTest.java
+++ b/src/test/java/io/papermc/paper/permissions/MinecraftCommandPermissionsTest.java
@@ -78,6 +_,13 @@
         "minecraft.command.gamemode.survival",
         "minecraft.command.gamemode.survival.other",
         // Purpur end - Skip junit tests for purpur commands
+
+        // Plazma start - Skip junit tests for plazma commands
+        "minecraft.command.gc",
+        "minecraft.command.heal",
+        "minecraft.command.heal.other",
+        // Plazma end - Skip junit tests for plazma commands
+
         "minecraft.command.selector"
     );
 
