--- a/ca/spottedleaf/moonrise/patches/chunk_system/queue/ChunkUnloadQueue.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/queue/ChunkUnloadQueue.java
@@ -123,8 +_,8 @@
                     final JsonObject coordinateJson = new JsonObject();
                     coordinates.add(coordinateJson);
 
-                    coordinateJson.addProperty("chunkX", Integer.valueOf(CoordinateUtils.getChunkX(coordinate)));
-                    coordinateJson.addProperty("chunkZ", Integer.valueOf(CoordinateUtils.getChunkZ(coordinate)));
+                    coordinateJson.addProperty("chunkX", CoordinateUtils.getChunkX(coordinate));
+                    coordinateJson.addProperty("chunkZ", CoordinateUtils.getChunkZ(coordinate));
                 }
             }
         }
