--- a/net/minecraft/world/level/storage/loot/functions/EnchantedCountIncreaseFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/EnchantedCountIncreaseFunction.java
@@ -31,7 +_,7 @@
                 instance.group(
                     Enchantment.CODEC.fieldOf("enchantment").forGetter(enchantedCountIncreaseFunction -> enchantedCountIncreaseFunction.enchantment),
                     NumberProviders.CODEC.fieldOf("count").forGetter(enchantedCountIncreaseFunction -> enchantedCountIncreaseFunction.value),
-                    Codec.INT.optionalFieldOf("limit", Integer.valueOf(0)).forGetter(enchantedCountIncreaseFunction -> enchantedCountIncreaseFunction.limit)
+                    Codec.INT.optionalFieldOf("limit", 0).forGetter(enchantedCountIncreaseFunction -> enchantedCountIncreaseFunction.limit)
                 )
             )
             .apply(instance, EnchantedCountIncreaseFunction::new)
