--- a/net/minecraft/world/level/storage/loot/functions/EnchantRandomlyFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/EnchantRandomlyFunction.java
@@ -34,7 +_,7 @@
                         .optionalFieldOf("options")
                         .forGetter(enchantRandomlyFunction -> enchantRandomlyFunction.options),
                     Codec.BOOL
-                        .optionalFieldOf("only_compatible", Boolean.valueOf(true))
+                        .optionalFieldOf("only_compatible", true)
                         .forGetter(enchantRandomlyFunction -> enchantRandomlyFunction.onlyCompatible)
                 )
             )
