--- a/net/minecraft/world/level/storage/loot/functions/ExplorationMapFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/ExplorationMapFunction.java
@@ -40,10 +_,10 @@
                     MapDecorationType.CODEC
                         .optionalFieldOf("decoration", DEFAULT_DECORATION)
                         .forGetter(explorationMapFunction -> explorationMapFunction.mapDecoration),
-                    Codec.BYTE.optionalFieldOf("zoom", Byte.valueOf((byte)2)).forGetter(explorationMapFunction -> explorationMapFunction.zoom),
-                    Codec.INT.optionalFieldOf("search_radius", Integer.valueOf(50)).forGetter(explorationMapFunction -> explorationMapFunction.searchRadius),
+                    Codec.BYTE.optionalFieldOf("zoom", (byte) 2).forGetter(explorationMapFunction -> explorationMapFunction.zoom),
+                    Codec.INT.optionalFieldOf("search_radius", 50).forGetter(explorationMapFunction -> explorationMapFunction.searchRadius),
                     Codec.BOOL
-                        .optionalFieldOf("skip_existing_chunks", Boolean.valueOf(true))
+                        .optionalFieldOf("skip_existing_chunks", true)
                         .forGetter(explorationMapFunction -> explorationMapFunction.skipKnownStructures)
                 )
             )
