--- a/net/minecraft/world/level/storage/loot/functions/SetAttributesFunction.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetAttributesFunction.java
@@ -31,7 +_,7 @@
             .and(
                 instance.group(
                     SetAttributesFunction.Modifier.CODEC.listOf().fieldOf("modifiers").forGetter(function -> function.modifiers),
-                    Codec.BOOL.optionalFieldOf("replace", Boolean.valueOf(true)).forGetter(function -> function.replace)
+                    Codec.BOOL.optionalFieldOf("replace", true).forGetter(function -> function.replace)
                 )
             )
             .apply(instance, SetAttributesFunction::new)
