--- a/net/minecraft/world/level/storage/loot/functions/SetContainerLootTable.java
+++ b/net/minecraft/world/level/storage/loot/functions/SetContainerLootTable.java
@@ -23,7 +_,7 @@
             .and(
                 instance.group(
                     ResourceKey.codec(Registries.LOOT_TABLE).fieldOf("name").forGetter(setContainerLootTable -> setContainerLootTable.name),
-                    Codec.LONG.optionalFieldOf("seed", Long.valueOf(0L)).forGetter(setContainerLootTable -> setContainerLootTable.seed),
+                    Codec.LONG.optionalFieldOf("seed", 0L).forGetter(setContainerLootTable -> setContainerLootTable.seed),
                     BuiltInRegistries.BLOCK_ENTITY_TYPE.holderByNameCodec().fieldOf("type").forGetter(setContainerLootTable -> setContainerLootTable.type)
                 )
             )
